// @zen-browser/surfer/dist/commands/build.js
'use strict'
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.build = void 0
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const execa_1 = __importDefault(require('execa'))
const node_fs_1 = require('node:fs')
const node_path_1 = require('node:path')
const __1 = require('..')
const constants_1 = require('../constants')
const mozconfig_1 = require('../constants/mozconfig')
const log_1 = require('../log')
const patch_check_1 = require('../middleware/patch-check')
const utils_1 = require('../utils')
const platform = {
  win32: 'windows',
  darwin: 'macos',
  linux: 'linux',
}
const applyConfig = async (os) => {
  log_1.log.info('Applying mozconfig...')
  const brandingKey = utils_1.dynamicConfig.get('brand')
  let changeset
  try {
    // Retrieve changeset
    const { stdout } = await (0, execa_1.default)('git', ['rev-parse', 'HEAD'])
    changeset = stdout.trim()
  } catch (error) {
    log_1.log.warning(
      'Surfer expects that you are building your browser with git as your version control'
    )
    log_1.log.warning(
      'If you are using some other version control system, please migrate to git'
    )
    log_1.log.warning('Otherwise, you can setup git in this folder by running:')
    log_1.log.warning('   |git init|')
    throw error
  }
  const templateOptions = {
    name: __1.config.name,
    vendor: __1.config.name,
    appId: __1.config.appId,
    brandingDir: (0, node_fs_1.existsSync)(
      (0, node_path_1.join)(
        constants_1.ENGINE_DIR,
        'browser',
        'branding',
        brandingKey
      )
    )
      ? 'browser/branding/' + brandingKey
      : 'browser/branding/unofficial',
    binName: __1.config.binaryName,
    changeset,
  }
  const commonConfig = (0, utils_1.stringTemplate)(
    (0, node_fs_1.readFileSync)(
      (0, node_path_1.resolve)(constants_1.CONFIGS_DIR, 'common', 'mozconfig')
    ).toString(),
    templateOptions
  )
  const osConfig = (0, utils_1.stringTemplate)(
    (0, node_fs_1.readFileSync)(
      (0, node_path_1.resolve)(constants_1.CONFIGS_DIR, os, 'mozconfig')
    ).toString(),
    templateOptions
  )
  // Allow a custom config to be placed in /mozconfig. This will not be committed
  // to origin
  let customConfig = (0, node_fs_1.existsSync)(
    (0, node_path_1.join)(process.cwd(), 'mozconfig')
  )
    ? (0, node_fs_1.readFileSync)(
        (0, node_path_1.join)(process.cwd(), 'mozconfig')
      ).toString()
    : ''
  customConfig = (0, utils_1.stringTemplate)(customConfig, templateOptions)
  const mergedConfig =
    `# This file is automatically generated. You should only modify this if you know what you are doing!\n\n` +
    commonConfig +
    '\n\n' +
    osConfig +
    '\n\n' +
    customConfig +
    '\n' +
    (0, mozconfig_1.internalMozconfg)(
      brandingKey,
      utils_1.dynamicConfig.get('buildMode')
    )
  ;(0, node_fs_1.writeFileSync)(
    (0, node_path_1.resolve)(constants_1.ENGINE_DIR, 'mozconfig'),
    mergedConfig
  )
  log_1.log.info(
    `Building config based on a ${utils_1.dynamicConfig.get('buildMode')} mode`
  )
  log_1.log.info(`Config for this \`${os}\` build:`)
  mergedConfig.split('\n').map((ln) => {
    if (ln.startsWith('mk') || ln.startsWith('ac') || ln.startsWith('export'))
      log_1.log.info(
        `\t${ln
          .replace(/mk_add_options /, '')
          .replace(/ac_add_options /, '')
          .replace(/export /, '')}`
      )
  })
  // We need to install the browser display version inside of browser/config/version.txt
  // and browser/config/version_display.txt
  const brandingConfig = __1.config.brands[brandingKey]
  const version = brandingConfig?.release?.displayVersion || '1.0.0'
  log_1.log.debug(`Writing ${version} to the browser version files`)
  ;(0, node_fs_1.writeFileSync)(
    (0, node_path_1.join)(constants_1.ENGINE_DIR, 'browser/config/version.txt'),
    version
  )
  ;(0, node_fs_1.writeFileSync)(
    (0, node_path_1.join)(
      constants_1.ENGINE_DIR,
      'browser/config/version_display.txt'
    ),
    version
  )
}
const genericBuild = async (os, fast = false, jobs) => {
  log_1.log.info(`Building for "${os}"...`)
  log_1.log.warning(
    `If you get any dependency errors, try running |${__1.bin_name} bootstrap|.`
  )
  const buildOptions = ['build']
  if (fast) {
    buildOptions.push('faster')
  }
  if (jobs) {
    buildOptions.push(`-j${jobs}`)
  }
  log_1.log.debug(`Running with build options ${buildOptions.join(', ')}`)
  log_1.log.debug(
    `Mach exists: ${(0, node_fs_1.existsSync)(
      (0, node_path_1.join)(constants_1.ENGINE_DIR, 'mach')
    )}`
  )
  log_1.log.debug(
    `Mach contents: \n ${(0, node_fs_1.readFileSync)(
      (0, node_path_1.join)(constants_1.ENGINE_DIR, 'mach')
    )}\n\n===END===`
  )
  return (
    await (0, utils_1.configDispatch)('./mach', {
      args: buildOptions,
      cwd: constants_1.ENGINE_DIR,
      killOnError: true,
    })
  ).success
}
const parseDate = (d) => {
  d /= 1000
  const h = Math.floor(d / 3600)
  const m = Math.floor((d % 3600) / 60)
  const s = Math.floor((d % 3600) % 60)
  const hDisplay = h > 0 ? h + (h == 1 ? ' hour, ' : ' hours, ') : ''
  const mDisplay = m > 0 ? m + (m == 1 ? ' minute, ' : ' minutes, ') : ''
  const sDisplay = s > 0 ? s + (s == 1 ? ' second' : ' seconds') : ''
  return hDisplay + mDisplay + sDisplay
}
const success = (date) => {
  // mach handles the success messages
  console.log()
  log_1.log.info(`Total build time: ${parseDate(Date.now() - date)}.`)
}
const build = async (options) => {
  const d = Date.now()
  // Host build
  const prettyHost = platform[process.surferPlatform]
  if (constants_1.BUILD_TARGETS.includes(prettyHost)) {
    if (!options.skipPatchCheck) await (0, patch_check_1.patchCheck)()
    await applyConfig(prettyHost)
    if (process.env.SURFER_MOZCONFIG_ONLY) {
      log_1.log.info('Mozconfig applied. Exiting...')
      process.exit(0)
    }
    log_1.log.info('Starting build...')
    let exit = await genericBuild(prettyHost, options.ui, options.jobs)
    process.exit(exit ? 0 : 1)
  }
}
exports.build = build
