// @zen-browser/surfer/dist/commands/license-check.js
'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.licenseCheck = exports.createTask = exports.isValidLicense = void 0
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const promises_1 = require('node:fs/promises')
const node_path_1 = require('node:path')
const constants_1 = require('../constants')
const fs_1 = require('../utils/fs')
const task_list_1 = require('../utils/task-list')
const ignoredFiles = new RegExp(
  '.*\\.(json|patch|md|jpeg|png|gif|tiff|ico|woff2|dep)'
)
const licenseIgnore = new RegExp('(//|#) Ignore license in this file', 'g')
const fixableFiles = [
  { regex: new RegExp('.*\\.(mj|j|t)s'), comment: '// ', commentClose: '\n' },
  {
    regex: new RegExp('.*\\.css'),
    commentOpen: '/*\n',
    comment: ' * ',
    commentClose: '\n */',
  },
  {
    regex: new RegExp('.*\\.(html|svg|xml)'),
    commentOpen: '<!--\n',
    comment: '   - ',
    commentClose: '\n   -->',
  },
  {
    regex: new RegExp('.*\\.py|moz\\.build|jar\\.mn'),
    commentOpen: '#\n',
    comment: '# ',
    commentClose: '\n',
  },
]
async function isValidLicense(path) {
  const file = await (0, promises_1.readFile)(path, { encoding: 'utf8' })
  const contents = file.split('\n')
  // We need to grab the top 5 lines just in case there are newlines in the
  // comment blocks
  const lines = [
    contents[0],
    contents[1],
    contents[2],
    contents[3],
    contents[4],
  ].join('\n')
  const hasLicense =
    (lines.includes('the Mozilla Public') &&
      lines.includes('If a copy of the MPL was') &&
      lines.includes('http://mozilla.org/MPL/2.0/')) ||
    licenseIgnore.test(contents.join('\n')) ||
    (lines.includes('Any copyright is dedicated to the Public') &&
      lines.includes('https://creativecommons.org/publicdomain'))
  return hasLicense || path.endsWith('.min.js') || path.endsWith('.min.mjs')
}
exports.isValidLicense = isValidLicense
function createTask(path, noFix) {
  return {
    skip: () => ignoredFiles.test(path),
    name: path.replace(constants_1.SRC_DIR, ''),
    task: async () => {
      const contents = await (0, promises_1.readFile)(path, {
        encoding: 'utf8',
      })
      const contentsSplitNewline = contents.split('\n')
      const hasLicense = await isValidLicense(path)
      if (hasLicense) {
        return
      }
      const fixable = fixableFiles.find(({ regex }) => regex.test(path))
      if (!fixable || noFix) {
        throw new Error(
          `${path} does not have a license. Please add the source code header`
        )
      }
      const mplHeader = // eslint-disable-next-line unicorn/prefer-module
        await (0, promises_1.readFile)(
          (0, node_path_1.join)(__dirname, 'license-check.txt'),
          {
            encoding: 'utf8',
          }
        )
      const { comment, commentOpen, commentClose } = fixable
      let header = mplHeader
        .split('\n')
        .map((ln) => (comment || '') + ln)
        .join('\n')
      if (commentOpen) {
        header = commentOpen + header + commentClose
      }
      await (0, promises_1.writeFile)(
        path,
        header + '\n' + contentsSplitNewline.join('\n')
      )
    },
  }
}
exports.createTask = createTask
const licenseCheck = async (options) => {
  const files = await (0, fs_1.walkDirectory)(constants_1.SRC_DIR)
  await new task_list_1.TaskList(
    files.map((file) => createTask(file, !options.fix))
  )
    .onError('inline')
    .run()
}
exports.licenseCheck = licenseCheck
