// @zen-browser/surfer/dist/commands/package.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCurrentBrandName = exports.surferPackage = void 0;
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const node_fs_1 = require("node:fs");
const promises_1 = require("node:fs/promises");
const node_path_1 = require("node:path");
const __1 = require("..");
const constants_1 = require("../constants");
const log_1 = require("../log");
const utils_1 = require("../utils");
const browser_1 = require("./updates/browser");
const fs_extra_1 = require("fs-extra");
const machPath = (0, node_path_1.resolve)(constants_1.ENGINE_DIR, 'mach');
async function getLocales() {
    // locales/supported-languages is a list of locales divided by newlines
    // open the file and split it by newlines
    let localesText = await (0, fs_extra_1.readFile)('locales/supported-languages', 'utf-8');
    const languageMaps = await (0, fs_extra_1.readFile)('locales/language-maps', 'utf-8');
    // Language maps contains a list of locale mappings for specific locales
    // e.g. "nb:nb-NO" means that "nb" should be mapped to "nb-NO"
    const mappings = {};
    for (const line of languageMaps.split('\n')) {
        const [key, value] = line.split(':');
        if (key && value) {
            mappings[key.trim()] = value.trim();
        }
    }
    localesText = localesText
        .split('\n')
        .map((locale) => locale.trim())
        .filter((locale) => locale.length > 0)
        .map((locale) => (mappings[locale] ? mappings[locale] : locale))
        .join('\n');
    log_1.log.info(`Found locales:\n${localesText}`);
    return localesText.split('\n');
}
const surferPackage = async () => {
    const brandingKey = utils_1.dynamicConfig.get('brand');
    const brandingDetails = __1.config.brands[brandingKey];
    const version = brandingDetails.release.displayVersion;
    const channel = brandingKey || 'unofficial';
    log_1.log.debug("Creating the dist directory if it doesn't exist");
    if (!(0, node_fs_1.existsSync)(constants_1.DIST_DIR))
        await (0, promises_1.mkdir)(constants_1.DIST_DIR, { recursive: true });
    // The engine directory must have been downloaded for this to be valid
    // TODO: Make this a reusable function that can be used by everything
    if (!process.env.JUST_MAR) {
        if (!(0, node_fs_1.existsSync)(constants_1.ENGINE_DIR)) {
            log_1.log.error(`Unable to locate any source directories.\nRun |${__1.bin_name} download| to generate the source directory.`);
        }
        if (!(0, node_fs_1.existsSync)(machPath)) {
            log_1.log.error(`Cannot locate the 'mach' binary within ${constants_1.ENGINE_DIR}`);
        }
        const arguments_ = ['package'];
        log_1.log.info(`Packaging \`${__1.config.binaryName}\` with args ${JSON.stringify(arguments_.slice(1, 0))}...`);
        await (0, utils_1.dispatch)(machPath, arguments_, constants_1.ENGINE_DIR, true);
        log_1.log.info('Copying language packs');
        await (0, utils_1.dispatch)(machPath, ['package-multi-locale', '--locales', ...(await getLocales())], constants_1.ENGINE_DIR, true);
        log_1.log.info('Copying results up');
        log_1.log.debug('Indexing files to copy');
        const filesInMozillaDistrobution = await (0, promises_1.readdir)((0, node_path_1.join)(constants_1.OBJ_DIR, 'dist'), {
            withFileTypes: true,
        });
        const files = filesInMozillaDistrobution
            .filter((entry) => entry.isFile())
            .map((entry) => entry.name);
        for (const file of files) {
            const destinationFile = (0, node_path_1.join)(constants_1.DIST_DIR, file);
            log_1.log.debug(`Copying ${file}`);
            if ((0, node_fs_1.existsSync)(destinationFile))
                await (0, promises_1.unlink)(destinationFile);
            await (0, promises_1.copyFile)((0, node_path_1.join)(constants_1.OBJ_DIR, 'dist', file), destinationFile);
        }
        // Windows has some special dist files that are available within the dist
        // directory.
        if (process.surferPlatform == 'win32') {
            const installerDistributionDirectory = (0, node_path_1.join)(constants_1.OBJ_DIR, 'dist');
            if (!(0, node_fs_1.existsSync)(installerDistributionDirectory)) {
                log_1.log.error(`Could not find windows installer files located at '${installerDistributionDirectory}'`);
            }
            const installerDistributionDirectoryContents = await (0, promises_1.readdir)(installerDistributionDirectory, { withFileTypes: true });
            const windowsInstallerFiles = installerDistributionDirectoryContents
                .filter((entry) => entry.isFile())
                .map((entry) => entry.name);
            for (const file of windowsInstallerFiles) {
                let newFileName = file;
                // There are some special cases that I want to reformat the name for
                if (file.includes('.installer.exe')) {
                    newFileName = `${__1.config.binaryName}.installer.exe`;
                }
                if (file.includes('.installer-stub.exe')) {
                    newFileName = `${__1.config.binaryName}.installer.pretty.exe`;
                    log_1.log.warning(`The installer ${newFileName} requires that your binaries are available from the internet and everything is correctly configured. I recommend you ship '${__1.config.binaryName}.installer.exe' if you have not set this up correctly yet`);
                }
                // Actually copy
                const destinationFile = (0, node_path_1.join)(constants_1.DIST_DIR, newFileName);
                log_1.log.debug(`Copying ${file}`);
                if ((0, node_fs_1.existsSync)(destinationFile))
                    await (0, promises_1.unlink)(destinationFile);
                await (0, promises_1.copyFile)((0, node_path_1.join)(installerDistributionDirectory, file), destinationFile);
            }
        }
    }
    const marPath = await createMarFile(version, channel, brandingDetails.release.github);
    utils_1.dynamicConfig.set('marPath', marPath);
    await (0, browser_1.generateBrowserUpdateFiles)();
    log_1.log.info();
    log_1.log.info(`Output written to ${constants_1.DIST_DIR}`);
    log_1.log.success('Packaging complected!');
};
exports.surferPackage = surferPackage;
function getCurrentBrandName() {
    const brand = utils_1.dynamicConfig.get('brand');
    if (brand == 'unofficial') {
        return 'Nightly';
    }
    return __1.config.brands[brand].brandShortName;
}
exports.getCurrentBrandName = getCurrentBrandName;
async function createMarFile(version, channel, github) {
    log_1.log.info(`Creating mar file...`);
    let marBinary = (0, utils_1.windowsPathToUnix)((0, node_path_1.join)(constants_1.OBJ_DIR, 'dist/host/bin', 'mar'));
    if (process.platform == 'win32') {
        marBinary += '.exe';
    }
    // On macos this should be
    // <obj dir>/dist/${binaryName}/${brandFullName}.app and on everything else,
    // the contents of the folder <obj dir>/dist/${binaryName}
    const binary = process.surferPlatform == 'darwin'
        ? process.env.JUST_MAR
            ? (0, node_path_1.join)(constants_1.OBJ_DIR, 'dist', `${getCurrentBrandName()}.app`)
            : (0, node_path_1.join)(constants_1.OBJ_DIR, 'dist', __1.config.binaryName, `${getCurrentBrandName()}.app`)
        : (0, node_path_1.join)(constants_1.OBJ_DIR, 'dist', __1.config.binaryName);
    const marPath = (0, node_path_1.resolve)(constants_1.DIST_DIR, 'output.mar');
    log_1.log.debug(`Writing MAR to ${constants_1.DIST_DIR} from ${binary}`);
    await (0, utils_1.configDispatch)('./tools/update-packaging/make_full_update.sh', {
        args: [
            // The mar output location
            (0, utils_1.windowsPathToUnix)(constants_1.DIST_DIR),
            (0, utils_1.windowsPathToUnix)(binary),
        ],
        cwd: constants_1.ENGINE_DIR,
        env: {
            MOZ_PRODUCT_VERSION: version,
            MAR_CHANNEL_ID: channel,
            MAR: process.env.MAR ? (0, utils_1.windowsPathToUnix)(process.env.MAR) : marBinary,
        },
        shell: process.env.SURFER_SIGNING_MODE ? 'unix' : 'default',
    });
    return marPath;
}
