// @zen-browser/surfer/dist/commands/setup-project.js
'use strict'
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.shouldSkipOptionalCopy =
  exports.templateDirectory =
  exports.setupProject =
    void 0
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const node_fs_1 = require('node:fs')
const promises_1 = require('node:fs/promises')
const node_path_1 = require('node:path')
const prompts_1 = __importDefault(require('prompts'))
const constants_1 = require('../constants')
const log_1 = require('../log')
const utils_1 = require('../utils')
// =============================================================================
// User interaction portion
async function setupProject() {
  try {
    if ((0, node_fs_1.existsSync)(utils_1.configPath)) {
      log_1.log.warning(
        'There is already a config file. This will overwrite it!'
      )
      await (0, utils_1.delay)(1000)
    }
    if (utils_1.configPath.includes('.optional')) {
      log_1.log.error(
        'The text ".optional" cannot be in the path to your custom browser'
      )
      process.exit(1)
    }
    // Ask user for assorted information
    const { product } = await (0, prompts_1.default)({
      type: 'select',
      name: 'product',
      message: 'Select a product to fork',
      choices: [
        {
          title: 'Firefox stable',
          description: 'Releases around every 4 weeks, fairly stable',
          value: utils_1.SupportedProducts.Firefox,
        },
        {
          title: 'Firefox extended support (older)',
          description:
            'The extended support version of Firefox. Will receive security updates for a longer period of time and less frequent, bigger, feature updates',
          value: utils_1.SupportedProducts.FirefoxESR,
        },
        {
          title: 'Firefox developer edition (Not recommended)',
          description: 'Tracks firefox beta, with a few config tweaks',
          value: utils_1.SupportedProducts.FirefoxDevelopment,
        },
        {
          title: 'Firefox beta (Not recommended)',
          description: 'Updates every 4 weeks. It will have unresolved bugs',
          value: utils_1.SupportedProducts.FirefoxBeta,
        },
        {
          title: 'Firefox Nightly (Not recommended)',
          description:
            'Updates daily, with many bugs. Practically impossible to track',
          value: utils_1.SupportedProducts.FirefoxNightly,
        },
      ],
    })
    if (typeof product === 'undefined') return
    const productVersion = await (0, utils_1.getLatestFF)(product)
    const { version, name, appId, vendor, ui, binaryName } = await (0,
    prompts_1.default)([
      {
        type: 'text',
        name: 'version',
        message: 'Enter the version of this product',
        initial: productVersion,
      },
      {
        type: 'text',
        name: 'name',
        message: 'Enter a product name',
        initial: 'Example browser',
      },
      {
        type: 'text',
        name: 'binaryName',
        message: 'Enter the name of the binary',
        initial: 'example-browser',
      },
      {
        type: 'text',
        name: 'vendor',
        message: 'Enter a vendor',
        initial: 'Example company',
      },
      {
        type: 'text',
        name: 'appId',
        message: 'Enter an appid',
        initial: 'com.example.browser',
        // Horrible validation to make sure people don't chose something entirely wrong
        validate: (t) => t.includes('.'),
      },
      {
        type: 'select',
        name: 'ui',
        message: 'Select a ui mode template',
        choices: [
          {
            title: 'None',
            description:
              'No files for the ui will be created, we will let you find that out on your own',
            value: 'none',
          },
          {
            title: 'UserChrome',
            value: 'uc',
          },
          // TODO: We also need to add extension based theming like the version
          // used in Pulse Browser
        ],
      },
    ])
    const config = {
      name,
      vendor,
      appId,
      binaryName,
      version: { product, version },
      buildOptions: {
        windowsUseSymbolicLinks: false,
      },
    }
    await copyRequired()
    if (ui === 'uc') {
      await copyOptional(['browser/themes'])
    }
    ;(0, node_fs_1.writeFileSync)(
      utils_1.configPath,
      JSON.stringify(config, undefined, 2)
    )
    // Append important stuff to gitignore
    const gitignore = (0, node_path_1.join)(
      utils_1.projectDirectory,
      '.gitignore'
    )
    let gitignoreContents = ''
    if ((0, node_fs_1.existsSync)(gitignore)) {
      gitignoreContents = (0, node_fs_1.readFileSync)(gitignore).toString()
    }
    gitignoreContents +=
      '\n.dotbuild/\n.surfer/\nengine/\nfirefox-*/\nnode_modules/\n'
    ;(0, node_fs_1.writeFileSync)(gitignore, gitignoreContents)
    log_1.log.success(
      'Project setup complete!',
      '',
      `You can start downloading the Firefox source code by running |${constants_1.BIN_NAME} download|`,
      'Or you can follow the getting started guide at https://docs.surfer.dev/getting-started/overview/'
    )
  } catch (error) {
    log_1.log.error(error)
  }
}
exports.setupProject = setupProject
// =============================================================================
// Filesystem templating
// eslint-disable-next-line unicorn/prefer-module
exports.templateDirectory = (0, node_path_1.join)(
  __dirname,
  '../..',
  'template'
)
/**
 * Copy files from the template directory that have .optional in their path,
 * based on the function parameters
 *
 * @param files The files that should be coppied
 */
async function copyOptional(files) {
  const directoryContents = await (0, utils_1.walkDirectory)(
    exports.templateDirectory
  )
  for (const file of directoryContents) {
    if (shouldSkipOptionalCopy(file, files)) continue
    const outLocation = (0, node_path_1.join)(
      utils_1.projectDirectory,
      file.replace(exports.templateDirectory, '')
    ).replace('.optional', '')
    if (!(0, node_fs_1.existsSync)(outLocation)) {
      ;(0, node_fs_1.mkdirSync)((0, node_path_1.dirname)(outLocation), {
        recursive: true,
      })
      await (0, promises_1.copyFile)(file, outLocation)
    }
  }
}
/**
 * Used to determine if a file should be copied or not. This is exported only so
 * that it can be unit tested
 *
 * @param file The file that should be copied
 * @param files A list of files / directories that we want to match to
 * @returns If the file should be skipped
 *
 * @private
 */
function shouldSkipOptionalCopy(file, files) {
  // We want to skip copying this file if:
  // - It is not optional
  // - It is not in the files array
  return (
    !file.includes('.optional') ||
    !files.map((f) => file.includes(f)).some(Boolean)
  )
}
exports.shouldSkipOptionalCopy = shouldSkipOptionalCopy
/**
 * Copy all non-optional files from the template directory
 */
async function copyRequired() {
  const directoryContents = await (0, utils_1.walkDirectory)(
    exports.templateDirectory
  )
  for (const file of directoryContents) {
    if (file.includes('.optional')) continue
    const outLocation = (0, node_path_1.join)(
      utils_1.projectDirectory,
      file.replace(exports.templateDirectory, '')
    )
    if (!(0, node_fs_1.existsSync)(outLocation)) {
      ;(0, node_fs_1.mkdirSync)((0, node_path_1.dirname)(outLocation), {
        recursive: true,
      })
      await (0, promises_1.copyFile)(file, outLocation)
    }
  }
}
