// 7z-iterator/dist/cjs/nextEntry.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return nextEntry;
    }
});
var _calloncefn = /*#__PURE__*/ _interop_require_default(require("call-once-fn"));
var _extractbaseiterator = require("extract-base-iterator");
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _FileEntryts = /*#__PURE__*/ _interop_require_default(require("./FileEntry.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function nextEntry(iterator, callback) {
    var internalIter = iterator._iterator;
    if (!internalIter) {
        callback(new Error('iterator missing'));
        return;
    }
    var entry = null;
    entry = internalIter.next();
    var nextCallback = (0, _calloncefn.default)(function(err, entry) {
        // keep processing
        if (entry) iterator.push(nextEntry);
        err ? callback(err) : callback(null, entry ? {
            done: false,
            value: entry
        } : {
            done: true,
            value: null
        });
    });
    // done: signal iteration is complete (guard against stale lock)
    if (!iterator.lock || iterator.isDone() || !entry) return callback(null, {
        done: true,
        value: null
    });
    // Skip anti-files (these mark files to delete in delta archives)
    if (entry.isAntiFile) {
        iterator.push(nextEntry);
        return callback(null, null);
    }
    // Determine type from entry
    var type = entry.type;
    // Default modes (decimal values for Node 0.8 compatibility)
    // 0o755 = 493, 0o644 = 420
    var defaultMode = type === 'directory' ? 493 : 420;
    // Build attributes from 7z entry
    // mtime must be timestamp (number) for FileAttributes compatibility
    var mtimeDate = entry.mtime || new Date();
    var attributes = {
        path: entry.path.split(_path.default.sep).filter(Boolean).join(_path.default.sep),
        basename: entry.name,
        mtime: mtimeDate.getTime(),
        mode: entry.mode !== undefined ? entry.mode : defaultMode
    };
    switch(type){
        case 'directory':
            attributes.type = 'directory';
            return nextCallback(null, new _extractbaseiterator.DirectoryEntry(attributes));
        case 'link':
            {
                // For symlinks, the file content IS the symlink target path
                // Read the content to get the linkpath for SymbolicLinkEntry
                var parser = internalIter.getParser();
                var stream = parser.getEntryStream(entry);
                var chunks = [];
                stream.on('data', function(chunk) {
                    chunks.push(chunk);
                });
                stream.on('end', function() {
                    var linkpath = Buffer.concat(chunks).toString('utf8');
                    var linkAttributes = {
                        path: attributes.path,
                        mtime: attributes.mtime,
                        mode: attributes.mode,
                        linkpath: linkpath
                    };
                    nextCallback(null, new _extractbaseiterator.SymbolicLinkEntry(linkAttributes));
                });
                stream.on('error', function(streamErr) {
                    nextCallback(streamErr);
                });
                return;
            }
        case 'file':
            {
                attributes.type = 'file';
                attributes.size = entry.size;
                var parser1 = internalIter.getParser();
                var stream1 = parser1.getEntryStream(entry);
                return nextCallback(null, new _FileEntryts.default(attributes, stream1, iterator.lock, entry._canStream));
            }
    }
    return callback(new Error("Unrecognized entry type: ".concat(type)));
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
