// 7z-iterator/dist/cjs/SevenZipIterator.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return SevenZipIterator;
    }
});
var _extractbaseiterator = /*#__PURE__*/ _interop_require_wildcard(require("extract-base-iterator"));
var _fsremovecompat = require("fs-remove-compat");
var _gracefulfs = /*#__PURE__*/ _interop_require_default(require("graceful-fs"));
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _queuecb = /*#__PURE__*/ _interop_require_default(require("queue-cb"));
var _shorthash = /*#__PURE__*/ _interop_require_default(require("short-hash"));
var _tempsuffix = /*#__PURE__*/ _interop_require_default(require("temp-suffix"));
var _compatts = require("./compat.js");
var _streamToSourcets = /*#__PURE__*/ _interop_require_default(require("./lib/streamToSource.js"));
var _nextEntryts = /*#__PURE__*/ _interop_require_default(require("./nextEntry.js"));
var _indexts = require("./sevenz/codecs/index.js");
var _SevenZipParserts = require("./sevenz/SevenZipParser.js");
function _assert_this_initialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _call_super(_this, derived, args) {
    derived = _get_prototype_of(derived);
    return _possible_constructor_return(_this, _is_native_reflect_construct() ? Reflect.construct(derived, args || [], _get_prototype_of(_this).constructor) : derived.apply(_this, args));
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _get_prototype_of(o) {
    _get_prototype_of = Object.setPrototypeOf ? Object.getPrototypeOf : function getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _get_prototype_of(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _set_prototype_of(subClass, superClass);
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function _possible_constructor_return(self, call) {
    if (call && (_type_of(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assert_this_initialized(self);
}
function _set_prototype_of(o, p) {
    _set_prototype_of = Object.setPrototypeOf || function setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _set_prototype_of(o, p);
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _is_native_reflect_construct() {
    try {
        var result = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
    } catch (_) {}
    return (_is_native_reflect_construct = function() {
        return !!result;
    })();
}
/**
 * Iterator wrapper around SevenZipParser entries
 */ var EntryIterator = /*#__PURE__*/ function() {
    "use strict";
    function EntryIterator(parser) {
        _class_call_check(this, EntryIterator);
        this.index = 0;
        this.parser = parser;
        this.entries = parser.getEntries();
    }
    var _proto = EntryIterator.prototype;
    _proto.next = function next() {
        if (this.index >= this.entries.length) {
            return null;
        }
        return this.entries[this.index++];
    };
    _proto.getParser = function getParser() {
        return this.parser;
    };
    return EntryIterator;
}();
var SevenZipIterator = /*#__PURE__*/ function(BaseIterator) {
    "use strict";
    _inherits(SevenZipIterator, BaseIterator);
    function SevenZipIterator(source) {
        var options = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
        _class_call_check(this, SevenZipIterator);
        var _this;
        _this = _call_super(this, SevenZipIterator, [
            options
        ]);
        _this.lock = new _extractbaseiterator.Lock();
        _this.lock.onDestroy = function(err) {
            return _extractbaseiterator.default.prototype.end.call(_this, err);
        };
        var queue = new _queuecb.default(1);
        var cancelled = false;
        var archiveSource = null;
        var setup = function() {
            cancelled = true;
        };
        _this.processing.push(setup);
        // Set password (or clear if not provided)
        (0, _indexts.setPassword)(options.password || null);
        if (typeof source === 'string') {
            // File path input - use FileSource directly
            queue.defer(function(cb) {
                _gracefulfs.default.stat(source, function(statErr, stats) {
                    if (_this.done || cancelled) return;
                    if (statErr) return cb(statErr);
                    _gracefulfs.default.open(source, 'r', function(err, fd) {
                        if (_this.done || cancelled) return;
                        if (err) return cb(err);
                        archiveSource = new _SevenZipParserts.FileSource(fd, stats.size);
                        // Register cleanup for file descriptor
                        _this.lock.registerCleanup(function() {
                            _gracefulfs.default.closeSync(fd);
                        });
                        cb();
                    });
                });
            });
        } else {
            // Stream input - write to temp file for random access
            // Register cleanup for source stream
            var stream = source;
            _this.lock.registerCleanup(function() {
                var s = stream;
                if (typeof s.destroy === 'function') s.destroy();
            });
            var tempPath = _path.default.join((0, _compatts.tmpdir)(), '7z-iterator', (0, _shorthash.default)(process.cwd()), (0, _tempsuffix.default)('tmp.7z'));
            queue.defer(function(cb) {
                (0, _streamToSourcets.default)(source, {
                    tempPath: tempPath
                }, function(err, result) {
                    if (_this.done || cancelled) return;
                    if (err) return cb(err);
                    if (!result) return cb(new Error('No result from streamToSource'));
                    archiveSource = result.source;
                    // Register cleanup for file descriptor
                    _this.lock.registerCleanup(function() {
                        _gracefulfs.default.closeSync(result.fd);
                    });
                    // Register cleanup for temp file
                    _this.lock.registerCleanup(function() {
                        try {
                            (0, _fsremovecompat.rmSync)(result.tempPath);
                        } catch (_e) {
                        /* ignore */ }
                    });
                    cb();
                });
            });
        }
        // Parse and build iterator
        queue.defer(function(cb) {
            if (_this.done || cancelled) return;
            if (!archiveSource) return cb(new Error('No archive source'));
            var parser = new _SevenZipParserts.SevenZipParser(archiveSource);
            parser.parse(function(parseErr) {
                if (parseErr) {
                    cb(parseErr);
                    return;
                }
                try {
                    _this._iterator = new EntryIterator(parser);
                    cb();
                } catch (err) {
                    cb(err);
                }
            });
        });
        // start processing
        queue.await(function(err) {
            _this.processing.remove(setup);
            if (_this.done || cancelled) return;
            err ? _this.end(err) : _this.push(_nextEntryts.default);
        });
        return _this;
    }
    var _proto = SevenZipIterator.prototype;
    _proto.end = function end(err) {
        if (this.lock) {
            var lock = this.lock;
            this.lock = null; // Clear before release to prevent re-entrancy
            lock.err = err;
            lock.release();
        }
        // Don't call base end here - Lock.__destroy() handles it
        this._iterator = null;
    };
    /**
   * Check if streaming extraction is available for any folder in this archive.
   * Streaming is possible when folders use codecs like BZip2, Deflate, or Copy
   * that can decompress incrementally without buffering the entire input.
   *
   * @returns true if at least one folder supports streaming
   */ _proto.canStream = function canStream() {
        if (!this._iterator) return false;
        var parser = this._iterator.getParser();
        if (!parser) return false;
        var entries = parser.getEntries();
        var checkedFolders = {};
        for(var i = 0; i < entries.length; i++){
            var folderIndex = entries[i]._folderIndex;
            if (folderIndex >= 0 && checkedFolders[folderIndex] === undefined) {
                checkedFolders[folderIndex] = parser.canStreamFolder(folderIndex);
                if (checkedFolders[folderIndex]) {
                    return true;
                }
            }
        }
        return false;
    };
    /**
   * Get entries sorted for optimal streaming extraction.
   *
   * Entries are sorted by:
   * 1. Folder index (process one folder at a time)
   * 2. Stream index within folder (for solid block streaming)
   *
   * This ordering allows multi-file solid folders to stream with
   * O(largest file) memory instead of O(folder size).
   *
   * @returns Array of entries in streaming order
   */ _proto.getStreamingOrder = function getStreamingOrder() {
        if (!this._iterator) return [];
        var parser = this._iterator.getParser();
        if (!parser) return [];
        var entries = parser.getEntries();
        // Create a copy and sort for streaming order
        var sorted = [];
        for(var i = 0; i < entries.length; i++){
            sorted.push(entries[i]);
        }
        sorted.sort(function(a, b) {
            // First by folder index
            if (a._folderIndex !== b._folderIndex) {
                return a._folderIndex - b._folderIndex;
            }
            // Then by stream index within folder
            return a._streamIndexInFolder - b._streamIndexInFolder;
        });
        return sorted;
    };
    return SevenZipIterator;
}(_extractbaseiterator.default);
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
