// extract-base-iterator/dist/cjs/shared/EntryStream.js
/**
 * EntryStream - Simple readable stream for entry content
 *
 * Extends Readable to emit 'data', 'end', 'error' events.
 * Node 0.8 compatible via readable-stream polyfill.
 *
 * This base class is designed to be used by:
 * - zip-iterator
 * - 7z-iterator
 * - tar-iterator
 * - Any other archive iterator library
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return EntryStream;
    }
});
var _stream = /*#__PURE__*/ _interop_require_default(require("stream"));
function _assert_this_initialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _call_super(_this, derived, args) {
    derived = _get_prototype_of(derived);
    return _possible_constructor_return(_this, _is_native_reflect_construct() ? Reflect.construct(derived, args || [], _get_prototype_of(_this).constructor) : derived.apply(_this, args));
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _get_prototype_of(o) {
    _get_prototype_of = Object.setPrototypeOf ? Object.getPrototypeOf : function getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _get_prototype_of(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _set_prototype_of(subClass, superClass);
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _possible_constructor_return(self, call) {
    if (call && (_type_of(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assert_this_initialized(self);
}
function _set_prototype_of(o, p) {
    _set_prototype_of = Object.setPrototypeOf || function setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _set_prototype_of(o, p);
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _is_native_reflect_construct() {
    try {
        var result = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
    } catch (_) {}
    return (_is_native_reflect_construct = function() {
        return !!result;
    })();
}
// Use native streams when available, readable-stream only for Node 0.x
var major = +process.versions.node.split('.')[0];
var ReadableBase;
if (major > 0) {
    ReadableBase = _stream.default.Readable;
} else {
    ReadableBase = require('readable-stream').Readable;
}
var EntryStream = /*#__PURE__*/ function(ReadableBase) {
    "use strict";
    _inherits(EntryStream, ReadableBase);
    function EntryStream() {
        _class_call_check(this, EntryStream);
        var _this;
        _this = _call_super(this, EntryStream, arguments), _this._ended = false;
        return _this;
    }
    var _proto = EntryStream.prototype;
    /**
   * Signal end of stream by pushing null
   */ _proto.end = function end() {
        if (this._ended) return;
        this._ended = true;
        this.push(null);
    };
    /**
   * Required by Readable - called when consumer wants data.
   * Data is pushed externally via push(), nothing to do here.
   */ _proto._read = function _read(_size) {
    // Data is pushed externally, nothing to do here
    };
    _create_class(EntryStream, [
        {
            key: "ended",
            get: /**
   * Check if stream has ended
   */ function get() {
                return this._ended;
            }
        }
    ]);
    return EntryStream;
}(ReadableBase);
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
