// extract-base-iterator/dist/cjs/shared/index.js
/**
 * Shared utilities for iterator libraries
 *
 * These utilities are designed to be used by:
 * - zip-iterator
 * - 7z-iterator
 * - tar-iterator
 * - Any other archive iterator library
 *
 * All utilities support Node.js 0.8+
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BufferList () {
        return _BufferListts.default;
    },
    get EntryStream () {
        return _EntryStreamts.default;
    },
    get Lock () {
        return _Lockts.default;
    },
    get MAX_SAFE_BUFFER_LENGTH () {
        return _compatts.MAX_SAFE_BUFFER_LENGTH;
    },
    get PassThrough () {
        return _compatts.PassThrough;
    },
    get Readable () {
        return _compatts.Readable;
    },
    get Transform () {
        return _compatts.Transform;
    },
    get Writable () {
        return _compatts.Writable;
    },
    get allocBuffer () {
        return _compatts.allocBuffer;
    },
    get allocBufferUnsafe () {
        return _compatts.allocBufferUnsafe;
    },
    get bufferCompare () {
        return _compatts.bufferCompare;
    },
    get bufferConcat () {
        return _compatts.bufferConcat;
    },
    get bufferEquals () {
        return _compatts.bufferEquals;
    },
    get bufferFrom () {
        return _compatts.bufferFrom;
    },
    get bufferSliceCopy () {
        return _compatts.bufferSliceCopy;
    },
    get canAllocateBufferSize () {
        return _compatts.canAllocateBufferSize;
    },
    get crc32 () {
        return _crc32ts.crc32;
    },
    get crc32Region () {
        return _crc32ts.crc32Region;
    },
    get createInflateRawStream () {
        return _compatts.createInflateRawStream;
    },
    get inflateRaw () {
        return _compatts.inflateRaw;
    },
    get isNaN () {
        return _compatts.isNaN;
    },
    get normalizePath () {
        return _normalizePathts.default;
    },
    get objectAssign () {
        return _compatts.objectAssign;
    },
    get readUInt64LE () {
        return _compatts.readUInt64LE;
    },
    get streamToString () {
        return _streamToStringts.default;
    },
    get stripPath () {
        return _stripPathts.default;
    },
    get verifyCrc32 () {
        return _crc32ts.verifyCrc32;
    },
    get verifyCrc32Region () {
        return _crc32ts.verifyCrc32Region;
    },
    get writeUInt64LE () {
        return _compatts.writeUInt64LE;
    }
});
var _BufferListts = /*#__PURE__*/ _interop_require_default(require("./BufferList.js"));
var _compatts = require("./compat.js");
var _crc32ts = require("./crc32.js");
var _EntryStreamts = /*#__PURE__*/ _interop_require_default(require("./EntryStream.js"));
var _Lockts = /*#__PURE__*/ _interop_require_default(require("./Lock.js"));
var _normalizePathts = /*#__PURE__*/ _interop_require_default(require("./normalizePath.js"));
var _streamToStringts = /*#__PURE__*/ _interop_require_default(require("./streamToString.js"));
var _stripPathts = /*#__PURE__*/ _interop_require_default(require("./stripPath.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
