// xz-compat/dist/cjs/native.js
/**
 * Native Acceleration Module
 *
 * Provides optional native acceleration via lzma-native on Node.js 10+.
 * Falls back gracefully to pure JS implementation on older Node versions
 * or when the native module is not available.
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get isNativeAvailable () {
        return isNativeAvailable;
    },
    get tryLoadNative () {
        return tryLoadNative;
    }
});
var _module = /*#__PURE__*/ _interop_require_default(require("module"));
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _url = /*#__PURE__*/ _interop_require_default(require("url"));
var _typests = require("./lzma/types.js");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
// Get __dirname for ES modules
var _require = typeof require === 'undefined' ? _module.default.createRequire(require("url").pathToFileURL(__filename).toString()) : require;
var __dirname = _path.default.dirname(typeof __filename !== 'undefined' ? __filename : _url.default.fileURLToPath(require("url").pathToFileURL(__filename).toString()));
// Get node_modules path (go up from dist/cjs to package root, then to node_modules)
var nodeModulesPath = _path.default.join(__dirname, '..', '..', 'node_modules');
var major = +process.versions.node.split('.')[0];
var nativeDisabled = process.env.LZMA_NATIVE_DISABLE === '1';
var NATIVE_PREBUILDS = [
    'darwin-arm64',
    'darwin-x64',
    'linux-arm64',
    'linux-x64',
    'win32-ia32',
    'win32-x64'
];
// Cache for native module loading result
var nativeModule = null;
var installationAttempted = false;
var sizeMismatchError = function(expected, actual) {
    return new Error("Native decode size mismatch (expected ".concat(expected, ", got ").concat(actual, ")"));
};
function collectStream(stream, input, expectedSize) {
    return new Promise(function(resolve, reject) {
        var chunks = [];
        stream.on('data', function(chunk) {
            chunks.push(chunk);
        });
        stream.once('error', reject);
        stream.once('end', function() {
            var output = chunks.length === 1 ? chunks[0] : Buffer.concat(chunks);
            if (typeof expectedSize === 'number' && expectedSize >= 0 && output.length !== expectedSize) {
                reject(sizeMismatchError(expectedSize, output.length));
                return;
            }
            resolve(output);
        });
        stream.end(input);
    });
}
function createRawDecoder(lzma, filterId, buildOptions) {
    return function(input, properties, unpackSize) {
        var filters = [
            {
                id: filterId,
                options: buildOptions(properties)
            }
        ];
        var stream = lzma.createStream('rawDecoder', {
            filters: filters
        });
        return collectStream(stream, input, unpackSize);
    };
}
function createNativeModule(bindings) {
    return {
        xz: {
            decompress: function(input) {
                return bindings.decompress(input);
            }
        },
        lzma: createRawDecoder(bindings, bindings.FILTER_LZMA1, function(properties) {
            if (!properties || properties.length < 5) throw new Error('LZMA requires 5-byte properties');
            var _parseProperties = (0, _typests.parseProperties)(properties), lc = _parseProperties.lc, lp = _parseProperties.lp, pb = _parseProperties.pb, dictionarySize = _parseProperties.dictionarySize;
            return {
                dictSize: dictionarySize,
                lc: lc,
                lp: lp,
                pb: pb
            };
        }),
        lzma2: createRawDecoder(bindings, bindings.FILTER_LZMA2, function(properties) {
            if (!properties || properties.length < 1) throw new Error('LZMA2 requires properties byte');
            return {
                dictSize: (0, _typests.parseLzma2DictionarySize)(properties[0])
            };
        })
    };
}
function tryLoadNative() {
    if (installationAttempted) return nativeModule;
    installationAttempted = true;
    if (nativeDisabled) return null;
    if (major < 14) return null;
    if (NATIVE_PREBUILDS.indexOf("".concat(process.platform, "-").concat(process.arch)) < 0) return null; // only supported prebuilds (or else tries to build from source on install)
    var load = function() {
        try {
            var bindings = _require('lzma-native');
            nativeModule = createNativeModule(bindings);
            return nativeModule;
        } catch (unused) {
            return null;
        }
    };
    var loaded = load();
    if (loaded) return loaded;
    try {
        console.log('Installing lzma-native for native acceleration...');
        var installModule = _require('install-module-linked').default;
        installModule.sync('lzma-native', nodeModulesPath, {});
        return load();
    } catch (unused) {
        return null;
    }
}
function isNativeAvailable() {
    return tryLoadNative() !== null;
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
