// @zen-browser/surfer/dist/commands/download/addon.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAddonsToMozBuild = exports.initializeAddon = exports.generateAddonMozBuild = exports.unpackAddon = exports.downloadAddon = exports.resolveAddonDownloadUrl = exports.getAddons = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = require("node:path");
const picomatch_1 = require("picomatch");
const __1 = require("../..");
const constants_1 = require("../../constants");
const log_1 = require("../../log");
const utils_1 = require("../../utils");
const download_1 = require("../../utils/download");
const store_1 = require("../../utils/store");
const discard_1 = require("../discard");
const axios_1 = __importDefault(require("axios"));
const getAddons = () => Object.keys(__1.config.addons).map((addon) => ({
    name: addon,
    ...__1.config.addons[addon],
}));
exports.getAddons = getAddons;
async function resolveAddonDownloadUrl(addon) {
    switch (addon.platform) {
        case 'url': {
            return addon.url;
        }
        case 'amo': {
            try {
                const mozillaData = await axios_1.default.get(`https://addons.mozilla.org/api/v4/addons/addon/${addon.amoId}/versions/`);
                return mozillaData.data.results[0].files[0].url;
            }
            catch (error) {
                log_1.log.warning('The following error occured whilst fetching amo addon metadata');
                log_1.log.error(error);
                return '';
            }
        }
        case 'github': {
            try {
                const githubData = await axios_1.default.get(`https://api.github.com/repos/${addon.repo}/releases/tags/${addon.version}`, {
                    headers: {
                        'User-Agent': 'surfer-build -> addon downloader',
                    },
                });
                const assets = githubData.data.assets;
                if (!assets)
                    throw `The GitHub releases api did not return any assets for ${addon.repo} -> ${addon.version}.`;
                const matchingFile = assets.find((asset) => (0, picomatch_1.isMatch)(asset.name, addon.fileGlob));
                const fileDownloadUrl = matchingFile?.browser_download_url;
                if (!matchingFile)
                    throw `The GitHub releases api did not provide any files that matched '${addon.fileGlob}'`;
                if (!fileDownloadUrl)
                    throw `The GitHub releases api did not provide a download url for '${matchingFile.name}'`;
                return (fileDownloadUrl ||
                    '{Release file does not include a file matching the glob}');
            }
            catch (error) {
                log_1.log.warning('The following error occurred whilst fetching github metadata');
                log_1.log.error(error);
                return '';
            }
        }
    }
}
exports.resolveAddonDownloadUrl = resolveAddonDownloadUrl;
async function downloadAddon(url, addon) {
    const temporaryFile = (0, node_path_1.join)(constants_1.MELON_TMP_DIR, addon.name + '.xpi');
    log_1.log.info(`Download addon from ${url}`);
    {
        const extensionCache = (0, store_1.readItem)(addon.name);
        if (extensionCache.isNone()) {
            // We haven't stored it in the cache, therefore we need to redownload
            // it
        }
        else {
            const cache = extensionCache.unwrap();
            if (cache.url == url && (0, node_fs_1.existsSync)(temporaryFile)) {
                log_1.log.info(`Using cached version of ${addon.name}`);
                return temporaryFile;
            }
        }
    }
    if ((0, node_fs_1.existsSync)(temporaryFile)) {
        (0, node_fs_1.unlinkSync)(temporaryFile);
    }
    await (0, download_1.downloadFileToLocation)(url, temporaryFile);
    // I do not know why, but this delay causes unzip to work reliably
    await (0, utils_1.delay)(200);
    return temporaryFile;
}
exports.downloadAddon = downloadAddon;
async function unpackAddon(path, addon) {
    const outPath = (0, node_path_1.join)(constants_1.ENGINE_DIR, 'browser', 'extensions', addon.name);
    if ((0, node_fs_1.existsSync)(outPath)) {
        log_1.log.info(`The extension ${addon.name} has already been unpacked... skipping`);
        return;
    }
    log_1.log.info(`Unpacking extension...`);
    // I do not know why, but this delay causes unzip to work reliably
    await (0, utils_1.delay)(200);
    (0, utils_1.ensureDirectory)(outPath);
    await (0, utils_1.configDispatch)('unzip', {
        args: [(0, utils_1.windowsPathToUnix)(path), '-d', (0, utils_1.windowsPathToUnix)(outPath)],
        killOnError: true,
        shell: 'unix',
    });
}
exports.unpackAddon = unpackAddon;
async function generateAddonMozBuild(addon) {
    const outPath = (0, node_path_1.join)(constants_1.ENGINE_DIR, 'browser', 'extensions', addon.name);
    log_1.log.info(`Generating addon mozbuild...`);
    const files = await (0, utils_1.walkDirectoryTree)(outPath);
    // Because the tree has the potential of being infinitely recursive, we
    // cannot possibly know the the type of the tree
    //
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    function runTree(tree, parent) {
        if (Array.isArray(tree)) {
            return tree
                .sort()
                .map((file) => `FINAL_TARGET_FILES.features["${addon.id}"]${parent} += ["${file
                .replace(outPath + '/', '')
                .replace(outPath, '')}"]`)
                .join('\n');
        }
        const current = tree['.']
            .sort()
            // Don't use windows path, which brick mozbuild
            .map((f) => (0, utils_1.windowsPathToUnix)(f))
            .map((f) => `FINAL_TARGET_FILES.features["${addon.id}"]${parent} += ["${f
            .replace(outPath + '/', '')
            .replace(outPath, '')}"]`)
            .join('\n');
        const children = Object.keys(tree)
            .filter((folder) => folder !== '.')
            .filter((folder) => typeof tree[folder] !== 'undefined')
            .map((folder) => runTree(tree[folder], `${parent}["${folder}"]`))
            .join('\n');
        return `${current}\n${children}`;
    }
    (0, node_fs_1.writeFileSync)((0, node_path_1.join)(outPath, 'moz.build'), `
DEFINES["MOZ_APP_VERSION"] = CONFIG["MOZ_APP_VERSION"]
DEFINES["MOZ_APP_MAXVERSION"] = CONFIG["MOZ_APP_MAXVERSION"]

${runTree(files, '')}`);
}
exports.generateAddonMozBuild = generateAddonMozBuild;
async function initializeAddon(addon) {
    const outPath = (0, node_path_1.join)(constants_1.ENGINE_DIR, 'browser', 'extensions', addon.name);
    log_1.log.info(`Initializing addon...`);
    await (0, utils_1.configDispatch)('git', {
        args: ['add', '-f', '.'],
        cwd: outPath,
    });
    await (0, utils_1.configDispatch)('git', {
        args: ['commit', '-m', addon.name],
        cwd: constants_1.ENGINE_DIR,
    });
}
exports.initializeAddon = initializeAddon;
async function addAddonsToMozBuild(addons) {
    log_1.log.info('Adding addons to mozbuild...');
    // Discard the file to make sure it has no changes
    await (0, discard_1.discard)('browser/extensions/moz.build');
    const path = (0, node_path_1.join)(constants_1.ENGINE_DIR, 'browser', 'extensions', 'moz.build');
    // Append all the files to the bottom
    (0, node_fs_1.writeFileSync)(path, `${(0, node_fs_1.readFileSync)(path).toString()}\nDIRS += [${addons
        .map((addon) => addon.name)
        .sort()
        .map((addon) => `"${addon}"`)
        .join(',')}]`);
}
exports.addAddonsToMozBuild = addAddonsToMozBuild;
