// @zen-browser/surfer/dist/commands/download/firefox.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadInternals = exports.setupFirefoxSource = exports.shouldSetupFirefoxSource = void 0;
const execa_1 = __importDefault(require("execa"));
const node_fs_1 = require("node:fs");
const promises_1 = require("node:fs/promises");
const node_path_1 = require("node:path");
const __1 = require("../..");
const constants_1 = require("../../constants");
const log_1 = require("../../log");
const command_exists_1 = require("../../utils/command-exists");
const download_1 = require("../../utils/download");
const fs_1 = require("../../utils/fs");
const init_1 = require("../init");
const __2 = require("../..");
const addon_1 = require("./addon");
const utils_1 = require("../../utils");
const fs_extra_1 = __importDefault(require("fs-extra"));
function shouldSetupFirefoxSource() {
    return !((0, node_fs_1.existsSync)(constants_1.ENGINE_DIR) &&
        (0, node_fs_1.existsSync)((0, node_path_1.resolve)(constants_1.ENGINE_DIR, 'toolkit', 'moz.build')));
}
exports.shouldSetupFirefoxSource = shouldSetupFirefoxSource;
async function setupFirefoxSource(version, candidateBuild, isCandidate = false) {
    const firefoxSourceTar = await downloadFirefoxSource(version, candidateBuild, isCandidate);
    await unpackFirefoxSource(firefoxSourceTar);
    if (!process.env.CI_SKIP_INIT) {
        log_1.log.info('Init firefox');
        await (0, init_1.init)(constants_1.ENGINE_DIR);
    }
}
exports.setupFirefoxSource = setupFirefoxSource;
async function unpackFirefoxSource(name) {
    log_1.log.info(`Unpacking Firefox...`);
    (0, fs_1.ensureDirectory)(constants_1.ENGINE_DIR);
    let tarExec = 'tar';
    // On MacOS, we need to use gnu tar, otherwise tar doesn't behave how we
    // would expect it to behave, so this section is responsible for handling
    // that
    //
    // If BSD tar adds --transform support in the future, we can use that
    // instead
    if (process.platform == 'darwin') {
        // GNU Tar doesn't come preinstalled on any MacOS machines, so we need to
        // check for it and ask for the user to install it if necessary
        if (!(0, command_exists_1.commandExistsSync)('gtar')) {
            throw new Error(`GNU Tar is required to extract Firefox's source on MacOS. Please install it using the command |brew install gnu-tar| or |sudo port install gnutar| and try again`);
        }
        tarExec = 'gtar';
    }
    log_1.log.info(`Unpacking ${(0, node_path_1.resolve)(constants_1.MELON_TMP_DIR, name)} to ${constants_1.ENGINE_DIR}`);
    if (process.platform === 'win32') {
        log_1.log.info('Unpacking Firefox source on Windows (7z)');
        await (0, execa_1.default)('7z', [
            'x',
            (0, node_path_1.resolve)(constants_1.MELON_TMP_DIR, name),
            '-o' + (0, node_path_1.resolve)(constants_1.MELON_TMP_DIR, name.replace('.tar.xz', '.tar')),
        ]);
        log_1.log.info('Unpacking Firefox source again without the .xz extension');
        await (0, execa_1.default)('7z', [
            'x',
            (0, node_path_1.resolve)(constants_1.MELON_TMP_DIR, name.replace('.tar.xz', '.tar')),
            '-o' + constants_1.MELON_TMP_DIR,
        ]);
        const archiveDir = (0, node_path_1.resolve)(constants_1.MELON_TMP_DIR, 'firefox-' + (0, utils_1.getFFVersionOrCandidate)());
        if ((0, node_fs_1.existsSync)(constants_1.ENGINE_DIR)) {
            // remove the existing engine directory
            fs_extra_1.default.removeSync(constants_1.ENGINE_DIR);
        }
        log_1.log.info('Moving Firefox source to engine directory');
        fs_extra_1.default.moveSync(archiveDir, constants_1.ENGINE_DIR);
        return;
    }
    await (0, execa_1.default)(tarExec, [
        '--strip-components=1',
        '-xf',
        (0, node_path_1.resolve)(constants_1.MELON_TMP_DIR, name),
        '-C',
        constants_1.ENGINE_DIR,
    ].filter(Boolean), {
        shell: constants_1.BASH_PATH,
    });
    log_1.log.info(`Unpacked Firefox source to ${constants_1.ENGINE_DIR}`);
}
async function downloadFirefoxSource(version, candidateBuild, isCandidate = false) {
    const filename = `firefox-${version}.source.tar.xz`;
    const getReleaseUri = (build) => {
        let base = `https://archive.mozilla.org/pub/firefox/releases/${version}/source/`;
        if (isCandidate) {
            console.log('Using candidate build');
            base = `https://archive.mozilla.org/pub/firefox/candidates/${version}-candidates/${build}/source/`;
        }
        return base + filename;
    };
    const fsParent = constants_1.MELON_TMP_DIR;
    const fsSaveLocation = (0, node_path_1.resolve)(fsParent, filename);
    log_1.log.info(`Locating Firefox release ${version}...`);
    await (0, fs_1.ensureDirectory)((0, node_path_1.dirname)(fsSaveLocation));
    if ((0, node_fs_1.existsSync)(fsSaveLocation)) {
        log_1.log.info('Using cached download');
        return filename;
    }
    // Do not re-download if there is already an existing workspace present
    if ((0, node_fs_1.existsSync)(constants_1.ENGINE_DIR))
        log_1.log.error(`Workspace already exists.\nRemove that workspace and run |${__1.bin_name} download ${version}| again.`);
    log_1.log.info(`Downloading Firefox release ${version}...`);
    // Try to download the second build first, as it is more likely to be the
    // correct build
    const url = getReleaseUri(`build${candidateBuild}`);
    await (0, download_1.downloadFileToLocation)(url, (0, node_path_1.resolve)(constants_1.MELON_TMP_DIR, filename));
    return filename;
}
async function downloadInternals({ version, force, isCandidate = (0, utils_1.shouldUseCandidate)(), }) {
    // Provide a legible error if there is no version specified
    if (!version) {
        log_1.log.error('You have not specified a version of firefox in your config file. This is required to build a firefox fork.');
        process.exit(1);
    }
    let candidateBuild = 1;
    if (isCandidate) {
        version = __2.config.version.candidate;
        candidateBuild = __2.config.version.candidateBuild;
    }
    if (force && (0, node_fs_1.existsSync)(constants_1.ENGINE_DIR)) {
        log_1.log.info('Removing existing workspace');
        (0, node_fs_1.rmSync)(constants_1.ENGINE_DIR, { recursive: true });
    }
    // If the engine directory is empty, we should delete it.
    const engineIsEmpty = (0, node_fs_1.existsSync)(constants_1.ENGINE_DIR) &&
        (await (0, promises_1.readdir)(constants_1.ENGINE_DIR).then((files) => files.length === 0));
    if (engineIsEmpty) {
        log_1.log.info("'engine/' is empty, it...");
        (0, node_fs_1.rmSync)(constants_1.ENGINE_DIR, { recursive: true });
    }
    if (!(0, node_fs_1.existsSync)(constants_1.ENGINE_DIR)) {
        await setupFirefoxSource(version, candidateBuild, isCandidate);
    }
    for (const addon of (0, addon_1.getAddons)()) {
        const downloadUrl = await (0, addon_1.resolveAddonDownloadUrl)(addon);
        const downloadedXPI = await (0, addon_1.downloadAddon)(downloadUrl, addon);
        await (0, addon_1.unpackAddon)(downloadedXPI, addon);
        await (0, addon_1.generateAddonMozBuild)(addon);
        await (0, addon_1.initializeAddon)(addon);
    }
    await (0, addon_1.addAddonsToMozBuild)((0, addon_1.getAddons)());
    if (!isCandidate) {
        __2.config.version.version = version;
    }
    else {
        __2.config.version.candidate = version;
    }
    (0, node_fs_1.writeFileSync)(utils_1.configPath, JSON.stringify(__2.config, undefined, 2));
}
exports.downloadInternals = downloadInternals;
