// 7z-iterator/dist/cjs/lib/streamToSource.js
// Stream to source conversion: writes stream to temp file for random access
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, /**
 * Convert a stream to a FileSource by writing to temp file
 *
 * 7z format requires random access for header parsing, so temp file is necessary for streams.
 * Writes directly to temp file for predictable O(1) memory usage during stream consumption.
 */ "default", {
    enumerable: true,
    get: function() {
        return streamToSource;
    }
});
var _calloncefn = /*#__PURE__*/ _interop_require_default(require("call-once-fn"));
var _extractbaseiterator = require("extract-base-iterator");
var _gracefulfs = /*#__PURE__*/ _interop_require_default(require("graceful-fs"));
var _mkdirpclassic = /*#__PURE__*/ _interop_require_default(require("mkdirp-classic"));
var _onone = /*#__PURE__*/ _interop_require_default(require("on-one"));
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _SevenZipParserts = require("../sevenz/SevenZipParser.js");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function streamToSource(stream, options, callback) {
    var tempPath = options.tempPath;
    var end = (0, _calloncefn.default)(callback);
    _mkdirpclassic.default.sync(_path.default.dirname(tempPath));
    var writeStream = _gracefulfs.default.createWriteStream(tempPath);
    function onData(chunk) {
        var buf = typeof chunk === 'string' ? (0, _extractbaseiterator.bufferFrom)(chunk) : chunk;
        writeStream.write(buf);
    }
    function onEnd() {
        writeStream.end(function() {
            _gracefulfs.default.open(tempPath, 'r', function(err, fd) {
                if (err) return end(err);
                _gracefulfs.default.stat(tempPath, function(statErr, stats) {
                    if (statErr) {
                        _gracefulfs.default.closeSync(fd);
                        return end(statErr);
                    }
                    end(null, {
                        source: new _SevenZipParserts.FileSource(fd, stats.size),
                        fd: fd,
                        tempPath: tempPath
                    });
                });
            });
        });
    }
    function onError(err) {
        writeStream.end();
        end(err);
    }
    stream.on('data', onData);
    (0, _onone.default)(stream, [
        'error'
    ], onError);
    (0, _onone.default)(stream, [
        'end',
        'close',
        'finish'
    ], onEnd);
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
