// @zen-browser/surfer/dist/commands/patches/branding-patch.js
'use strict'
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.apply = exports.get = exports.BRANDING_DIR = void 0
const node_fs_1 = require('node:fs')
const promises_1 = require('node:fs/promises')
const modern_async_1 = require('modern-async')
const node_path_1 = require('node:path')
const sharp_1 = __importDefault(require('sharp'))
const async_icns_1 = __importDefault(require('async-icns'))
const __1 = require('../..')
const constants_1 = require('../../constants')
const log_1 = require('../../log')
const utils_1 = require('../../utils')
const setup_project_1 = require('../setup-project')
// =============================================================================
// Pure constants
exports.BRANDING_DIR = (0, node_path_1.join)(
  constants_1.CONFIGS_DIR,
  'branding'
)
const BRANDING_STORE = (0, node_path_1.join)(
  constants_1.ENGINE_DIR,
  'browser',
  'branding'
)
const BRANDING_FF = (0, node_path_1.join)(BRANDING_STORE, 'unofficial')
const SHARED_NSH = (0, node_path_1.join)(
  constants_1.ENGINE_DIR,
  'browser',
  'installer',
  'windows',
  'nsis',
  'shared.nsh'
)
const REQUIRED_FILES = [
  'logo.png',
  'logo-mac.png',
  'firefox.ico',
  'firefox64.ico',
]
const BRANDING_NSIS = 'branding.nsi'
const CSS_REPLACE_REGEX = new RegExp(
  '#130829|hsla\\(235, 43%, 10%, .5\\)',
  'gm'
)
// =============================================================================
// Utility Functions
function checkForFaults(name, configPath) {
  if (!(0, node_fs_1.existsSync)(configPath)) {
    throw new Error(`Branding ${name} does not exist`)
  }
  const requiredFiles = REQUIRED_FILES.map((file) =>
    (0, node_path_1.join)(configPath, file)
  )
  const requiredFilesExist = (0, utils_1.filesExist)(requiredFiles)
  if (!requiredFilesExist) {
    throw new Error(
      `Missing some of the required files: ${requiredFiles
        .filter((file) => !(0, node_fs_1.existsSync)(file))
        .join(', ')}`
    )
  }
}
function constructConfig(name) {
  return {
    brandingGenericName: __1.config.name,
    brandingVendor: __1.config.vendor,
    ...utils_1.defaultBrandsConfig,
    ...__1.config.brands[name],
  }
}
// =============================================================================
// Main code
async function setupImages(configPath, outputPath) {
  log_1.log.debug('Generating icons')
  // Firefox doesn't use 512 by 512, but we need it to generate ico files later
  await (0, modern_async_1.every)(
    [16, 22, 24, 32, 48, 64, 128, 256, 512],
    async (size) => {
      const logoPath = (0, node_path_1.join)(configPath, `logo${size}.png`)
      if (!(0, utils_1.filesExist)([logoPath]))
        throw new Error(`Missing logo${size}.png`)
      const outputPathLogo = (0, node_path_1.join)(
        outputPath,
        `default${size}.png`
      )
      await (0, promises_1.copyFile)(logoPath, outputPathLogo)
      return true
    }
  )
  // TODO: Custom MacOS icon support
  if (process.surferPlatform == 'darwin') {
    log_1.log.debug('Generating Mac Icons')
    log_1.log.debug(
      `Using MacOS icon: ${(0, node_path_1.join)(configPath, 'logo-mac.png')}`
    )
    log_1.log.debug(`Output path: ${outputPath}`)
    const temporary = (0, node_path_1.join)(
      constants_1.MELON_TMP_DIR,
      'macos_icon_info.iconset'
    )
    if ((0, node_fs_1.existsSync)(temporary))
      await (0, promises_1.rm)(temporary, { recursive: true })
    await async_icns_1.default.convert({
      input: (0, node_path_1.join)(configPath, 'logo-mac.png'),
      output: (0, node_path_1.join)(outputPath, 'firefox.icns'),
      sizes: [16, 32, 64, 128, 256, 512],
      tmpDirectory: temporary,
    })
  }
  ;(0, node_fs_1.mkdirSync)((0, node_path_1.join)(outputPath, 'content'), {
    recursive: true,
  })
  await (0, sharp_1.default)((0, node_path_1.join)(configPath, 'logo.png'))
    .resize(512, 512)
    .toFile((0, node_path_1.join)(outputPath, 'content', 'about-logo.png'))
  await (0, sharp_1.default)((0, node_path_1.join)(configPath, 'logo.png'))
    .resize(1024, 1024)
    .toFile((0, node_path_1.join)(outputPath, 'content', 'about-logo@2x.png'))
  // Register logo in cache
  await (0, utils_1.addHash)((0, node_path_1.join)(configPath, 'logo.png'))
}
async function setupLocale(outputPath, brandingConfig) {
  for (const file of await (0, utils_1.walkDirectory)(
    (0, node_path_1.join)(
      setup_project_1.templateDirectory,
      'branding.optional'
    )
  )) {
    const fileContents = await (0, promises_1.readFile)(
      (0, utils_1.windowsPathToUnix)(file),
      {
        encoding: 'utf8',
      }
    )
    const universalPath =
      // We want to avoid the pain that windows is going to throw at us with its
      // weird paths
      (0, utils_1.windowsPathToUnix)(file)
        // We want to remove all of the extra folders that surround this from the
        // template folder
        .replace(
          (0, utils_1.windowsPathToUnix)(
            (0, node_path_1.join)(
              setup_project_1.templateDirectory,
              'branding.optional'
            ) + '/'
          ),
          ''
        )
    const sourceFolderPath = (0, node_path_1.join)(outputPath, universalPath)
    await (0, promises_1.mkdir)((0, node_path_1.dirname)(sourceFolderPath), {
      recursive: true,
    })
    await (0, promises_1.writeFile)(
      sourceFolderPath,
      (0, utils_1.stringTemplate)(fileContents, brandingConfig)
    )
  }
}
async function copyMozFiles(outputPath, brandingConfig) {
  const firefoxBrandingDirectoryContents = await (0, utils_1.walkDirectory)(
    BRANDING_FF
  )
  const files = firefoxBrandingDirectoryContents.filter(
    (file) =>
      !(0, node_fs_1.existsSync)(
        (0, node_path_1.join)(outputPath, file.replace(BRANDING_FF, ''))
      )
  )
  const css = files.filter((file) =>
    (0, node_path_1.extname)(file).includes('css')
  )
  const everythingElse = files.filter(
    (file) => !css.includes(file) && !file.includes(BRANDING_NSIS)
  )
  for (const [contents, path] of css
    .map((filePath) => [
      (0, node_fs_1.readFileSync)(filePath).toString(),
      (0, node_path_1.join)(outputPath, filePath.replace(BRANDING_FF, '')),
    ])
    .map(([contents, path]) => [
      contents.replace(CSS_REPLACE_REGEX, 'var(--theme-bg)') +
        `:root { --theme-bg: ${brandingConfig.backgroundColor} }`,
      path,
    ])) {
    ;(0, node_fs_1.mkdirSync)((0, node_path_1.dirname)(path), {
      recursive: true,
    })
    ;(0, node_fs_1.writeFileSync)(path, contents)
  }
  const brandingNsis = files.filter((file) => file.includes(BRANDING_NSIS))
  console.assert(
    brandingNsis.length == 1,
    'There should only be one branding.nsi file'
  )
  const outputBrandingNsis = (0, node_path_1.join)(
    outputPath,
    brandingNsis[0].replace(BRANDING_FF, '')
  )
  const configureProfileBrandingPath = (0, node_path_1.join)(
    outputPath,
    'pref',
    'firefox-branding.js'
  )
  log_1.log.debug('Configuring branding.nsi into ' + outputBrandingNsis)
  configureBrandingNsis(outputBrandingNsis, brandingConfig)
  // Copy everything else from the default firefox branding directory
  for (const file of everythingElse) {
    ;(0, utils_1.mkdirpSync)(
      (0, node_path_1.dirname)(
        (0, node_path_1.join)(outputPath, file.replace(BRANDING_FF, ''))
      )
    )
    ;(0, node_fs_1.copyFileSync)(
      file,
      (0, node_path_1.join)(outputPath, file.replace(BRANDING_FF, ''))
    )
  }
  configureProfileBranding(configureProfileBrandingPath, brandingConfig)
}
function get() {
  if (!(0, node_fs_1.existsSync)(exports.BRANDING_DIR)) return []
  return (0, node_fs_1.readdirSync)(exports.BRANDING_DIR).filter((file) =>
    (0, node_fs_1.lstatSync)(
      (0, node_path_1.join)(exports.BRANDING_DIR, file)
    ).isDirectory()
  )
}
exports.get = get
async function apply(name) {
  const configPath = (0, node_path_1.join)(exports.BRANDING_DIR, name)
  const outputPath = (0, node_path_1.join)(BRANDING_STORE, name)
  checkForFaults(name, configPath)
  const brandingConfig = constructConfig(name)
  // Remove the output path if it exists and recreate it
  ;(0, utils_1.ensureEmpty)(outputPath)
  await setupImages(configPath, outputPath)
  await setupLocale(outputPath, { ...brandingConfig, appId: __1.config.appId })
  await copyMozFiles(outputPath, brandingConfig)
  await addOptionalIcons(configPath, outputPath)
  setUpdateURLs()
}
exports.apply = apply
function configureBrandingNsis(brandingNsis, brandingConfig) {
  ;(0, node_fs_1.writeFileSync)(
    brandingNsis,
    `
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

# NSIS branding defines for official release builds.
# The nightly build branding.nsi is located in browser/installer/windows/nsis/
# The unofficial build branding.nsi is located in browser/branding/unofficial/

# BrandFullNameInternal is used for some registry and file system values
# instead of BrandFullName and typically should not be modified.
!define BrandFullNameInternal "${brandingConfig.brandFullName}"
!define BrandFullName         "${brandingConfig.brandFullName}"
!define CompanyName           "${brandingConfig.brandingVendor}"
!define URLInfoAbout          "https://zen-browser.app"
!define URLUpdateInfo         "https://zen-browser.app/release-notes/\${AppVersion}"
!define HelpLink              "https://github.com/zen-browser/desktop/issues"

; The OFFICIAL define is a workaround to support different urls for Release and
; Stable since they share the same branding when building with other branches that
; set the update channel to stable.
!define OFFICIAL
!define URLStubDownloadX86 "https://download.mozilla.org/?os=win&lang=\${AB_CD}&product=firefox-latest"
!define URLStubDownloadAMD64 "https://download.mozilla.org/?os=win64&lang=\${AB_CD}&product=firefox-latest"
!define URLStubDownloadAArch64 "https://download.mozilla.org/?os=win64-aarch64&lang=\${AB_CD}&product=firefox-latest"
!define URLManualDownload "https://zen-browser.app/download"
!define URLSystemRequirements "https://www.mozilla.org/firefox/system-requirements/"
!define Channel "stable"

# The installer's certificate name and issuer expected by the stub installer
!define CertNameDownload   "${brandingConfig.brandFullName}"
!define CertIssuerDownload "DigiCert SHA2 Assured ID Code Signing CA"

# Dialog units are used so the UI displays correctly with the system's DPI
# settings. These are tweaked to look good with the en-US strings; ideally
# we would customize them for each locale but we don't really have a way to
# implement that and it would be a ton of work for the localizers.
!define PROFILE_CLEANUP_LABEL_TOP "50u"
!define PROFILE_CLEANUP_LABEL_LEFT "22u"
!define PROFILE_CLEANUP_LABEL_WIDTH "175u"
!define PROFILE_CLEANUP_LABEL_HEIGHT "100u"
!define PROFILE_CLEANUP_LABEL_ALIGN "left"
!define PROFILE_CLEANUP_CHECKBOX_LEFT "22u"
!define PROFILE_CLEANUP_CHECKBOX_WIDTH "175u"
!define PROFILE_CLEANUP_BUTTON_LEFT "22u"
!define INSTALL_HEADER_TOP "70u"
!define INSTALL_HEADER_LEFT "22u"
!define INSTALL_HEADER_WIDTH "180u"
!define INSTALL_HEADER_HEIGHT "100u"
!define INSTALL_BODY_LEFT "22u"
!define INSTALL_BODY_WIDTH "180u"
!define INSTALL_INSTALLING_TOP "115u"
!define INSTALL_INSTALLING_LEFT "270u"
!define INSTALL_INSTALLING_WIDTH "150u"
!define INSTALL_PROGRESS_BAR_TOP "100u"
!define INSTALL_PROGRESS_BAR_LEFT "270u"
!define INSTALL_PROGRESS_BAR_WIDTH "150u"
!define INSTALL_PROGRESS_BAR_HEIGHT "12u"

!define PROFILE_CLEANUP_CHECKBOX_TOP_MARGIN "12u"
!define PROFILE_CLEANUP_BUTTON_TOP_MARGIN "12u"
!define PROFILE_CLEANUP_BUTTON_X_PADDING "80u"
!define PROFILE_CLEANUP_BUTTON_Y_PADDING "8u"
!define INSTALL_BODY_TOP_MARGIN "20u"

# Font settings that can be customized for each channel
!define INSTALL_HEADER_FONT_SIZE 20
!define INSTALL_HEADER_FONT_WEIGHT 600
!define INSTALL_INSTALLING_FONT_SIZE 15
!define INSTALL_INSTALLING_FONT_WEIGHT 600

# UI Colors that can be customized for each channel
!define COMMON_TEXT_COLOR 0x000000
!define COMMON_BACKGROUND_COLOR 0xFFFFFF
!define INSTALL_INSTALLING_TEXT_COLOR 0xFFFFFF
# This color is written as 0x00BBGGRR because it's actually a COLORREF value.
!define PROGRESS_BAR_BACKGROUND_COLOR 0xFFAA00
`
  )
  ;(0, node_fs_1.writeFileSync)(
    SHARED_NSH,
    (0, node_fs_1.readFileSync)(SHARED_NSH)
      .toString()
      .replace(
        '"Publisher" "Mozilla"',
        `"Publisher" "${brandingConfig.brandingVendor}"`
      )
  )
}
function addOptionalIcons(brandingPath, outputPath) {
  // move all icons in the top directory and inside "content/" into the branding directory
  const icons = (0, node_fs_1.readdirSync)(brandingPath)
  const iconsContent = (0, node_fs_1.readdirSync)(
    (0, node_path_1.join)(brandingPath, 'content')
  )
  for (const icon of icons) {
    if (icon.includes('content')) continue
    log_1.log.info(`Copying ${icon} to ${outputPath}`)
    ;(0, node_fs_1.copyFileSync)(
      (0, node_path_1.join)(brandingPath, icon),
      (0, node_path_1.join)(outputPath, icon)
    )
  }
  for (const icon of iconsContent) {
    log_1.log.info(`Copying ${icon} to ${outputPath}`)
    ;(0, node_fs_1.copyFileSync)(
      (0, node_path_1.join)(brandingPath, 'content', icon),
      (0, node_path_1.join)(outputPath, 'content', icon)
    )
  }
}
function configureProfileBranding(brandingPath, brandingConfig) {
  ;(0, node_fs_1.writeFileSync)(
    brandingPath,
    `
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

pref("startup.homepage_override_url", "https://zen-browser.app/whatsnew?v=%VERSION%");
pref("startup.homepage_welcome_url", "https://zen-browser.app/welcome/");
pref("startup.homepage_welcome_url.additional", "https://zen-browser.app/privacy-policy/");

// Give the user x seconds to react before showing the big UI. default=192 hours
pref("app.update.promptWaitTime", 691200);
// app.update.url.manual: URL user can browse to manually if for some reason
// all update installation attempts fail.
// app.update.url.details: a default value for the "More information about this
// update" link supplied in the "An update is available" page of the update
// wizard.
pref("app.update.url.manual", "https://zen-browser.app/download/");
pref("app.update.url.details", "https://zen-browser.app/release-notes/latest/");
pref("app.releaseNotesURL", "https://zen-browser.app/whatsnew/");
pref("app.releaseNotesURL.aboutDialog", "https://www.zen-browser.app/release-notes/%VERSION%/");
pref("app.releaseNotesURL.prompt", "https://zen-browser.app/release-notes/%VERSION%/");

// Number of usages of the web console.
// If this is less than 5, then pasting code into the web console is disabled
pref("devtools.selfxss.count", 5);
`
  )
}
function setUpdateURLs() {
  const baseURL = `URL=https://@MOZ_APPUPDATE_HOST@/updates/browser/%BUILD_TARGET%/%CHANNEL%/update.xml`
  const appIni = (0, node_path_1.join)(
    constants_1.ENGINE_DIR,
    'build',
    'application.ini.in'
  )
  const appIniContents = (0, node_fs_1.readFileSync)(appIni).toString()
  const updatedAppIni = appIniContents.replace(/URL=.*update.xml/g, baseURL)
  ;(0, node_fs_1.writeFileSync)(appIni, updatedAppIni)
}
