// @zen-browser/surfer/dist/commands/patches/command.js
'use strict'
var __createBinding =
  (this && this.__createBinding) ||
  (Object.create
    ? function (o, m, k, k2) {
        if (k2 === undefined) k2 = k
        var desc = Object.getOwnPropertyDescriptor(m, k)
        if (
          !desc ||
          ('get' in desc ? !m.__esModule : desc.writable || desc.configurable)
        ) {
          desc = {
            enumerable: true,
            get: function () {
              return m[k]
            },
          }
        }
        Object.defineProperty(o, k2, desc)
      }
    : function (o, m, k, k2) {
        if (k2 === undefined) k2 = k
        o[k2] = m[k]
      })
var __setModuleDefault =
  (this && this.__setModuleDefault) ||
  (Object.create
    ? function (o, v) {
        Object.defineProperty(o, 'default', { enumerable: true, value: v })
      }
    : function (o, v) {
        o['default'] = v
      })
var __importStar =
  (this && this.__importStar) ||
  function (mod) {
    if (mod && mod.__esModule) return mod
    var result = {}
    if (mod != null)
      for (var k in mod)
        if (k !== 'default' && Object.prototype.hasOwnProperty.call(mod, k))
          __createBinding(result, mod, k)
    __setModuleDefault(result, mod)
    return result
  }
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.applyPatches = void 0
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const node_path_1 = require('node:path')
const node_fs_1 = require('node:fs')
const tiny_glob_1 = __importDefault(require('tiny-glob'))
const constants_1 = require('../../constants')
const gitPatch = __importStar(require('./git-patch'))
const copyPatch = __importStar(require('./copy-patches'))
const brandingPatch = __importStar(require('./branding-patch'))
const patch_check_1 = require('../../middleware/patch-check')
const utils_1 = require('../../utils')
const task_list_1 = require('../../utils/task-list')
const promises_1 = require('node:fs/promises')
function patchMethod(name, patches, patchFunction) {
  return {
    name: `Apply ${patches.length} ${name} patches`,
    long: true,
    task: () =>
      new task_list_1.TaskList(
        patches.map((patch, index) => ({
          name: `Apply ${patch.name}`,
          task: () => patchFunction(patch, index),
          skip: patch.skip,
        }))
      ),
  }
}
function importMelonPatches() {
  return patchMethod(
    'branding',
    [
      ...brandingPatch.get().map((name) => ({
        type: 'branding',
        name,
        value: name,
        skip: async () => {
          const logoCheck = (0, utils_1.checkHash)(
            (0, node_path_1.join)(brandingPatch.BRANDING_DIR, name, 'logo.png')
          )
          const macosInstallerCheck = (0, utils_1.checkHash)(
            (0, node_path_1.join)(
              brandingPatch.BRANDING_DIR,
              name,
              'MacOSInstaller.svg'
            )
          )
          if (
            (await logoCheck) &&
            (await macosInstallerCheck) &&
            (0, node_fs_1.existsSync)(
              (0, node_path_1.join)(
                constants_1.ENGINE_DIR,
                'browser/branding',
                name
              )
            )
          ) {
            return true
          }
          return false
        },
      })),
    ],
    async (patch) => await brandingPatch.apply(patch.value)
  )
}
async function importFolders() {
  return patchMethod(
    'folder',
    await copyPatch.get(),
    async (patch) => await copyPatch.apply(patch)
  )
}
async function importGitPatch() {
  let patches = await (0, tiny_glob_1.default)('**/*.patch', {
    filesOnly: true,
    cwd: constants_1.SRC_DIR,
  })
  patches = patches.map((path) =>
    (0, node_path_1.join)(constants_1.SRC_DIR, path)
  )
  await (0, promises_1.writeFile)(
    patch_check_1.patchCountFile,
    patches.length.toString()
  )
  return patchMethod(
    'git',
    patches.map((path) => ({ name: path, path })),
    async (patch) => await gitPatch.apply(patch.path)
  )
}
async function importInternalPatch() {
  const patches = await (0, tiny_glob_1.default)('*.patch', {
    filesOnly: true,
    cwd: constants_1.PATCHES_DIR,
  })
  const structuredPatches = patches.map((path) => ({
    name: path,
    path: (0, node_path_1.join)(constants_1.PATCHES_DIR, path),
  }))
  return patchMethod(
    'surfer',
    structuredPatches,
    async (patch) => await gitPatch.apply(patch.path)
  )
}
async function applyPatches() {
  const canDoBrandingPatch = process.env.SURFER_NO_BRANDING_PATCH !== 'true'
  let tasks = [
    await importInternalPatch(),
    canDoBrandingPatch ? importMelonPatches() : undefined,
    await importFolders(),
    await importGitPatch(),
  ].filter((task) => task !== undefined)
  await new task_list_1.TaskList(tasks).run()
}
exports.applyPatches = applyPatches
