// @zen-browser/surfer/dist/commands/patches/copy-patches.js
'use strict'
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.apply = exports.get = exports.copyManual = void 0
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const node_fs_1 = require('node:fs')
const node_fs_2 = require('node:fs')
const fs_extra_1 = require('fs-extra')
const promises_1 = require('node:fs/promises')
const node_path_1 = require('node:path')
const tiny_glob_1 = __importDefault(require('tiny-glob'))
const utils_1 = require('../../utils')
const __1 = require('../..')
const constants_1 = require('../../constants')
// =============================================================================
// Utilities
const getChunked = (location) => location.replace(/\\/g, '/').split('/')
const copyManual = async (name, patchName) => {
  let dest = constants_1.ENGINE_DIR
  if (patchName === 'tests') {
    dest = (0, node_path_1.resolve)(dest, 'browser', 'base', 'zen-components')
  }
  const placeToCheck =
    patchName === 'tests' ? constants_1.CURRENT_DIR : constants_1.SRC_DIR
  // If the file exists and is not a symlink, we want to replace it with a
  // symlink to our file, so remove it
  if (
    (0, node_fs_1.existsSync)(
      (0, node_path_1.resolve)(dest, ...getChunked(name))
    ) &&
    !(0, node_fs_2.lstatSync)(
      (0, node_path_1.resolve)(dest, ...getChunked(name))
    ).isSymbolicLink()
  ) {
    await (0, fs_extra_1.remove)(
      (0, node_path_1.resolve)(dest, ...getChunked(name))
    )
  }
  try {
    if (
      process.platform == 'win32' &&
      !__1.config.buildOptions.windowsUseSymbolicLinks
    ) {
      // Make the directory if it doesn't already exist.
      await (0, utils_1.mkdirp)(
        (0, node_path_1.dirname)(
          (0, node_path_1.resolve)(dest, ...getChunked(name))
        )
      )
      // By default, windows users do not have access to the permissions to create
      // symbolic links. As a work around, we will just copy the files instead
      await (0, promises_1.copyFile)(
        (0, node_path_1.resolve)(placeToCheck, ...getChunked(name)),
        (0, node_path_1.resolve)(dest, ...getChunked(name))
      )
    } else {
      // Create the symlink
      await (0, fs_extra_1.ensureSymlink)(
        (0, node_path_1.resolve)(placeToCheck, ...getChunked(name)),
        (0, node_path_1.resolve)(dest, ...getChunked(name))
      )
    }
  } catch (e) {
    console.info('name: ', name)
    console.info('patchName: ', patchName)
    console.error(e) // Just in case we have an error
  }
  const gitignore = (0, node_fs_2.readFileSync)(
    (0, node_path_1.resolve)(constants_1.ENGINE_DIR, '.gitignore')
  ).toString()
  if (!gitignore.includes(getChunked(name).join('/')))
    (0, utils_1.appendToFileSync)(
      (0, node_path_1.resolve)(constants_1.ENGINE_DIR, '.gitignore'),
      `\n${getChunked(name).join('/')}`
    )
}
exports.copyManual = copyManual
// =============================================================================
// Exports
async function get() {
  const allFilesInSource = await (0, tiny_glob_1.default)('**/*', {
    filesOnly: true,
    cwd: constants_1.SRC_DIR,
  })
  const files = allFilesInSource.filter(
    (f) => !(f.endsWith('.patch') || f.split('/').includes('node_modules'))
  )
  const manualPatches = []
  files.map((index) => {
    const group = index.split('/')[0]
    if (!manualPatches.some((m) => m.name == group)) {
      manualPatches.push({
        name: group,
        src: files.filter((f) => f.split('/')[0] == group),
      })
    }
  })
  await (0, utils_1.ensureDirectory)('./tests')
  const testFiles = await (0, tiny_glob_1.default)('./tests/**/*', {
    filesOnly: true,
    cwd: '.',
  })
  const testFilesGrouped = testFiles.filter(
    (f) => !(f.endsWith('.patch') || f.split('/').includes('node_modules'))
  )
  testFilesGrouped.map((index) => {
    const group = index.split('/')[0]
    if (!manualPatches.some((m) => m.name == group)) {
      manualPatches.push({
        name: group,
        src: testFilesGrouped.filter((f) => f.split('/')[0] == group),
      })
    }
  })
  return manualPatches
}
exports.get = get
async function apply({ src, name }) {
  for (const item of src) {
    await (0, exports.copyManual)(item, name)
  }
}
exports.apply = apply
