// fast-extract/dist/cjs/streams/transforms/IteratorTransform.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return createIteratorTransform;
    }
});
var _extractbaseiterator = require("extract-base-iterator");
function _assert_this_initialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _call_super(_this, derived, args) {
    derived = _get_prototype_of(derived);
    return _possible_constructor_return(_this, _is_native_reflect_construct() ? Reflect.construct(derived, args || [], _get_prototype_of(_this).constructor) : derived.apply(_this, args));
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _get_prototype_of(o) {
    _get_prototype_of = Object.setPrototypeOf ? Object.getPrototypeOf : function getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _get_prototype_of(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _set_prototype_of(subClass, superClass);
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
function _possible_constructor_return(self, call) {
    if (call && (_type_of(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assert_this_initialized(self);
}
function _set_prototype_of(o, p) {
    _set_prototype_of = Object.setPrototypeOf || function setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _set_prototype_of(o, p);
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _is_native_reflect_construct() {
    try {
        var result = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
    } catch (_) {}
    return (_is_native_reflect_construct = function() {
        return !!result;
    })();
}
function createIteratorTransform(IteratorClass) {
    var IteratorTransform = /*#__PURE__*/ function(Transform) {
        "use strict";
        _inherits(IteratorTransform, Transform);
        function IteratorTransform(options) {
            _class_call_check(this, IteratorTransform);
            var _this;
            var _ref;
            var concurrency = (_ref = options === null || options === void 0 ? void 0 : options.concurrency) !== null && _ref !== void 0 ? _ref : Infinity;
            options = options ? _object_spread_props(_object_spread({}, options), {
                objectMode: true
            }) : {
                objectMode: true
            };
            _this = _call_super(this, IteratorTransform, [
                options
            ]);
            _this._concurrency = concurrency;
            return _this;
        }
        var _proto = IteratorTransform.prototype;
        _proto._transform = function _transform(chunk, encoding, callback) {
            var _this = this;
            if (this._stream) {
                this._stream.write(chunk, encoding, callback);
                return;
            }
            this._stream = new _extractbaseiterator.PassThrough();
            this._iterator = new IteratorClass(this._stream);
            var onEntry = function(entry) {
                _this.push(entry);
            };
            var onDone = function(err) {
                if (!_this._iterator) return;
                err || _this.push(null);
                _this._stream = null;
                _this._iterator.destroy();
                _this._iterator = null;
                _this._callback ? _this._callback(err) : _this.end(err);
                _this._callback = null;
            };
            this._iterator.forEach(onEntry, {
                concurrency: this._concurrency
            }, onDone);
            this._stream.write(chunk, encoding, callback);
        };
        _proto._flush = function _flush(callback) {
            if (!this._stream) return callback();
            this._callback = callback;
            this._stream.end();
            this._stream = null;
        };
        _proto.destroy = function destroy(error) {
            if (this._stream) {
                this._stream.end();
                this._stream = null;
            }
            if (this._iterator) {
                var iterator = this._iterator;
                this._iterator = null;
                iterator.destroy(error);
                this.end(error);
            }
            return this;
        };
        return IteratorTransform;
    }(_extractbaseiterator.Transform);
    return IteratorTransform;
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
