// 7z-iterator/dist/cjs/sevenz/ArchiveSource.js
/**
 * ArchiveSource - Abstraction for reading 7z archive data
 *
 * Provides a common interface for reading archive data from either
 * a file descriptor or an in-memory buffer.
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BufferSource () {
        return BufferSource;
    },
    get FileSource () {
        return FileSource;
    }
});
var _extractbaseiterator = require("extract-base-iterator");
var _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
// Helper to create a Readable stream compatible with Node 0.8
function createReadableStream(readFn) {
    var stream = new _extractbaseiterator.Readable();
    stream._read = function() {
        readFn(this);
    };
    return stream;
}
var BufferSource = /*#__PURE__*/ function() {
    "use strict";
    function BufferSource(buffer) {
        _class_call_check(this, BufferSource);
        this.buffer = buffer;
    }
    var _proto = BufferSource.prototype;
    _proto.read = function read(position, length) {
        return this.buffer.slice(position, position + length);
    };
    _proto.getSize = function getSize() {
        return this.buffer.length;
    };
    _proto.close = function close() {
    // Nothing to close for buffer
    };
    /**
   * Create a readable stream for a portion of the buffer.
   * Streams the data in chunks to avoid blocking.
   */ _proto.createReadStream = function createReadStream(offset, length) {
        var buffer = this.buffer;
        var end = Math.min(offset + length, buffer.length);
        var currentPos = offset;
        var chunkSize = 65536; // 64KB chunks
        return createReadableStream(function(stream) {
            if (currentPos >= end) {
                stream.push(null);
                return;
            }
            var toRead = Math.min(chunkSize, end - currentPos);
            var chunk = buffer.slice(currentPos, currentPos + toRead);
            currentPos += toRead;
            stream.push(chunk);
        });
    };
    return BufferSource;
}();
var FileSource = /*#__PURE__*/ function() {
    "use strict";
    function FileSource(fd, size) {
        _class_call_check(this, FileSource);
        this.fd = fd;
        this.size = size;
    }
    var _proto = FileSource.prototype;
    _proto.read = function read(position, length) {
        // For small reads that fit in a single buffer, return directly
        // This is efficient on all Node versions
        if ((0, _extractbaseiterator.canAllocateBufferSize)(length)) {
            return this.readChunk(position, length);
        }
        // For large reads, return a BufferList to avoid large contiguous allocation
        // This enables LZMA1 decompression on old Node versions
        var result = new _extractbaseiterator.BufferList();
        var totalBytesRead = 0;
        var currentPos = position;
        while(totalBytesRead < length){
            var remaining = length - totalBytesRead;
            // Use safe chunk size that works on all Node versions
            var chunkSize = Math.min(remaining, (0, _extractbaseiterator.canAllocateBufferSize)(remaining) ? remaining : 256 * 1024 * 1024);
            var chunk = this.readChunk(currentPos, chunkSize);
            result.append(chunk);
            totalBytesRead += chunk.length;
            currentPos += chunk.length;
            if (chunk.length < chunkSize) {
                break;
            }
        }
        return result;
    };
    _proto.readChunk = function readChunk(position, length) {
        var buf = (0, _extractbaseiterator.allocBuffer)(length);
        var bytesRead = _fs.default.readSync(this.fd, buf, 0, length, position);
        if (bytesRead < length) {
            return buf.slice(0, bytesRead);
        }
        return buf;
    };
    _proto.getSize = function getSize() {
        return this.size;
    };
    _proto.close = function close() {
        try {
            _fs.default.closeSync(this.fd);
        } catch (_e) {
        // Ignore close errors
        }
    };
    /**
   * Create a readable stream for a portion of the file.
   * Uses async fs.read() to avoid blocking the event loop.
   */ _proto.createReadStream = function createReadStream(offset, length) {
        var fd = this.fd;
        var bytesRead = 0;
        var reading = false;
        var finished = false;
        var chunkSize = 65536; // 64KB chunks
        var _streamRef = null;
        var stream = createReadableStream(function(s) {
            _streamRef = s;
            if (reading || finished) return; // Prevent re-entrant reads
            var toRead = Math.min(chunkSize, length - bytesRead);
            if (toRead <= 0) {
                finished = true;
                s.push(null);
                return;
            }
            reading = true;
            var buffer = (0, _extractbaseiterator.allocBuffer)(toRead);
            var currentOffset = offset + bytesRead;
            _fs.default.read(fd, buffer, 0, toRead, currentOffset, function(err, n) {
                reading = false;
                if (err) {
                    // Emit error for Node 0.8 compatibility (no destroy method)
                    s.emit('error', err);
                    finished = true;
                    s.push(null);
                    return;
                }
                if (n === 0) {
                    finished = true;
                    s.push(null);
                } else {
                    bytesRead += n;
                    s.push(buffer.slice(0, n));
                }
            });
        });
        return stream;
    };
    return FileSource;
}();
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
