// 7z-iterator/dist/cjs/sevenz/FolderStreamSplitter.js
/**
 * FolderStreamSplitter - Splits a decompressed folder stream into individual file streams
 *
 * For multi-file solid archives, the folder is decompressed as a single stream.
 * This class splits that stream into individual file streams based on known file boundaries.
 *
 * Features:
 * - Lazy stream creation (streams created on first access)
 * - Backpressure propagation (returns false when downstream is full)
 * - Running CRC verification per file
 * - Automatic cleanup of completed streams
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "FolderStreamSplitter", {
    enumerable: true,
    get: function() {
        return FolderStreamSplitter;
    }
});
var _extractbaseiterator = require("extract-base-iterator");
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
var FolderStreamSplitter = /*#__PURE__*/ function() {
    "use strict";
    function FolderStreamSplitter(options) {
        _class_call_check(this, FolderStreamSplitter);
        var fileSizes = options.fileSizes;
        var verifyCrc = options.verifyCrc !== undefined ? options.verifyCrc : true;
        var expectedCrcs = options.expectedCrcs || [];
        this.verifyCrc = verifyCrc;
        this.expectedCrcs = expectedCrcs;
        this.currentFileIndex = 0;
        this.bytesWritten = 0;
        this.finished = false;
        this.error = null;
        this.drainCallbacks = [];
        this._needsDrain = false;
        // Calculate cumulative boundaries
        this.fileBoundaries = [
            0
        ];
        for(var i = 0; i < fileSizes.length; i++){
            this.fileBoundaries.push(this.fileBoundaries[this.fileBoundaries.length - 1] + fileSizes[i]);
        }
        // Initialize streams array (lazy creation - all null initially)
        this.fileStreams = [];
        this.fileCrcs = [];
        for(var i1 = 0; i1 < fileSizes.length; i1++){
            this.fileStreams.push(null);
            this.fileCrcs.push(0);
        }
        // Set first file boundary
        this.currentFileEnd = this.fileBoundaries[1] || 0;
    }
    var _proto = FolderStreamSplitter.prototype;
    /**
   * Write decompressed data chunk. Data is routed to appropriate file stream(s).
   * Returns false if backpressure should be applied (downstream is full).
   */ _proto.write = function write(chunk) {
        var _this = this;
        if (this.finished || this.error) return true;
        var offset = 0;
        var canContinue = true;
        while(offset < chunk.length && this.currentFileIndex < this.fileStreams.length){
            var remaining = chunk.length - offset;
            var neededForFile = this.currentFileEnd - this.bytesWritten;
            var toWrite = Math.min(remaining, neededForFile);
            if (toWrite > 0) {
                var fileChunk = chunk.slice(offset, offset + toWrite);
                // Ensure stream exists (lazy creation)
                var fileStream = this.ensureFileStream(this.currentFileIndex);
                // Update CRC
                if (this.verifyCrc) {
                    this.fileCrcs[this.currentFileIndex] = (0, _extractbaseiterator.crc32)(fileChunk, this.fileCrcs[this.currentFileIndex]);
                }
                // Write to file stream, track backpressure
                if (!fileStream.write(fileChunk)) {
                    canContinue = false;
                    this._needsDrain = true;
                    fileStream.once('drain', function() {
                        _this._needsDrain = false;
                        _this.notifyDrain();
                    });
                }
            }
            this.bytesWritten += toWrite;
            offset += toWrite;
            // Check if current file is complete
            if (this.bytesWritten >= this.currentFileEnd) {
                this.finishCurrentFile();
            }
        }
        return canContinue;
    };
    /**
   * Ensure stream exists for file index (lazy creation)
   */ _proto.ensureFileStream = function ensureFileStream(fileIndex) {
        var stream = this.fileStreams[fileIndex];
        if (!stream) {
            stream = new _extractbaseiterator.PassThrough();
            this.fileStreams[fileIndex] = stream;
        }
        return stream;
    };
    /**
   * Complete current file and move to next
   */ _proto.finishCurrentFile = function finishCurrentFile() {
        var fileStream = this.fileStreams[this.currentFileIndex];
        // Verify CRC if enabled
        if (this.verifyCrc) {
            var expectedCrc = this.expectedCrcs[this.currentFileIndex];
            if (expectedCrc !== undefined && this.fileCrcs[this.currentFileIndex] !== expectedCrc) {
                var err = new Error("CRC mismatch for file ".concat(this.currentFileIndex, ": expected ").concat(expectedCrc.toString(16), ", got ").concat(this.fileCrcs[this.currentFileIndex].toString(16)));
                this.emitError(err);
                return;
            }
        }
        // End this file's stream
        if (fileStream) {
            fileStream.end();
        }
        // Release reference for GC
        this.fileStreams[this.currentFileIndex] = null;
        // Move to next file
        this.currentFileIndex++;
        if (this.currentFileIndex < this.fileBoundaries.length - 1) {
            this.currentFileEnd = this.fileBoundaries[this.currentFileIndex + 1];
        }
    };
    /**
   * Signal end of decompressed data
   */ _proto.end = function end() {
        if (this.finished) return;
        this.finished = true;
        // End any remaining streams
        for(var i = this.currentFileIndex; i < this.fileStreams.length; i++){
            var stream = this.fileStreams[i];
            if (stream) {
                stream.end();
            }
            this.fileStreams[i] = null;
        }
    };
    /**
   * Emit error to all pending file streams
   */ _proto.emitError = function emitError(err) {
        this.error = err;
        for(var i = this.currentFileIndex; i < this.fileStreams.length; i++){
            var stream = this.fileStreams[i];
            if (stream) {
                stream.emit('error', err);
                stream.end();
            }
            this.fileStreams[i] = null;
        }
    };
    /**
   * Get the stream for a specific file by index.
   * Stream is created lazily on first access.
   */ _proto.getFileStream = function getFileStream(fileIndex) {
        if (fileIndex < 0 || fileIndex >= this.fileBoundaries.length - 1) {
            throw new Error("Invalid file index: ".concat(fileIndex));
        }
        // Check if file already completed
        if (fileIndex < this.currentFileIndex) {
            throw new Error("File ".concat(fileIndex, " already completed - streams must be accessed in order"));
        }
        return this.ensureFileStream(fileIndex);
    };
    /**
   * Register callback for when backpressure clears
   */ _proto.onDrain = function onDrain(callback) {
        if (!this._needsDrain) {
            callback();
        } else {
            this.drainCallbacks.push(callback);
        }
    };
    /**
   * Notify all drain callbacks
   */ _proto.notifyDrain = function notifyDrain() {
        var callbacks = this.drainCallbacks;
        this.drainCallbacks = [];
        for(var i = 0; i < callbacks.length; i++){
            callbacks[i]();
        }
    };
    /**
   * Check if a specific file's stream has been fully written
   */ _proto.isFileComplete = function isFileComplete(fileIndex) {
        return fileIndex < this.currentFileIndex;
    };
    /**
   * Check if splitter has encountered an error
   */ _proto.getError = function getError() {
        return this.error;
    };
    _create_class(FolderStreamSplitter, [
        {
            key: "fileCount",
            get: /**
   * Get total number of files in this folder
   */ function get() {
                return this.fileBoundaries.length - 1;
            }
        }
    ]);
    return FolderStreamSplitter;
}();
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
