// 7z-iterator/dist/cjs/sevenz/headers.js
// 7z header parsing
// Reference: https://py7zr.readthedocs.io/en/latest/archive_format.html
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get parseEncodedHeader () {
        return parseEncodedHeader;
    },
    get parseHeaderContent () {
        return parseHeaderContent;
    },
    get parseSignatureHeader () {
        return parseSignatureHeader;
    }
});
var _extractbaseiterator = require("extract-base-iterator");
var _constantsts = require("./constants.js");
var _NumberCodects = require("./NumberCodec.js");
function parseSignatureHeader(buf) {
    // Verify magic bytes
    if (!(0, _extractbaseiterator.bufferEquals)(buf, 0, _constantsts.SEVENZ_MAGIC)) {
        throw (0, _constantsts.createCodedError)('Not a valid 7z archive', _constantsts.ErrorCode.INVALID_SIGNATURE);
    }
    // Read version
    var majorVersion = buf[6];
    var minorVersion = buf[7];
    // Version check - we support 0.x (current is 0.4)
    if (majorVersion > 0) {
        throw (0, _constantsts.createCodedError)("Unsupported 7z version: ".concat(majorVersion, ".").concat(minorVersion), _constantsts.ErrorCode.UNSUPPORTED_VERSION);
    }
    // Read start header CRC (CRC of the next 20 bytes)
    var startHeaderCRC = buf.readUInt32LE(8);
    // Verify start header CRC
    if (!(0, _extractbaseiterator.verifyCrc32Region)(buf, 12, 20, startHeaderCRC)) {
        throw (0, _constantsts.createCodedError)('Start header CRC mismatch', _constantsts.ErrorCode.CRC_MISMATCH);
    }
    // Read next header location
    var nextHeaderOffset = (0, _extractbaseiterator.readUInt64LE)(buf, 12);
    var nextHeaderSize = (0, _extractbaseiterator.readUInt64LE)(buf, 20);
    var nextHeaderCRC = buf.readUInt32LE(28);
    return {
        majorVersion: majorVersion,
        minorVersion: minorVersion,
        startHeaderCRC: startHeaderCRC,
        nextHeaderOffset: nextHeaderOffset,
        nextHeaderSize: nextHeaderSize,
        nextHeaderCRC: nextHeaderCRC
    };
}
function parseEncodedHeader(buf, expectedCRC) {
    // Verify CRC
    if (!(0, _extractbaseiterator.verifyCrc32Region)(buf, 0, buf.length, expectedCRC)) {
        throw (0, _constantsts.createCodedError)('Encoded header CRC mismatch', _constantsts.ErrorCode.CRC_MISMATCH);
    }
    var offset = 0;
    // Read property ID
    var propertyId = buf[offset++];
    // Handle kEncodedHeader - means the header itself is compressed
    if (propertyId === _constantsts.PropertyId.kEncodedHeader) {
        // Return indicator that we need to decompress
        throw (0, _constantsts.createCodedError)('Compressed header - needs decompression first', _constantsts.ErrorCode.COMPRESSED_HEADER);
    }
    // Should be kHeader
    if (propertyId !== _constantsts.PropertyId.kHeader) {
        throw (0, _constantsts.createCodedError)("Expected kHeader, got ".concat(propertyId), _constantsts.ErrorCode.CORRUPT_HEADER);
    }
    // Parse header contents (after kHeader byte)
    return parseHeaderContent(buf, offset);
}
function parseHeaderContent(buf, offset) {
    var result = {
        filesInfo: []
    };
    // Parse header contents
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        switch(propertyId){
            case _constantsts.PropertyId.kArchiveProperties:
                offset = skipArchiveProperties(buf, offset);
                break;
            case _constantsts.PropertyId.kAdditionalStreamsInfo:
                // Additional streams - skip for now
                offset = skipStreamsInfo(buf, offset);
                break;
            case _constantsts.PropertyId.kMainStreamsInfo:
                {
                    var streamsResult = parseStreamsInfo(buf, offset);
                    result.streamsInfo = streamsResult.info;
                    offset = streamsResult.offset;
                    break;
                }
            case _constantsts.PropertyId.kFilesInfo:
                {
                    var filesResult = parseFilesInfo(buf, offset);
                    result.filesInfo = filesResult.files;
                    offset = filesResult.offset;
                    break;
                }
            default:
                throw (0, _constantsts.createCodedError)("Unknown property ID in header: ".concat(propertyId), _constantsts.ErrorCode.CORRUPT_HEADER);
        }
    }
    return result;
}
/**
 * Parse StreamsInfo block
 */ function parseStreamsInfo(buf, offset) {
    var info = {
        packPos: 0,
        packSizes: [],
        folders: [],
        numUnpackStreamsPerFolder: [],
        unpackSizes: []
    };
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        switch(propertyId){
            case _constantsts.PropertyId.kPackInfo:
                {
                    var packResult = parsePackInfo(buf, offset);
                    info.packPos = packResult.packPos;
                    info.packSizes = packResult.packSizes;
                    info.packCRCs = packResult.packCRCs;
                    offset = packResult.offset;
                    break;
                }
            case _constantsts.PropertyId.kUnpackInfo:
                {
                    var unpackResult = parseUnpackInfo(buf, offset);
                    info.folders = unpackResult.folders;
                    offset = unpackResult.offset;
                    break;
                }
            case _constantsts.PropertyId.kSubStreamsInfo:
                {
                    var subResult = parseSubStreamsInfo(buf, offset, info.folders);
                    info.numUnpackStreamsPerFolder = subResult.numUnpackStreamsPerFolder;
                    info.unpackSizes = subResult.unpackSizes;
                    info.unpackCRCs = subResult.unpackCRCs;
                    offset = subResult.offset;
                    break;
                }
            default:
                throw (0, _constantsts.createCodedError)("Unknown property ID in StreamsInfo: ".concat(propertyId), _constantsts.ErrorCode.CORRUPT_HEADER);
        }
    }
    // If no SubStreamsInfo, each folder produces one file
    if (info.unpackSizes.length === 0 && info.folders.length > 0) {
        for(var i = 0; i < info.folders.length; i++){
            var folder = info.folders[i];
            // Get the final unpack size (last coder's output)
            var finalSize = folder.unpackSizes[folder.unpackSizes.length - 1];
            info.unpackSizes.push(finalSize);
            info.numUnpackStreamsPerFolder.push(1);
        }
    }
    return {
        info: info,
        offset: offset
    };
}
/**
 * Parse PackInfo block
 */ function parsePackInfo(buf, offset) {
    // Pack position
    var packPosResult = (0, _NumberCodects.readNumber)(buf, offset);
    var packPos = packPosResult.value;
    offset += packPosResult.bytesRead;
    // Number of pack streams
    var numPackResult = (0, _NumberCodects.readNumber)(buf, offset);
    var numPackStreams = numPackResult.value;
    offset += numPackResult.bytesRead;
    var packSizes = [];
    var packCRCs;
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        if (propertyId === _constantsts.PropertyId.kSize) {
            for(var i = 0; i < numPackStreams; i++){
                var sizeResult = (0, _NumberCodects.readNumber)(buf, offset);
                packSizes.push(sizeResult.value);
                offset += sizeResult.bytesRead;
            }
        } else if (propertyId === _constantsts.PropertyId.kCRC) {
            packCRCs = [];
            var definedResult = (0, _NumberCodects.readDefinedVector)(buf, offset, numPackStreams);
            offset += definedResult.bytesRead;
            for(var j = 0; j < numPackStreams; j++){
                if (definedResult.defined[j]) {
                    packCRCs.push(buf.readUInt32LE(offset));
                    offset += 4;
                } else {
                    packCRCs.push(0);
                }
            }
        }
    }
    return {
        packPos: packPos,
        packSizes: packSizes,
        packCRCs: packCRCs,
        offset: offset
    };
}
/**
 * Parse UnpackInfo block
 */ function parseUnpackInfo(buf, offset) {
    var folders = [];
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        if (propertyId === _constantsts.PropertyId.kFolder) {
            // Number of folders
            var numFoldersResult = (0, _NumberCodects.readNumber)(buf, offset);
            var numFolders = numFoldersResult.value;
            offset += numFoldersResult.bytesRead;
            // External flag
            var external = buf[offset++];
            if (external !== 0) {
                throw (0, _constantsts.createCodedError)('External folders not supported', _constantsts.ErrorCode.CORRUPT_HEADER);
            }
            // Parse each folder
            for(var i = 0; i < numFolders; i++){
                var folderResult = parseFolder(buf, offset);
                folders.push(folderResult.folder);
                offset = folderResult.offset;
            }
        } else if (propertyId === _constantsts.PropertyId.kCodersUnpackSize) {
            // Unpack sizes for each coder output
            for(var j = 0; j < folders.length; j++){
                var folder = folders[j];
                folder.unpackSizes = [];
                // One unpack size per coder output stream
                var numOutputs = 0;
                for(var k = 0; k < folder.coders.length; k++){
                    numOutputs += folder.coders[k].numOutStreams;
                }
                for(var l = 0; l < numOutputs; l++){
                    var sizeResult = (0, _NumberCodects.readNumber)(buf, offset);
                    folder.unpackSizes.push(sizeResult.value);
                    offset += sizeResult.bytesRead;
                }
            }
        } else if (propertyId === _constantsts.PropertyId.kCRC) {
            // CRCs for folders
            var definedResult = (0, _NumberCodects.readDefinedVector)(buf, offset, folders.length);
            offset += definedResult.bytesRead;
            for(var m = 0; m < folders.length; m++){
                folders[m].hasCRC = definedResult.defined[m];
                if (definedResult.defined[m]) {
                    folders[m].unpackCRC = buf.readUInt32LE(offset);
                    offset += 4;
                }
            }
        }
    }
    return {
        folders: folders,
        offset: offset
    };
}
/**
 * Parse a single Folder structure
 */ function parseFolder(buf, offset) {
    // Number of coders
    var numCodersResult = (0, _NumberCodects.readNumber)(buf, offset);
    var numCoders = numCodersResult.value;
    offset += numCodersResult.bytesRead;
    var coders = [];
    var numInStreamsTotal = 0;
    var numOutStreamsTotal = 0;
    for(var i = 0; i < numCoders; i++){
        var flags = buf[offset++];
        var idSize = flags & 0x0f;
        var isComplex = (flags & 0x10) !== 0;
        var hasAttributes = (flags & 0x20) !== 0;
        // Read codec ID
        var id = [];
        for(var j = 0; j < idSize; j++){
            id.push(buf[offset++]);
        }
        var numInStreams = 1;
        var numOutStreams = 1;
        if (isComplex) {
            var inResult = (0, _NumberCodects.readNumber)(buf, offset);
            numInStreams = inResult.value;
            offset += inResult.bytesRead;
            var outResult = (0, _NumberCodects.readNumber)(buf, offset);
            numOutStreams = outResult.value;
            offset += outResult.bytesRead;
        }
        var properties = void 0;
        if (hasAttributes) {
            var propsLenResult = (0, _NumberCodects.readNumber)(buf, offset);
            offset += propsLenResult.bytesRead;
            properties = buf.slice(offset, offset + propsLenResult.value);
            offset += propsLenResult.value;
        }
        coders.push({
            id: id,
            numInStreams: numInStreams,
            numOutStreams: numOutStreams,
            properties: properties
        });
        numInStreamsTotal += numInStreams;
        numOutStreamsTotal += numOutStreams;
    }
    // Bind pairs
    var numBindPairs = numOutStreamsTotal - 1;
    var bindPairs = [];
    for(var k = 0; k < numBindPairs; k++){
        var inIndexResult = (0, _NumberCodects.readNumber)(buf, offset);
        offset += inIndexResult.bytesRead;
        var outIndexResult = (0, _NumberCodects.readNumber)(buf, offset);
        offset += outIndexResult.bytesRead;
        bindPairs.push({
            inIndex: inIndexResult.value,
            outIndex: outIndexResult.value
        });
    }
    // Packed stream indices
    var numPackedStreams = numInStreamsTotal - numBindPairs;
    var packedStreams = [];
    if (numPackedStreams === 1) {
        // Find the unbound input stream
        for(var m = 0; m < numInStreamsTotal; m++){
            var isBound = false;
            for(var n = 0; n < bindPairs.length; n++){
                if (bindPairs[n].inIndex === m) {
                    isBound = true;
                    break;
                }
            }
            if (!isBound) {
                packedStreams.push(m);
                break;
            }
        }
    } else {
        for(var p = 0; p < numPackedStreams; p++){
            var indexResult = (0, _NumberCodects.readNumber)(buf, offset);
            packedStreams.push(indexResult.value);
            offset += indexResult.bytesRead;
        }
    }
    return {
        folder: {
            coders: coders,
            bindPairs: bindPairs,
            packedStreams: packedStreams,
            unpackSizes: [],
            hasCRC: false
        },
        offset: offset
    };
}
/**
 * Parse SubStreamsInfo block
 */ function parseSubStreamsInfo(buf, offset, folders) {
    var numUnpackStreamsPerFolder = [];
    var unpackSizes = [];
    var unpackCRCs;
    // Default: 1 file per folder
    for(var i = 0; i < folders.length; i++){
        numUnpackStreamsPerFolder.push(1);
    }
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        if (propertyId === _constantsts.PropertyId.kNumUnpackStream) {
            for(var j = 0; j < folders.length; j++){
                var numResult = (0, _NumberCodects.readNumber)(buf, offset);
                numUnpackStreamsPerFolder[j] = numResult.value;
                offset += numResult.bytesRead;
            }
        } else if (propertyId === _constantsts.PropertyId.kSize) {
            for(var k = 0; k < folders.length; k++){
                var numStreams = numUnpackStreamsPerFolder[k];
                if (numStreams === 0) continue;
                // Read sizes for all but last stream in folder (last is calculated)
                var remaining = folders[k].unpackSizes[folders[k].unpackSizes.length - 1];
                for(var l = 0; l < numStreams - 1; l++){
                    var sizeResult = (0, _NumberCodects.readNumber)(buf, offset);
                    unpackSizes.push(sizeResult.value);
                    remaining -= sizeResult.value;
                    offset += sizeResult.bytesRead;
                }
                // Last stream size is remainder
                unpackSizes.push(remaining);
            }
        } else if (propertyId === _constantsts.PropertyId.kCRC) {
            // Count files that need CRC
            var numFiles = 0;
            for(var m = 0; m < folders.length; m++){
                var numStreamsInFolder = numUnpackStreamsPerFolder[m];
                // Only count if folder doesn't have CRC or has multiple streams
                if (!folders[m].hasCRC || numStreamsInFolder > 1) {
                    numFiles += numStreamsInFolder;
                }
            }
            unpackCRCs = [];
            var definedResult = (0, _NumberCodects.readDefinedVector)(buf, offset, numFiles);
            offset += definedResult.bytesRead;
            for(var n = 0; n < numFiles; n++){
                if (definedResult.defined[n]) {
                    unpackCRCs.push(buf.readUInt32LE(offset));
                    offset += 4;
                } else {
                    unpackCRCs.push(0);
                }
            }
        }
    }
    // If no sizes specified, use folder unpack sizes
    if (unpackSizes.length === 0) {
        for(var p = 0; p < folders.length; p++){
            var folder = folders[p];
            unpackSizes.push(folder.unpackSizes[folder.unpackSizes.length - 1]);
        }
    }
    return {
        numUnpackStreamsPerFolder: numUnpackStreamsPerFolder,
        unpackSizes: unpackSizes,
        unpackCRCs: unpackCRCs,
        offset: offset
    };
}
/**
 * Parse FilesInfo block
 */ function parseFilesInfo(buf, offset) {
    // Number of files
    var numFilesResult = (0, _NumberCodects.readNumber)(buf, offset);
    var numFiles = numFilesResult.value;
    offset += numFilesResult.bytesRead;
    // Initialize files array
    var files = [];
    for(var i = 0; i < numFiles; i++){
        files.push({
            name: '',
            size: 0,
            isDirectory: false,
            isAntiFile: false,
            hasStream: true
        });
    }
    var emptyStreamFlags = [];
    var emptyFileFlags = [];
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        // Read property size
        var propSizeResult = (0, _NumberCodects.readNumber)(buf, offset);
        var propSize = propSizeResult.value;
        offset += propSizeResult.bytesRead;
        var propEnd = offset + propSize;
        switch(propertyId){
            case _constantsts.PropertyId.kEmptyStream:
                emptyStreamFlags = readBoolVector(buf, offset, numFiles);
                // Mark files that don't have streams
                for(var j = 0; j < numFiles; j++){
                    files[j].hasStream = !emptyStreamFlags[j];
                }
                break;
            case _constantsts.PropertyId.kEmptyFile:
                {
                    var numEmptyStreams = 0;
                    for(var k = 0; k < emptyStreamFlags.length; k++){
                        if (emptyStreamFlags[k]) numEmptyStreams++;
                    }
                    emptyFileFlags = readBoolVector(buf, offset, numEmptyStreams);
                    break;
                }
            case _constantsts.PropertyId.kAnti:
                {
                    var numAnti = 0;
                    for(var l = 0; l < emptyStreamFlags.length; l++){
                        if (emptyStreamFlags[l]) numAnti++;
                    }
                    var antiFlags = readBoolVector(buf, offset, numAnti);
                    var antiIdx = 0;
                    for(var m = 0; m < numFiles; m++){
                        if (emptyStreamFlags[m]) {
                            files[m].isAntiFile = antiFlags[antiIdx++];
                        }
                    }
                    break;
                }
            case _constantsts.PropertyId.kName:
                offset = parseFileNames(buf, offset, files);
                break;
            case _constantsts.PropertyId.kCTime:
                offset = parseFileTimes(buf, offset, files, 'ctime');
                break;
            case _constantsts.PropertyId.kATime:
                offset = parseFileTimes(buf, offset, files, 'atime');
                break;
            case _constantsts.PropertyId.kMTime:
                offset = parseFileTimes(buf, offset, files, 'mtime');
                break;
            case _constantsts.PropertyId.kWinAttributes:
                offset = parseAttributes(buf, offset, files);
                break;
            case _constantsts.PropertyId.kDummy:
                break;
            default:
                break;
        }
        offset = propEnd;
    }
    // Determine directories from empty stream + not empty file
    var emptyIdx = 0;
    for(var n = 0; n < numFiles; n++){
        if (emptyStreamFlags[n]) {
            // Empty stream - could be directory or empty file
            if (emptyIdx < emptyFileFlags.length && emptyFileFlags[emptyIdx]) {
                files[n].isDirectory = false; // Empty file
            } else {
                files[n].isDirectory = true; // Directory
            }
            emptyIdx++;
        }
    }
    return {
        files: files,
        offset: offset
    };
}
/**
 * Read a boolean vector (bit-packed)
 */ function readBoolVector(buf, offset, count) {
    var result = [];
    var byteIdx = 0;
    var bitMask = 0x80;
    for(var i = 0; i < count; i++){
        result.push((buf[offset + byteIdx] & bitMask) !== 0);
        bitMask = bitMask >>> 1;
        if (bitMask === 0) {
            bitMask = 0x80;
            byteIdx++;
        }
    }
    return result;
}
/**
 * Parse file names (UTF-16LE encoded)
 */ function parseFileNames(buf, offset, files) {
    // External flag
    var external = buf[offset++];
    if (external !== 0) {
        throw (0, _constantsts.createCodedError)('External file names not supported', _constantsts.ErrorCode.CORRUPT_HEADER);
    }
    // Names are UTF-16LE, null-terminated
    for(var i = 0; i < files.length; i++){
        var nameChars = [];
        while(offset < buf.length){
            var charCode = buf.readUInt16LE(offset);
            offset += 2;
            if (charCode === 0) break;
            nameChars.push(charCode);
        }
        files[i].name = String.fromCharCode.apply(null, nameChars);
    }
    return offset;
}
/**
 * Parse file times (Windows FILETIME format)
 */ function parseFileTimes(buf, offset, files, timeType) {
    // Read defined vector (allDefined byte + optional bitmask)
    var definedResult = (0, _NumberCodects.readDefinedVector)(buf, offset, files.length);
    offset += definedResult.bytesRead;
    // External flag - 0x00 means data follows inline, non-zero means external stream
    var external = buf[offset++];
    if (external !== 0) {
        throw (0, _constantsts.createCodedError)('External file times not supported', _constantsts.ErrorCode.UNSUPPORTED_FEATURE);
    }
    // Read times
    for(var i = 0; i < files.length; i++){
        if (definedResult.defined[i]) {
            var filetime = (0, _extractbaseiterator.readUInt64LE)(buf, offset);
            offset += 8;
            // Convert FILETIME (100ns since 1601) to JavaScript Date
            // FILETIME epoch: 1601-01-01
            // JS Date epoch: 1970-01-01
            // Difference: 11644473600 seconds
            var ms = filetime / 10000 - 11644473600000;
            files[i][timeType] = new Date(ms);
        }
    }
    return offset;
}
/**
 * Parse Windows file attributes
 */ function parseAttributes(buf, offset, files) {
    // Read defined vector (allDefined byte + optional bitmask)
    var definedResult = (0, _NumberCodects.readDefinedVector)(buf, offset, files.length);
    offset += definedResult.bytesRead;
    // External flag - 0x00 means data follows inline, non-zero means external stream
    var external = buf[offset++];
    if (external !== 0) {
        throw (0, _constantsts.createCodedError)('External file attributes not supported', _constantsts.ErrorCode.UNSUPPORTED_FEATURE);
    }
    // Read attributes
    for(var i = 0; i < files.length; i++){
        if (definedResult.defined[i]) {
            files[i].attributes = buf.readUInt32LE(offset);
            offset += 4;
        }
    }
    return offset;
}
/**
 * Skip archive properties block
 */ function skipArchiveProperties(buf, offset) {
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        var sizeResult = (0, _NumberCodects.readNumber)(buf, offset);
        offset += sizeResult.bytesRead + sizeResult.value;
    }
    return offset;
}
/**
 * Skip streams info block (for additional streams)
 */ function skipStreamsInfo(buf, offset) {
    while(offset < buf.length){
        var propertyId = buf[offset++];
        if (propertyId === _constantsts.PropertyId.kEnd) {
            break;
        }
        // For nested structures, recurse
        if (propertyId === _constantsts.PropertyId.kPackInfo || propertyId === _constantsts.PropertyId.kUnpackInfo || propertyId === _constantsts.PropertyId.kSubStreamsInfo) {
            offset = skipStreamsInfo(buf, offset);
        }
    }
    return offset;
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
