// 7z-iterator/dist/cjs/sevenz/NumberCodec.js
// Variable-length integer encoding for 7z format
// Reference: https://py7zr.readthedocs.io/en/latest/archive_format.html
//
// 7z uses a space-efficient encoding where the first byte determines length:
// 0xxxxxxx                    -> 1 byte  (0-127)
// 10xxxxxx xxxxxxxx           -> 2 bytes (0-16383)
// 110xxxxx + 2 bytes          -> 3 bytes
// 1110xxxx + 3 bytes          -> 4 bytes
// 11110xxx + 4 bytes          -> 5 bytes
// 111110xx + 5 bytes          -> 6 bytes
// 1111110x + 6 bytes          -> 7 bytes
// 11111110 + 7 bytes          -> 8 bytes
// 11111111 + 8 bytes          -> 9 bytes (full 64-bit)
//
// NOTE: Returns JavaScript number which is accurate up to 2^53 - 1 (~9 PB).
// This covers all practical file sizes.
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get encodedSize () {
        return encodedSize;
    },
    get readBoolean () {
        return readBoolean;
    },
    get readDefinedVector () {
        return readDefinedVector;
    },
    get readNumber () {
        return readNumber;
    },
    get readNumberArray () {
        return readNumberArray;
    },
    get readRawNumber () {
        return readRawNumber;
    }
});
var _extractbaseiterator = require("extract-base-iterator");
function readNumber(buf, offset) {
    var firstByte = buf[offset];
    // Special case: 0xFF means 8 extra bytes (full 64-bit value)
    if (firstByte === 0xff) {
        return {
            value: (0, _extractbaseiterator.readUInt64LE)(buf, offset + 1),
            bytesRead: 9
        };
    }
    // Determine number of extra bytes based on first byte value thresholds
    // This matches the 7z format specification
    var extraBytes = 0;
    var mask = 0x80;
    if (firstByte <= 0x7f) {
        extraBytes = 0;
        mask = 0x80;
    } else if (firstByte <= 0xbf) {
        extraBytes = 1;
        mask = 0x40;
    } else if (firstByte <= 0xdf) {
        extraBytes = 2;
        mask = 0x20;
    } else if (firstByte <= 0xef) {
        extraBytes = 3;
        mask = 0x10;
    } else if (firstByte <= 0xf7) {
        extraBytes = 4;
        mask = 0x08;
    } else if (firstByte <= 0xfb) {
        extraBytes = 5;
        mask = 0x04;
    } else if (firstByte <= 0xfd) {
        extraBytes = 6;
        mask = 0x02;
    } else {
        // 0xFE
        extraBytes = 7;
        mask = 0x01;
    }
    // Get high part from first byte (bits below the length indicator)
    var highPart = firstByte & mask - 1;
    // Read extra bytes as LITTLE-ENDIAN
    var value = 0;
    for(var i = 0; i < extraBytes; i++){
        value += buf[offset + 1 + i] * Math.pow(256, i);
    }
    // Combine: value + (highPart << (extraBytes * 8))
    value += highPart * Math.pow(256, extraBytes);
    return {
        value: value,
        bytesRead: 1 + extraBytes
    };
}
function readRawNumber(buf, offset) {
    return (0, _extractbaseiterator.readUInt64LE)(buf, offset);
}
function encodedSize(value) {
    if (value < 0x80) return 1; // 7 bits
    if (value < 0x4000) return 2; // 14 bits
    if (value < 0x200000) return 3; // 21 bits
    if (value < 0x10000000) return 4; // 28 bits
    if (value < 0x800000000) return 5; // 35 bits
    if (value < 0x40000000000) return 6; // 42 bits
    if (value < 0x2000000000000) return 7; // 49 bits
    // 2^56 = 72057594037927936 (use calculated value to avoid precision loss)
    if (value < 72057594037927936) return 8; // 56 bits
    return 9; // 64 bits
}
function readBoolean(buf, offset) {
    return buf[offset] !== 0;
}
function readDefinedVector(buf, offset, count) {
    var allDefined = buf[offset] !== 0;
    var bytesRead = 1;
    var defined = [];
    if (allDefined) {
        // All items are defined
        for(var i = 0; i < count; i++){
            defined.push(true);
        }
    } else {
        // Read bitmask
        var bitsNeeded = count;
        var bytesNeeded = Math.ceil(bitsNeeded / 8);
        for(var byteIdx = 0; byteIdx < bytesNeeded; byteIdx++){
            var byte = buf[offset + 1 + byteIdx];
            for(var bit = 7; bit >= 0 && defined.length < count; bit--){
                defined.push((byte & 1 << bit) !== 0);
            }
        }
        bytesRead += bytesNeeded;
    }
    return {
        defined: defined,
        bytesRead: bytesRead
    };
}
function readNumberArray(buf, offset, count) {
    var values = [];
    var totalBytesRead = 0;
    for(var i = 0; i < count; i++){
        var result = readNumber(buf, offset + totalBytesRead);
        values.push(result.value);
        totalBytesRead += result.bytesRead;
    }
    return {
        values: values,
        bytesRead: totalBytesRead
    };
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
