// 7z-iterator/dist/cjs/sevenz/SevenZipParser.js
/**
 * SevenZipParser - Main 7z archive parser
 *
 * Handles reading archive structure and providing file streams.
 *
 * Parser Flow:
 * 1. Read signature header (32 bytes) to get header location
 * 2. Read encoded header from nextHeaderOffset
 * 3. If header is compressed, decompress it first
 * 4. Parse streams info (folder structure, pack positions)
 * 5. Parse files info (names, sizes, attributes)
 * 6. Build entry list for iteration
 *
 * Decompression:
 * - 7z uses "folders" as decompression units
 * - Solid archives: multiple files share one folder (decompress once)
 * - Non-solid: one file per folder
 * - Supports LZMA, LZMA2, COPY, BCJ2, and other codecs
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get BufferSource () {
        return _ArchiveSourcets.BufferSource;
    },
    get FileSource () {
        return _ArchiveSourcets.FileSource;
    },
    get SevenZipParser () {
        return SevenZipParser;
    }
});
var _calloncefn = /*#__PURE__*/ _interop_require_default(require("call-once-fn"));
var _extractbaseiterator = require("extract-base-iterator");
var _deferts = require("../lib/defer.js");
var _indexts = require("./codecs/index.js");
var _FolderStreamSplitterts = require("./FolderStreamSplitter.js");
var _constantsts = require("./constants.js");
var _headersts = require("./headers.js");
var _NumberCodects = require("./NumberCodec.js");
var _ArchiveSourcets = require("./ArchiveSource.js");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _async_to_generator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _instanceof(left, right) {
    if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
        return !!right[Symbol.hasInstance](left);
    } else {
        return left instanceof right;
    }
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _ts_generator(thisArg, body) {
    var f, y, t, _ = {
        label: 0,
        sent: function() {
            if (t[0] & 1) throw t[1];
            return t[1];
        },
        trys: [],
        ops: []
    }, g = Object.create((typeof Iterator === "function" ? Iterator : Object).prototype), d = Object.defineProperty;
    return d(g, "next", {
        value: verb(0)
    }), d(g, "throw", {
        value: verb(1)
    }), d(g, "return", {
        value: verb(2)
    }), typeof Symbol === "function" && d(g, Symbol.iterator, {
        value: function() {
            return this;
        }
    }), g;
    function verb(n) {
        return function(v) {
            return step([
                n,
                v
            ]);
        };
    }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while(g && (g = 0, op[0] && (_ = 0)), _)try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [
                op[0] & 2,
                t.value
            ];
            switch(op[0]){
                case 0:
                case 1:
                    t = op;
                    break;
                case 4:
                    _.label++;
                    return {
                        value: op[1],
                        done: false
                    };
                case 5:
                    _.label++;
                    y = op[1];
                    op = [
                        0
                    ];
                    continue;
                case 7:
                    op = _.ops.pop();
                    _.trys.pop();
                    continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                        _ = 0;
                        continue;
                    }
                    if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
                        _.label = op[1];
                        break;
                    }
                    if (op[0] === 6 && _.label < t[1]) {
                        _.label = t[1];
                        t = op;
                        break;
                    }
                    if (t && _.label < t[2]) {
                        _.label = t[2];
                        _.ops.push(op);
                        break;
                    }
                    if (t[2]) _.ops.pop();
                    _.trys.pop();
                    continue;
            }
            op = body.call(thisArg, _);
        } catch (e) {
            op = [
                6,
                e
            ];
            y = 0;
        } finally{
            f = t = 0;
        }
        if (op[0] & 5) throw op[1];
        return {
            value: op[0] ? op[1] : void 0,
            done: true
        };
    }
}
var SevenZipParser = /*#__PURE__*/ function() {
    "use strict";
    function SevenZipParser(source) {
        _class_call_check(this, SevenZipParser);
        this.signature = null;
        this.streamsInfo = null;
        this.filesInfo = [];
        this.entries = [];
        this.parsed = false;
        // Smart cache for decompressed solid blocks
        // Only caches when multiple files share a block, releases when last file extracted
        this.decompressedCache = {};
        // Track files per folder and how many have been extracted
        this.filesPerFolder = {};
        this.extractedPerFolder = {};
        // Splitter cache for multi-file folder streaming (Phase 2)
        this.folderSplitters = {};
        this.pendingFolders = {};
        this.source = source;
    }
    var _proto = SevenZipParser.prototype;
    /**
   * Convert BufferLike to Buffer (for small data like headers)
   */ _proto.toBuffer = function toBuffer(buf) {
        return Buffer.isBuffer(buf) ? buf : buf.toBuffer();
    };
    /**
   * Decode using codec - accepts BufferLike for LZMA1 support
   */ _proto.decodeWithCodec = function decodeWithCodec(codec, input, properties, unpackSize, callback) {
        var done = (0, _calloncefn.default)(callback);
        try {
            codec.decode(input, properties, unpackSize, function(err, result) {
                if (err) return done(err);
                if (!result) return done((0, _constantsts.createCodedError)('Decoder returned no data', _constantsts.ErrorCode.DECOMPRESSION_FAILED));
                done(null, result);
            });
        } catch (err) {
            done(err);
        }
    };
    /**
   * Parse the archive structure
   * Must be called before iterating entries
   */ _proto.parse = function parse(callback) {
        var _this = this;
        if (this.parsed) {
            if (typeof callback === 'function') return callback(null);
            return Promise.resolve();
        }
        if (typeof callback === 'function') return this.parseInternal(callback);
        return new Promise(function(resolve, reject) {
            return _this.parseInternal(function(err) {
                return err ? reject(err) : resolve();
            });
        });
    };
    _proto.parseInternal = function parseInternal(callback) {
        var _this = this;
        if (this.parsed) {
            callback(null);
            return;
        }
        var signature;
        var headerBuf;
        try {
            var sigBuf = this.source.read(0, _constantsts.SIGNATURE_HEADER_SIZE);
            if (sigBuf.length < _constantsts.SIGNATURE_HEADER_SIZE) {
                callback((0, _constantsts.createCodedError)('Archive too small', _constantsts.ErrorCode.TRUNCATED_ARCHIVE));
                return;
            }
            signature = (0, _headersts.parseSignatureHeader)(this.toBuffer(sigBuf));
            this.signature = signature;
            var headerOffset = _constantsts.SIGNATURE_HEADER_SIZE + signature.nextHeaderOffset;
            headerBuf = this.source.read(headerOffset, signature.nextHeaderSize);
            if (headerBuf.length < signature.nextHeaderSize) {
                callback((0, _constantsts.createCodedError)('Truncated header', _constantsts.ErrorCode.TRUNCATED_ARCHIVE));
                return;
            }
        } catch (err) {
            callback(err);
            return;
        }
        var finalize = function() {
            try {
                _this.buildEntries();
                _this.parsed = true;
                callback(null);
            } catch (err) {
                callback(err);
            }
        };
        try {
            var _ref;
            var _this_signature;
            var headerResult = (0, _headersts.parseEncodedHeader)(this.toBuffer(headerBuf), (_ref = (_this_signature = this.signature) === null || _this_signature === void 0 ? void 0 : _this_signature.nextHeaderCRC) !== null && _ref !== void 0 ? _ref : 0);
            this.streamsInfo = headerResult.streamsInfo || null;
            this.filesInfo = headerResult.filesInfo;
            finalize();
        } catch (err) {
            var codedErr = err;
            if (codedErr && codedErr.code === _constantsts.ErrorCode.COMPRESSED_HEADER) {
                this.handleCompressedHeader(headerBuf, function(headerErr) {
                    if (headerErr) {
                        callback(headerErr);
                        return;
                    }
                    finalize();
                });
            } else {
                callback(err);
            }
        }
    };
    /**
   * Handle compressed header (kEncodedHeader)
   */ _proto.handleCompressedHeader = function handleCompressedHeader(headerBuf, callback) {
        var _this = this;
        // Parse the encoded header info to get decompression parameters
        // Convert to Buffer for header parsing (small data)
        var headerBuffer = this.toBuffer(headerBuf);
        var offset = 1; // Skip kEncodedHeader byte
        var propertyId = headerBuffer[offset++];
        if (propertyId !== _constantsts.PropertyId.kMainStreamsInfo && propertyId !== _constantsts.PropertyId.kPackInfo) {
            callback((0, _constantsts.createCodedError)('Expected StreamsInfo in encoded header', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        var packInfoResult;
        try {
            packInfoResult = this.parseEncodedHeaderStreams(headerBuffer, 1);
        } catch (err) {
            callback(err);
            return;
        }
        var codec = (0, _indexts.getCodec)(packInfoResult.codecId);
        var candidates = [];
        var compressedStart = _constantsts.SIGNATURE_HEADER_SIZE + packInfoResult.packPos;
        candidates.push(this.source.read(compressedStart, packInfoResult.packSize));
        if (this.signature) {
            var packAreaEnd = _constantsts.SIGNATURE_HEADER_SIZE + this.signature.nextHeaderOffset;
            var searchStart = packAreaEnd - packInfoResult.packSize;
            var searchEnd = Math.max(_constantsts.SIGNATURE_HEADER_SIZE, compressedStart - 100000);
            var scanChunkSize = 4096;
            for(var chunkStart = searchStart; chunkStart >= searchEnd; chunkStart -= scanChunkSize){
                var chunkRaw = this.source.read(chunkStart, scanChunkSize + packInfoResult.packSize);
                // Convert to Buffer for scanning (small data)
                var chunk = this.toBuffer(chunkRaw);
                var limit = Math.min(chunk.length, scanChunkSize);
                for(var i = 0; i < limit; i++){
                    if (chunk[i] === 0x00) {
                        var end = i + packInfoResult.packSize;
                        if (end <= chunk.length) {
                            var candidateData = chunk.slice(i, end);
                            if (candidateData.length === packInfoResult.packSize) {
                                candidates.push(candidateData);
                            }
                        }
                    }
                }
            }
        }
        var tryCandidate = function(index) {
            if (index >= candidates.length) {
                callback((0, _constantsts.createCodedError)('Failed to decompress header - could not find valid LZMA data', _constantsts.ErrorCode.CORRUPT_HEADER));
                return;
            }
            _this.decodeWithCodec(codec, candidates[index], packInfoResult.properties, packInfoResult.unpackSize, function(err, decompressed) {
                if (err || !decompressed) {
                    tryCandidate(index + 1);
                    return;
                }
                if (packInfoResult.unpackCRC !== undefined) {
                    var actualCRC = (0, _extractbaseiterator.crc32)(decompressed);
                    if (actualCRC !== packInfoResult.unpackCRC) {
                        tryCandidate(index + 1);
                        return;
                    }
                }
                _this.parseDecompressedHeader(decompressed, callback);
            });
        };
        tryCandidate(0);
    };
    _proto.parseDecompressedHeader = function parseDecompressedHeader(decompressedHeader, callback) {
        var decompOffset = 0;
        var headerId = decompressedHeader[decompOffset++];
        if (headerId !== _constantsts.PropertyId.kHeader) {
            callback((0, _constantsts.createCodedError)('Expected kHeader in decompressed header', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        try {
            var result = (0, _headersts.parseHeaderContent)(decompressedHeader, decompOffset);
            this.streamsInfo = result.streamsInfo || null;
            this.filesInfo = result.filesInfo;
            callback(null);
        } catch (err) {
            callback(err);
        }
    };
    /**
   * Parse streams info from encoded header block
   * This is a simplified parser for the header's own compression info
   */ _proto.parseEncodedHeaderStreams = function parseEncodedHeaderStreams(buf, offset) {
        // This is a simplified parser for the encoded header's own streams info
        var packPos = 0;
        var packSize = 0;
        var unpackSize = 0;
        var codecId = [];
        var properties;
        var unpackCRC;
        while(offset < buf.length){
            var propertyId = buf[offset++];
            if (propertyId === _constantsts.PropertyId.kEnd) {
                break;
            }
            switch(propertyId){
                case _constantsts.PropertyId.kPackInfo:
                    {
                        var packPosResult = (0, _NumberCodects.readNumber)(buf, offset);
                        packPos = packPosResult.value;
                        offset += packPosResult.bytesRead;
                        var numPackResult = (0, _NumberCodects.readNumber)(buf, offset);
                        offset += numPackResult.bytesRead;
                        // Read until kEnd
                        while(buf[offset] !== _constantsts.PropertyId.kEnd){
                            if (buf[offset] === _constantsts.PropertyId.kSize) {
                                offset++;
                                var sizeResult = (0, _NumberCodects.readNumber)(buf, offset);
                                packSize = sizeResult.value;
                                offset += sizeResult.bytesRead;
                            } else {
                                offset++;
                            }
                        }
                        offset++; // Skip kEnd
                        break;
                    }
                case _constantsts.PropertyId.kUnpackInfo:
                    // Find folder/coder info
                    while(offset < buf.length && buf[offset] !== _constantsts.PropertyId.kEnd){
                        if (buf[offset] === _constantsts.PropertyId.kFolder) {
                            offset++;
                            var numFoldersResult = (0, _NumberCodects.readNumber)(buf, offset);
                            offset += numFoldersResult.bytesRead;
                            offset++; // external flag
                            // Parse coder
                            var numCodersResult = (0, _NumberCodects.readNumber)(buf, offset);
                            offset += numCodersResult.bytesRead;
                            var flags = buf[offset++];
                            var idSize = flags & 0x0f;
                            var hasAttributes = (flags & 0x20) !== 0;
                            codecId = [];
                            for(var i = 0; i < idSize; i++){
                                codecId.push(buf[offset++]);
                            }
                            if (hasAttributes) {
                                var propsLenResult = (0, _NumberCodects.readNumber)(buf, offset);
                                offset += propsLenResult.bytesRead;
                                properties = buf.slice(offset, offset + propsLenResult.value);
                                offset += propsLenResult.value;
                            }
                        } else if (buf[offset] === _constantsts.PropertyId.kCodersUnpackSize) {
                            offset++;
                            // Read unpack size - needed for LZMA decoder
                            var unpackSizeResult = (0, _NumberCodects.readNumber)(buf, offset);
                            unpackSize = unpackSizeResult.value;
                            offset += unpackSizeResult.bytesRead;
                        } else if (buf[offset] === _constantsts.PropertyId.kCRC) {
                            offset++;
                            var allDefined = buf[offset++];
                            if (allDefined) {
                                unpackCRC = buf.readUInt32LE(offset);
                                offset += 4;
                            }
                        } else {
                            offset++;
                        }
                    }
                    if (buf[offset] === _constantsts.PropertyId.kEnd) offset++;
                    break;
            }
        }
        return {
            packPos: packPos,
            packSize: packSize,
            unpackSize: unpackSize,
            codecId: codecId,
            properties: properties,
            unpackCRC: unpackCRC
        };
    };
    /**
   * Build the entries list from parsed file info
   */ _proto.buildEntries = function buildEntries() {
        this.entries = [];
        if (!this.streamsInfo) {
            // No streams info - just create entries from file info
            for(var i = 0; i < this.filesInfo.length; i++){
                var file = this.filesInfo[i];
                this.entries.push(this.createEntry(file, 0, 0, 0));
            }
            return;
        }
        // Use the properly parsed numUnpackStreamsPerFolder from the archive header
        var streamsPerFolder = this.streamsInfo.numUnpackStreamsPerFolder;
        // Initialize files per folder count (for smart caching)
        for(var f = 0; f < streamsPerFolder.length; f++){
            this.filesPerFolder[f] = streamsPerFolder[f];
            this.extractedPerFolder[f] = 0;
        }
        // Now build entries with proper folder/stream tracking
        var streamIndex = 0;
        var folderIndex = 0;
        var streamInFolder = 0;
        var folderStreamCount = streamsPerFolder[0] || 0;
        for(var j = 0; j < this.filesInfo.length; j++){
            var fileInfo = this.filesInfo[j];
            // Get size from unpackSizes for files with streams
            var size = 0;
            if (fileInfo.hasStream && streamIndex < this.streamsInfo.unpackSizes.length) {
                size = this.streamsInfo.unpackSizes[streamIndex];
            }
            var entry = this.createEntry(fileInfo, size, folderIndex, streamInFolder);
            entry._streamIndex = streamIndex;
            // Set CRC if available
            if (fileInfo.hasStream && this.streamsInfo.unpackCRCs && this.streamsInfo.unpackCRCs[streamIndex] !== undefined) {
                entry._crc = this.streamsInfo.unpackCRCs[streamIndex];
            }
            this.entries.push(entry);
            // Advance stream tracking for files with streams
            if (fileInfo.hasStream) {
                streamIndex++;
                streamInFolder++;
                // Check if we've exhausted streams in this folder
                if (streamInFolder >= folderStreamCount) {
                    folderIndex++;
                    streamInFolder = 0;
                    folderStreamCount = streamsPerFolder[folderIndex] || 0;
                }
            }
        }
        // Set _canStream for all entries now that we have complete folder info
        // This must be done after all entries are built because canStreamFolder
        // relies on the folder structure being fully parsed
        for(var i1 = 0; i1 < this.entries.length; i1++){
            var entry1 = this.entries[i1];
            if (entry1._hasStream && entry1._folderIndex >= 0) {
                entry1._canStream = this.canStreamFolder(entry1._folderIndex);
            }
        }
    };
    /**
   * Create an entry from file info
   */ _proto.createEntry = function createEntry(file, size, folderIndex, streamInFolder) {
        // Determine entry type
        // Note: 7z format doesn't natively support symlinks. p7zip with -snl stores
        // symlinks as regular files with the target path as content.
        var type = 'file';
        if (file.isDirectory) {
            type = 'directory';
        }
        // Calculate mode from Windows attributes
        var mode;
        if (file.attributes !== undefined) {
            // Check for Unix extension bit
            if ((file.attributes & _constantsts.FileAttribute.UNIX_EXTENSION) !== 0) {
                mode = file.attributes >>> 16 & 0xffff;
                // Check for symlink (S_IFLNK = 0xA000)
                // Note: Most 7z implementations don't preserve symlink mode bits
                if ((mode & 0xf000) === 0xa000) {
                    type = 'link';
                }
            } else if (file.isDirectory) {
                mode = 493; // 0o755
            } else {
                mode = 420; // 0o644
            }
        }
        return {
            name: getBaseName(file.name),
            path: file.name,
            type: type,
            size: size,
            mtime: file.mtime,
            atime: file.atime,
            ctime: file.ctime,
            mode: mode,
            isAntiFile: file.isAntiFile,
            _folderIndex: folderIndex,
            _streamIndex: 0,
            _streamIndexInFolder: streamInFolder,
            _hasStream: file.hasStream,
            _canStream: false
        };
    };
    /**
   * Get the list of entries
   */ _proto.getEntries = function getEntries() {
        if (!this.parsed) {
            throw new Error('SevenZipParser has not been parsed yet. Call parse(callback) before accessing entries.');
        }
        return this.entries;
    };
    /**
   * Get a readable stream for an entry's content.
   * Returns immediately - decompression happens when data is read (proper streaming).
   * Uses true streaming for codecs that support it, buffered for others.
   */ _proto.getEntryStream = function getEntryStream(entry) {
        if (!entry._hasStream || entry.type === 'directory') {
            // Return empty stream for directories and empty files
            var emptyStream = new _extractbaseiterator.PassThrough();
            emptyStream.end();
            return emptyStream;
        }
        if (!this.streamsInfo) {
            throw (0, _constantsts.createCodedError)('No streams info available', _constantsts.ErrorCode.CORRUPT_HEADER);
        }
        // Get folder info
        var folder = this.streamsInfo.folders[entry._folderIndex];
        if (!folder) {
            throw (0, _constantsts.createCodedError)('Invalid folder index', _constantsts.ErrorCode.CORRUPT_HEADER);
        }
        // Check codec support
        for(var i = 0; i < folder.coders.length; i++){
            var coder = folder.coders[i];
            if (!(0, _indexts.isCodecSupported)(coder.id)) {
                var codecName = (0, _indexts.getCodecName)(coder.id);
                throw (0, _constantsts.createCodedError)("Unsupported codec: ".concat(codecName), _constantsts.ErrorCode.UNSUPPORTED_CODEC);
            }
        }
        // Use true streaming for single-file folders that support it.
        // Multi-file folders use buffered approach because streaming requires
        // accessing files in order, which doesn't work with concurrent extraction.
        var filesInFolder = this.filesPerFolder[entry._folderIndex] || 1;
        if (entry._canStream && filesInFolder === 1) {
            return this._getEntryStreamStreaming(entry);
        }
        return this._getEntryStreamBuffered(entry);
    };
    /**
   * True streaming: data flows through without buffering entire folder.
   * Only used for single-file folders with streamable codecs (BZip2, Deflate, LZMA2).
   */ _proto._getEntryStreamStreaming = function _getEntryStreamStreaming(entry) {
        var _this = this;
        var started = false;
        var destroyed = false;
        var folderStream = null;
        var stream = new _extractbaseiterator.PassThrough();
        var originalRead = stream._read.bind(stream);
        stream._read = function(size) {
            if (!started && !destroyed) {
                started = true;
                (0, _deferts.defer)(function() {
                    if (destroyed) return;
                    try {
                        var crcValue = 0;
                        var verifyCrc = entry._crc !== undefined;
                        folderStream = _this.streamFolder(entry._folderIndex);
                        folderStream.output.on('data', function(chunk) {
                            if (destroyed) return;
                            if (verifyCrc) {
                                crcValue = (0, _extractbaseiterator.crc32)(chunk, crcValue);
                            }
                            if (!stream.write(chunk)) {
                                folderStream === null || folderStream === void 0 ? void 0 : folderStream.pause();
                                stream.once('drain', function() {
                                    return folderStream === null || folderStream === void 0 ? void 0 : folderStream.resume();
                                });
                            }
                        });
                        folderStream.output.on('end', function() {
                            if (destroyed) return;
                            if (verifyCrc && crcValue !== entry._crc) {
                                var _entry__crc;
                                stream.destroy((0, _constantsts.createCodedError)("CRC mismatch for ".concat(entry.path, ": expected ").concat((_entry__crc = entry._crc) === null || _entry__crc === void 0 ? void 0 : _entry__crc.toString(16), ", got ").concat(crcValue.toString(16)), _constantsts.ErrorCode.CRC_MISMATCH));
                                return;
                            }
                            stream.end();
                            _this.extractedPerFolder[entry._folderIndex] = (_this.extractedPerFolder[entry._folderIndex] || 0) + 1;
                        });
                        folderStream.output.on('error', function(err) {
                            if (!destroyed) stream.destroy(err);
                        });
                    } catch (err) {
                        if (!destroyed) {
                            stream.destroy(err);
                        }
                    }
                });
            }
            return originalRead(size);
        };
        // Override destroy to clean up folder stream
        // IMPORTANT: Emit error synchronously BEFORE calling original destroy.
        // On older Node, destroy() emits 'finish' and 'end' before 'error',
        // which causes piped streams to complete successfully before the error fires.
        var streamWithDestroy = stream;
        var originalDestroy = typeof streamWithDestroy.destroy === 'function' ? streamWithDestroy.destroy.bind(stream) : null;
        streamWithDestroy.destroy = function(err) {
            destroyed = true;
            if (err) stream.emit('error', err);
            if (folderStream) folderStream.destroy();
            if (originalDestroy) return originalDestroy();
            return stream;
        };
        return stream;
    };
    /**
   * Buffered extraction: decompress entire folder, slice out file.
   * Used for codecs that don't support incremental streaming (LZMA1, BCJ2).
   */ _proto._getEntryStreamBuffered = function _getEntryStreamBuffered(entry) {
        var _this = this;
        if (!this.streamsInfo) {
            throw (0, _constantsts.createCodedError)('No streams info available', _constantsts.ErrorCode.CORRUPT_HEADER);
        }
        var streamsInfo = this.streamsInfo;
        var folderIdx = entry._folderIndex;
        var started = false;
        var destroyed = false;
        var stream = new _extractbaseiterator.PassThrough();
        var originalRead = stream._read.bind(stream);
        stream._read = function(size) {
            if (!started && !destroyed) {
                started = true;
                (0, _deferts.defer)(function() {
                    if (destroyed) return;
                    _this.getDecompressedFolder(folderIdx, function(err, data) {
                        if (destroyed) return;
                        if (err || !data) {
                            stream.destroy(err || (0, _constantsts.createCodedError)('Unable to decompress folder', _constantsts.ErrorCode.DECOMPRESSION_FAILED));
                            return;
                        }
                        try {
                            var fileStart = 0;
                            for(var m = 0; m < entry._streamIndexInFolder; m++){
                                var prevStreamGlobalIndex = entry._streamIndex - entry._streamIndexInFolder + m;
                                fileStart += streamsInfo.unpackSizes[prevStreamGlobalIndex];
                            }
                            var fileSize = entry.size;
                            if (fileStart + fileSize > data.length) {
                                stream.destroy((0, _constantsts.createCodedError)("File data out of bounds: offset ".concat(fileStart, " + size ").concat(fileSize, " > decompressed length ").concat(data.length), _constantsts.ErrorCode.DECOMPRESSION_FAILED));
                                return;
                            }
                            var fileData = data.slice(fileStart, fileStart + fileSize);
                            if (entry._crc !== undefined) {
                                var actualCRC = (0, _extractbaseiterator.crc32)(fileData);
                                if (actualCRC !== entry._crc) {
                                    stream.destroy((0, _constantsts.createCodedError)("CRC mismatch for ".concat(entry.path, ": expected ").concat(entry._crc.toString(16), ", got ").concat(actualCRC.toString(16)), _constantsts.ErrorCode.CRC_MISMATCH));
                                    return;
                                }
                            }
                            _this.extractedPerFolder[folderIdx] = (_this.extractedPerFolder[folderIdx] || 0) + 1;
                            if (_this.extractedPerFolder[folderIdx] >= _this.filesPerFolder[folderIdx]) {
                                delete _this.decompressedCache[folderIdx];
                            }
                            if (!destroyed) {
                                stream.push(fileData);
                                stream.push(null);
                            }
                        } catch (decodeErr) {
                            stream.destroy(decodeErr);
                        }
                    });
                });
            }
            return originalRead(size);
        };
        // Override destroy to set destroyed flag
        // IMPORTANT: Emit error synchronously BEFORE calling original destroy.
        // On older Node, destroy() emits 'finish' and 'end' before 'error',
        // which causes piped streams to complete successfully before the error fires.
        var streamWithDestroy = stream;
        var originalDestroy = typeof streamWithDestroy.destroy === 'function' ? streamWithDestroy.destroy.bind(stream) : null;
        streamWithDestroy.destroy = function(err) {
            destroyed = true;
            if (err) stream.emit('error', err);
            if (originalDestroy) return originalDestroy();
            return stream;
        };
        return stream;
    };
    /**
   * Check if a folder uses BCJ2 codec
   */ _proto.folderHasBcj2 = function folderHasBcj2(folder) {
        for(var i = 0; i < folder.coders.length; i++){
            if ((0, _indexts.isBcj2Codec)(folder.coders[i].id)) {
                return true;
            }
        }
        return false;
    };
    /**
   * Get decompressed data for a folder, with smart caching for solid archives
   * Only caches when multiple files share a block, releases when last file extracted
   */ _proto.getDecompressedFolder = function getDecompressedFolder(folderIndex, callback) {
        var _this = this;
        if (this.decompressedCache[folderIndex]) {
            callback(null, this.decompressedCache[folderIndex]);
            return;
        }
        if (this.pendingFolders[folderIndex]) {
            this.pendingFolders[folderIndex].push(callback);
            return;
        }
        if (!this.streamsInfo) {
            callback((0, _constantsts.createCodedError)('No streams info available', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        this.pendingFolders[folderIndex] = [
            callback
        ];
        this.decodeFolderData(folderIndex, function(err, data) {
            var waiters = _this.pendingFolders[folderIndex] || [];
            delete _this.pendingFolders[folderIndex];
            if (err || !data) {
                for(var i = 0; i < waiters.length; i++){
                    waiters[i](err || (0, _constantsts.createCodedError)('Decoder returned no data', _constantsts.ErrorCode.DECOMPRESSION_FAILED));
                }
                return;
            }
            if (_this.shouldCacheFolder(folderIndex)) {
                _this.decompressedCache[folderIndex] = data;
            }
            for(var i1 = 0; i1 < waiters.length; i1++){
                waiters[i1](null, data);
            }
        });
    };
    _proto.shouldCacheFolder = function shouldCacheFolder(folderIndex) {
        var filesInFolder = this.filesPerFolder[folderIndex] || 1;
        var extractedFromFolder = this.extractedPerFolder[folderIndex] || 0;
        return filesInFolder - extractedFromFolder > 1;
    };
    _proto.decodeFolderData = function decodeFolderData(folderIndex, callback) {
        if (!this.streamsInfo) {
            callback((0, _constantsts.createCodedError)('No streams info available', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        var folder = this.streamsInfo.folders[folderIndex];
        if (!folder) {
            callback((0, _constantsts.createCodedError)('Invalid folder index', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        if (this.folderHasBcj2(folder)) {
            this.decompressBcj2Folder(folderIndex, callback);
            return;
        }
        var packDataResult = this.readPackedData(folderIndex);
        if (_instanceof(packDataResult, Error)) {
            callback(packDataResult);
            return;
        }
        this.decodeFolderCoders(folder, packDataResult, 0, callback);
    };
    _proto.readPackedData = function readPackedData(folderIndex) {
        if (!this.streamsInfo) {
            return (0, _constantsts.createCodedError)('No streams info available', _constantsts.ErrorCode.CORRUPT_HEADER);
        }
        var folder = this.streamsInfo.folders[folderIndex];
        if (!folder) {
            return (0, _constantsts.createCodedError)('Invalid folder index', _constantsts.ErrorCode.CORRUPT_HEADER);
        }
        var signedHeaderSize = _constantsts.SIGNATURE_HEADER_SIZE;
        var signedPackPos = this.streamsInfo.packPos;
        var packPos = Math.max(signedHeaderSize, 0) + Math.max(signedPackPos, 0);
        var packStreamIndex = 0;
        for(var j = 0; j < folderIndex; j++){
            packStreamIndex += this.streamsInfo.folders[j].packedStreams.length;
        }
        for(var k = 0; k < packStreamIndex; k++){
            var size = this.streamsInfo.packSizes[k];
            if (packPos + size < packPos) {
                return (0, _constantsts.createCodedError)("Pack position overflow at index ".concat(k), _constantsts.ErrorCode.CORRUPT_ARCHIVE);
            }
            packPos += size;
        }
        var packSize = this.streamsInfo.packSizes[packStreamIndex];
        if (packSize < 0 || packSize > Number.MAX_SAFE_INTEGER) {
            return (0, _constantsts.createCodedError)("Invalid pack size: ".concat(packSize), _constantsts.ErrorCode.CORRUPT_ARCHIVE);
        }
        if (packPos < 0 || packPos > Number.MAX_SAFE_INTEGER) {
            return (0, _constantsts.createCodedError)("Invalid pack position: ".concat(packPos), _constantsts.ErrorCode.CORRUPT_ARCHIVE);
        }
        return this.source.read(packPos, packSize);
    };
    _proto.decodeFolderCoders = function decodeFolderCoders(folder, input, index, callback) {
        var _this = this;
        if (index >= folder.coders.length) {
            // Convert BufferList to Buffer for final output
            callback(null, this.toBuffer(input));
            return;
        }
        var coderInfo = folder.coders[index];
        var codec = (0, _indexts.getCodec)(coderInfo.id);
        var unpackSize = folder.unpackSizes[index];
        if (unpackSize < 0 || unpackSize > Number.MAX_SAFE_INTEGER) {
            callback((0, _constantsts.createCodedError)("Invalid unpack size: ".concat(unpackSize), _constantsts.ErrorCode.CORRUPT_ARCHIVE));
            return;
        }
        this.decodeWithCodec(codec, input, coderInfo.properties, unpackSize, function(err, output) {
            if (err || !output) {
                callback(err || (0, _constantsts.createCodedError)('Decoder returned no data', _constantsts.ErrorCode.DECOMPRESSION_FAILED));
                return;
            }
            _this.decodeFolderCoders(folder, output, index + 1, callback);
        });
    };
    /**
   * Decompress a BCJ2 folder with multi-stream handling
   * BCJ2 uses 4 input streams: main, call, jump, range coder
   */ _proto.decompressBcj2Folder = function decompressBcj2Folder(folderIndex, callback) {
        var _this = this;
        if (!this.streamsInfo) {
            callback((0, _constantsts.createCodedError)('No streams info available', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        var folder = this.streamsInfo.folders[folderIndex];
        if (!folder) {
            callback((0, _constantsts.createCodedError)('Invalid folder index', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        var packPos = _constantsts.SIGNATURE_HEADER_SIZE + this.streamsInfo.packPos;
        var packStreamIndex = 0;
        for(var j = 0; j < folderIndex; j++){
            packStreamIndex += this.streamsInfo.folders[j].packedStreams.length;
        }
        for(var k = 0; k < packStreamIndex; k++){
            packPos += this.streamsInfo.packSizes[k];
        }
        var numPackStreams = folder.packedStreams.length;
        var packStreams = [];
        var currentPos = packPos;
        for(var p = 0; p < numPackStreams; p++){
            var size = this.streamsInfo.packSizes[packStreamIndex + p];
            packStreams.push(this.source.read(currentPos, size));
            currentPos += size;
        }
        var coderOutputs = {};
        var bcj2CoderIndex = -1;
        for(var c = 0; c < folder.coders.length; c++){
            if ((0, _indexts.isBcj2Codec)(folder.coders[c].id)) {
                bcj2CoderIndex = c;
                break;
            }
        }
        if (bcj2CoderIndex === -1) {
            callback((0, _constantsts.createCodedError)('BCJ2 coder not found in folder', _constantsts.ErrorCode.CORRUPT_HEADER));
            return;
        }
        var inputToPackStream = {};
        for(var pi = 0; pi < folder.packedStreams.length; pi++){
            inputToPackStream[folder.packedStreams[pi]] = pi;
        }
        var processOrder = this.getCoderProcessOrder(folder, bcj2CoderIndex);
        var processNext = function(orderIndex) {
            if (orderIndex >= processOrder.length) {
                _this.finishBcj2Decode(folder, bcj2CoderIndex, coderOutputs, inputToPackStream, packStreams, callback);
                return;
            }
            var coderIdx = processOrder[orderIndex];
            if (coderIdx === bcj2CoderIndex) {
                processNext(orderIndex + 1);
                return;
            }
            var coder = folder.coders[coderIdx];
            var codec = (0, _indexts.getCodec)(coder.id);
            var coderInputStart = 0;
            for(var ci2 = 0; ci2 < coderIdx; ci2++){
                coderInputStart += folder.coders[ci2].numInStreams;
            }
            var inputIdx = coderInputStart;
            var packStreamIdx = inputToPackStream[inputIdx];
            var inputData = packStreams[packStreamIdx];
            var unpackSize = folder.unpackSizes[coderIdx];
            _this.decodeWithCodec(codec, inputData, coder.properties, unpackSize, function(err, outputData) {
                if (err || !outputData) {
                    callback(err || (0, _constantsts.createCodedError)('Decoder returned no data', _constantsts.ErrorCode.DECOMPRESSION_FAILED));
                    return;
                }
                var coderOutputStart = 0;
                for(var co2 = 0; co2 < coderIdx; co2++){
                    coderOutputStart += folder.coders[co2].numOutStreams;
                }
                coderOutputs[coderOutputStart] = outputData;
                processNext(orderIndex + 1);
            });
        };
        processNext(0);
    };
    _proto.finishBcj2Decode = function finishBcj2Decode(folder, bcj2CoderIndex, coderOutputs, inputToPackStream, packStreams, callback) {
        var bcj2InputStart = 0;
        for(var ci3 = 0; ci3 < bcj2CoderIndex; ci3++){
            bcj2InputStart += folder.coders[ci3].numInStreams;
        }
        var bcj2Inputs = [];
        for(var bi = 0; bi < 4; bi++){
            var globalIdx = bcj2InputStart + bi;
            var boundOutput = -1;
            for(var bp2 = 0; bp2 < folder.bindPairs.length; bp2++){
                if (folder.bindPairs[bp2].inIndex === globalIdx) {
                    boundOutput = folder.bindPairs[bp2].outIndex;
                    break;
                }
            }
            if (boundOutput >= 0) {
                bcj2Inputs.push(coderOutputs[boundOutput]);
            } else {
                var psIdx = inputToPackStream[globalIdx];
                // Convert BufferList to Buffer for BCJ2 (small data, safe to concatenate)
                bcj2Inputs.push(this.toBuffer(packStreams[psIdx]));
            }
        }
        var bcj2OutputStart = 0;
        for(var co3 = 0; co3 < bcj2CoderIndex; co3++){
            bcj2OutputStart += folder.coders[co3].numOutStreams;
        }
        var bcj2UnpackSize = folder.unpackSizes[bcj2OutputStart];
        try {
            var result = (0, _indexts.decodeBcj2Multi)(bcj2Inputs, undefined, bcj2UnpackSize);
            callback(null, result);
        } catch (err) {
            callback(err);
        } finally{
            for(var key in coderOutputs){
                delete coderOutputs[key];
            }
            packStreams.length = 0;
        }
    };
    /**
   * Get processing order for coders (dependency order)
   */ _proto.getCoderProcessOrder = function getCoderProcessOrder(folder, excludeIdx) {
        var order = [];
        var processed = {};
        // Simple approach: process coders that don't depend on unprocessed outputs
        var changed = true;
        while(changed){
            changed = false;
            for(var c = 0; c < folder.coders.length; c++){
                if (processed[c] || c === excludeIdx) continue;
                // Check if all inputs are satisfied
                var inputStart = 0;
                for(var i = 0; i < c; i++){
                    inputStart += folder.coders[i].numInStreams;
                }
                var canProcess = true;
                for(var inp = 0; inp < folder.coders[c].numInStreams; inp++){
                    var globalIdx = inputStart + inp;
                    // Check if bound to an unprocessed coder
                    for(var bp = 0; bp < folder.bindPairs.length; bp++){
                        if (folder.bindPairs[bp].inIndex === globalIdx) {
                            // Find which coder produces this output
                            var outIdx = folder.bindPairs[bp].outIndex;
                            var outStart = 0;
                            for(var oc = 0; oc < folder.coders.length; oc++){
                                var numOut = folder.coders[oc].numOutStreams;
                                if (outIdx < outStart + numOut) {
                                    if (!processed[oc] && oc !== excludeIdx) {
                                        canProcess = false;
                                    }
                                    break;
                                }
                                outStart += numOut;
                            }
                        }
                    }
                }
                if (canProcess) {
                    order.push(c);
                    processed[c] = true;
                    changed = true;
                }
            }
        }
        return order;
    };
    /**
   * Close the parser and release resources
   */ _proto.close = function close() {
        if (this.source) {
            this.source.close();
        }
    };
    // ============================================================
    // STREAMING METHODS (Phase 1+)
    // ============================================================
    /**
   * Check if a codec supports true streaming decompression.
   *
   * Only codecs that process data incrementally (not buffering entire input) qualify.
   * @param codecId - The codec ID as an array of bytes
   * @returns true if the codec can stream
   */ _proto.codecSupportsStreaming = function codecSupportsStreaming(codecId) {
        // Convert to string key for comparison
        var key = codecId.map(function(b) {
            return b.toString(16).toUpperCase();
        }).join('-');
        // BZip2 - unbzip2-stream processes blocks incrementally
        if (key === '4-2-2') return true;
        // Copy/Store - PassThrough, obviously streams
        if (key === '0') return true;
        // Deflate - now uses zlib.createInflateRaw() which streams
        if (key === '4-1-8') return true;
        // Delta - now uses streaming Transform (Phase 2.5)
        if (key === '3') return true;
        // BCJ x86 - now uses streaming Transform (Phase 3.5)
        if (key === '3-3-1-3') return true;
        // BCJ ARM - now uses streaming Transform (Phase 3.5)
        if (key === '3-3-1-5') return true;
        // LZMA2 - now uses streaming Transform (Phase 5)
        if (key === '21') return true;
        // LZMA - still buffer-based (cannot stream - no chunk boundaries in LZMA1 format)
        // Other BCJ variants (ARM64, ARMT, IA64, PPC, SPARC) - still buffer-based
        // BCJ2 - multi-stream architecture, never streamable
        return false;
    };
    /**
   * Check if a folder can be streamed (vs buffered).
   *
   * Streaming is possible when ALL codecs in the chain support streaming.
   * BCJ2 folders are never streamable due to their 4-stream architecture.
   *
   * @param folderIndex - Index of the folder to check
   * @returns true if the folder can be streamed
   */ _proto.canStreamFolder = function canStreamFolder(folderIndex) {
        if (!this.streamsInfo) return false;
        var folder = this.streamsInfo.folders[folderIndex];
        if (!folder) return false;
        // BCJ2 requires special multi-stream handling - not streamable
        if (this.folderHasBcj2(folder)) {
            return false;
        }
        // Check if ALL codecs in chain support streaming
        for(var i = 0; i < folder.coders.length; i++){
            if (!this.codecSupportsStreaming(folder.coders[i].id)) {
                return false;
            }
        }
        return true;
    };
    /**
   * Stream a folder's decompression.
   *
   * Creates a pipeline: packed data → codec decoders → output stream
   *
   * @param folderIndex - Index of folder to decompress
   * @returns Object with output stream and control methods
   */ _proto.streamFolder = function streamFolder(folderIndex) {
        if (!this.streamsInfo) {
            throw (0, _constantsts.createCodedError)('No streams info available', _constantsts.ErrorCode.CORRUPT_HEADER);
        }
        if (!this.canStreamFolder(folderIndex)) {
            throw (0, _constantsts.createCodedError)('Folder does not support streaming', _constantsts.ErrorCode.UNSUPPORTED_CODEC);
        }
        var folder = this.streamsInfo.folders[folderIndex];
        // Calculate packed data position
        var packPos = _constantsts.SIGNATURE_HEADER_SIZE + this.streamsInfo.packPos;
        // Find which pack stream this folder uses
        var packStreamIndex = 0;
        for(var j = 0; j < folderIndex; j++){
            packStreamIndex += this.streamsInfo.folders[j].packedStreams.length;
        }
        // Calculate position of this pack stream
        for(var k = 0; k < packStreamIndex; k++){
            packPos += this.streamsInfo.packSizes[k];
        }
        var packSize = this.streamsInfo.packSizes[packStreamIndex];
        // Create readable stream from packed data
        var packedStream = this.source.createReadStream(packPos, packSize);
        // Build codec pipeline
        var stream = packedStream;
        var decoders = [];
        for(var i = 0; i < folder.coders.length; i++){
            var coderInfo = folder.coders[i];
            var codec = (0, _indexts.getCodec)(coderInfo.id);
            var unpackSize = folder.unpackSizes[i];
            var decoder = codec.createDecoder(coderInfo.properties, unpackSize);
            decoders.push(decoder);
            stream = stream.pipe(decoder);
        }
        return {
            output: stream,
            pause: function() {
                return packedStream.pause();
            },
            resume: function() {
                return packedStream.resume();
            },
            destroy: function(err) {
                // Check for destroy method existence (not available in Node 4 and earlier)
                var ps = packedStream;
                if (typeof ps.destroy === 'function') ps.destroy(err);
                for(var i = 0; i < decoders.length; i++){
                    var d = decoders[i];
                    if (typeof d.destroy === 'function') d.destroy(err);
                }
            }
        };
    };
    /**
   * Get a streaming entry stream (Promise-based API).
   *
   * For streamable folders: Returns a true streaming decompression
   * For non-streamable folders: Falls back to buffered extraction
   *
   * @param entry - The entry to get stream for
   * @returns Promise resolving to readable stream
   */ _proto.getEntryStreamStreaming = function getEntryStreamStreaming(entry) {
        return _async_to_generator(function() {
            var emptyStream, folderIndex, filesInFolder;
            return _ts_generator(this, function(_state) {
                if (!entry._hasStream || entry.type === 'directory') {
                    emptyStream = new _extractbaseiterator.PassThrough();
                    emptyStream.end();
                    return [
                        2,
                        emptyStream
                    ];
                }
                folderIndex = entry._folderIndex;
                // Fall back to buffered if not streamable
                if (!this.canStreamFolder(folderIndex)) {
                    return [
                        2,
                        this.getEntryStream(entry)
                    ];
                }
                filesInFolder = this.filesPerFolder[folderIndex] || 1;
                if (filesInFolder === 1) {
                    // Single file - direct streaming
                    return [
                        2,
                        this.getEntryStreamDirect(entry)
                    ];
                }
                // Multi-file folders use FolderStreamSplitter (Phase 2)
                return [
                    2,
                    this.getEntryStreamFromSplitter(entry)
                ];
            });
        }).call(this);
    };
    /**
   * Direct streaming for single-file folders.
   * Pipes folder decompression directly to output with CRC verification.
   */ _proto.getEntryStreamDirect = function getEntryStreamDirect(entry) {
        var _this = this;
        return new Promise(function(resolve, reject) {
            var outputStream = new _extractbaseiterator.PassThrough();
            var crcValue = 0;
            var verifyCrc = entry._crc !== undefined;
            try {
                var folderStream = _this.streamFolder(entry._folderIndex);
                folderStream.output.on('data', function(chunk) {
                    if (verifyCrc) {
                        crcValue = (0, _extractbaseiterator.crc32)(chunk, crcValue);
                    }
                    // Handle backpressure
                    if (!outputStream.write(chunk)) {
                        folderStream.pause();
                        outputStream.once('drain', function() {
                            return folderStream.resume();
                        });
                    }
                });
                folderStream.output.on('end', function() {
                    // Verify CRC
                    if (verifyCrc && crcValue !== entry._crc) {
                        var _entry__crc;
                        var _$err = (0, _constantsts.createCodedError)("CRC mismatch for ".concat(entry.path, ": expected ").concat((_entry__crc = entry._crc) === null || _entry__crc === void 0 ? void 0 : _entry__crc.toString(16), ", got ").concat(crcValue.toString(16)), _constantsts.ErrorCode.CRC_MISMATCH);
                        outputStream.destroy(_$err);
                        return;
                    }
                    outputStream.end();
                    // Track extraction
                    _this.extractedPerFolder[entry._folderIndex] = (_this.extractedPerFolder[entry._folderIndex] || 0) + 1;
                });
                folderStream.output.on('error', function(err) {
                    outputStream.destroy(err);
                });
                resolve(outputStream);
            } catch (err) {
                reject(err);
            }
        });
    };
    /**
   * Get stream from folder splitter (for multi-file folders).
   * Creates splitter on first access, reuses for subsequent files in same folder.
   */ _proto.getEntryStreamFromSplitter = function getEntryStreamFromSplitter(entry) {
        var _this = this;
        return new Promise(function(resolve, reject) {
            var folderIndex = entry._folderIndex;
            // Get or create splitter for this folder
            var splitter = _this.folderSplitters[folderIndex];
            if (!splitter) {
                // Create new splitter with file sizes and CRCs
                var folderInfo = _this.getFolderFileInfo(folderIndex);
                splitter = new _FolderStreamSplitterts.FolderStreamSplitter({
                    fileSizes: folderInfo.fileSizes,
                    verifyCrc: true,
                    expectedCrcs: folderInfo.expectedCrcs
                });
                _this.folderSplitters[folderIndex] = splitter;
                // Start streaming the folder
                var folderStream;
                try {
                    folderStream = _this.streamFolder(folderIndex);
                } catch (err) {
                    delete _this.folderSplitters[folderIndex];
                    reject(err);
                    return;
                }
                folderStream.output.on('data', function(chunk) {
                    // Handle backpressure from splitter
                    if (!(splitter === null || splitter === void 0 ? void 0 : splitter.write(chunk))) {
                        folderStream.pause();
                        splitter === null || splitter === void 0 ? void 0 : splitter.onDrain(function() {
                            folderStream.resume();
                        });
                    }
                });
                folderStream.output.on('end', function() {
                    splitter === null || splitter === void 0 ? void 0 : splitter.end();
                    delete _this.folderSplitters[folderIndex];
                });
                folderStream.output.on('error', function(_err) {
                    splitter === null || splitter === void 0 ? void 0 : splitter.end();
                    delete _this.folderSplitters[folderIndex];
                });
            }
            // Get this entry's stream from splitter
            try {
                var fileStream = splitter.getFileStream(entry._streamIndexInFolder);
                // Track extraction when stream ends
                fileStream.on('end', function() {
                    _this.extractedPerFolder[folderIndex] = (_this.extractedPerFolder[folderIndex] || 0) + 1;
                });
                resolve(fileStream);
            } catch (err) {
                reject(err);
            }
        });
    };
    /**
   * Get file sizes and CRCs for all files in a folder (in stream order).
   * Used by FolderStreamSplitter to know file boundaries.
   */ _proto.getFolderFileInfo = function getFolderFileInfo(folderIndex) {
        var fileSizes = [];
        var expectedCrcs = [];
        // Collect entries in this folder, sorted by stream index
        var folderEntries = [];
        for(var i = 0; i < this.entries.length; i++){
            var e = this.entries[i];
            if (e._folderIndex === folderIndex && e._hasStream) {
                folderEntries.push(e);
            }
        }
        // Sort by stream index within folder
        folderEntries.sort(function(a, b) {
            return a._streamIndexInFolder - b._streamIndexInFolder;
        });
        for(var i1 = 0; i1 < folderEntries.length; i1++){
            var entry = folderEntries[i1];
            fileSizes.push(entry.size);
            expectedCrcs.push(entry._crc);
        }
        return {
            fileSizes: fileSizes,
            expectedCrcs: expectedCrcs
        };
    };
    return SevenZipParser;
}();
/**
 * Get base name from a path
 */ function getBaseName(path) {
    var lastSlash = path.lastIndexOf('/');
    var lastBackslash = path.lastIndexOf('\\');
    var lastSep = Math.max(lastSlash, lastBackslash);
    return lastSep >= 0 ? path.slice(lastSep + 1) : path;
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
