// function-exec-sync/dist/esm/workers/runFunction.cjs
const compat = require('async-compat');
const fs = require('fs');
const { loadModule } = require('module-compat');
const path = require('path');
const serialize = require(path.join(__dirname, '..', 'serialize-javascript.js'));
const input = process.argv[2];
const output = process.argv[3];
function writeResult(result) {
    fs.writeFile(output, serialize(result), 'utf8', ()=>{
        process.exit(0);
    });
}
function writeError(error) {
    const errorObj = {
        message: error.message,
        stack: error.stack
    };
    for(const key in error)errorObj[key] = error[key];
    writeResult({
        error: errorObj
    });
}
// get data
try {
    // biome-ignore lint/security/noGlobalEval: Deserialize worker data
    const workerData = eval(`(${fs.readFileSync(input, 'utf8')})`);
    // set up env
    if (process.cwd() !== workerData.cwd) process.chdir(workerData.cwd);
    for(const key in workerData.env)process.env[key] = workerData.env[key];
    // load and call function
    const loadOptions = {
        moduleType: workerData.moduleType || 'auto',
        interop: workerData.interop || 'default'
    };
    loadModule(workerData.filePath, loadOptions, (err, fn)=>{
        if (err) {
            writeError(err);
            return;
        }
        if (typeof fn !== 'function') {
            writeResult({
                value: fn
            });
        } else {
            const args = [
                fn,
                workerData.callbacks
            ].concat(workerData.args);
            args.push((err, value)=>{
                err ? writeError(err) : writeResult({
                    value
                });
            });
            compat.asyncFunction.apply(null, args);
        }
    });
} catch (err) {
    writeError(err);
}
// CJS worker - uses CJS build of module-compat
// Only supports loading CJS modules (or ESM on Node 23+ via require)
export { };
