// fs-remove-compat/dist/cjs/fallback/rm.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, /**
 * Fallback rm implementation for Node < 14.14.
 * Matches Node.js fs.rm API.
 */ "default", {
    enumerable: true,
    get: function() {
        return fallbackRm;
    }
});
var _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
var _fixWinEPERMts = require("./fixWinEPERM.js");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var RETRYABLE_CODES = [
    'EBUSY',
    'EMFILE',
    'ENFILE',
    'ENOTEMPTY',
    'EPERM'
];
/**
 * Remove a file asynchronously with retry logic.
 */ function unlinkWithRetry(path, options, attempt, callback) {
    _fs.default.unlink(path, function(err) {
        if (!err) return callback();
        // Handle ENOENT
        if (err.code === 'ENOENT') {
            if (options.force) return callback();
            return callback(err);
        }
        // Try EPERM fix on Windows
        if ((0, _fixWinEPERMts.shouldFixEPERM)(err)) {
            return (0, _fixWinEPERMts.fixWinEPERM)(path, err, function(fixErr) {
                if (!fixErr) return callback();
                // Fall through to retry logic
                retryOrFail(path, options, attempt, err, callback, unlinkWithRetry);
            });
        }
        retryOrFail(path, options, attempt, err, callback, unlinkWithRetry);
    });
}
/**
 * Remove a directory asynchronously with retry logic.
 */ function rmdirWithRetry(path, options, attempt, callback) {
    _fs.default.rmdir(path, function(err) {
        if (!err) return callback();
        if (err.code === 'ENOENT') {
            if (options.force) return callback();
            return callback(err);
        }
        retryOrFail(path, options, attempt, err, callback, rmdirWithRetry);
    });
}
/**
 * Retry logic helper.
 */ function retryOrFail(path, options, attempt, err, callback, retryFn) {
    if (RETRYABLE_CODES.indexOf(err.code || '') === -1 || attempt >= options.maxRetries) {
        callback(err);
        return;
    }
    setTimeout(function() {
        retryFn(path, options, attempt + 1, callback);
    }, options.retryDelay);
}
/**
 * Remove directory contents recursively.
 */ function rmdirRecursive(path, options, callback) {
    _fs.default.readdir(path, function(readErr, entries) {
        var _loop = function(i) {
            var entry = entries[i];
            var entryPath = "".concat(path, "/").concat(entry);
            _fs.default.lstat(entryPath, function(statErr, stats) {
                if (hasError) return;
                if (statErr) {
                    if (statErr.code === 'ENOENT' && options.force) {
                        pending--;
                        if (pending === 0) rmdirWithRetry(path, options, 0, callback);
                        return;
                    }
                    hasError = true;
                    return callback(statErr);
                }
                var onDone = function(err) {
                    if (hasError) return;
                    if (err) {
                        hasError = true;
                        return callback(err);
                    }
                    pending--;
                    if (pending === 0) {
                        rmdirWithRetry(path, options, 0, callback);
                    }
                };
                if (stats.isDirectory()) {
                    rmdirRecursive(entryPath, options, onDone);
                } else {
                    unlinkWithRetry(entryPath, options, 0, onDone);
                }
            });
        };
        if (readErr) {
            if (readErr.code === 'ENOENT' && options.force) return callback();
            return callback(readErr);
        }
        var pending = entries.length;
        if (pending === 0) {
            return rmdirWithRetry(path, options, 0, callback);
        }
        var hasError = false;
        for(var i = 0; i < entries.length; i++)_loop(i);
    });
}
function fallbackRm(path, options, callback) {
    var _ref, _ref1, _ref2, _ref3;
    var opts = {
        recursive: (_ref = options === null || options === void 0 ? void 0 : options.recursive) !== null && _ref !== void 0 ? _ref : false,
        force: (_ref1 = options === null || options === void 0 ? void 0 : options.force) !== null && _ref1 !== void 0 ? _ref1 : false,
        maxRetries: (_ref2 = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref2 !== void 0 ? _ref2 : 0,
        retryDelay: (_ref3 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref3 !== void 0 ? _ref3 : 100
    };
    _fs.default.lstat(path, function(statErr, stats) {
        if (statErr) {
            if (statErr.code === 'ENOENT') {
                if (opts.force) return callback();
                return callback(statErr);
            }
            return callback(statErr);
        }
        if (stats.isDirectory()) {
            if (!opts.recursive) {
                var err = new Error("EISDIR: illegal operation on a directory, rm '".concat(path, "'"));
                err.code = 'EISDIR';
                err.syscall = 'rm';
                err.path = path;
                return callback(err);
            }
            rmdirRecursive(path, opts, callback);
        } else {
            unlinkWithRetry(path, opts, 0, callback);
        }
    });
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
