// fs-remove-compat/dist/cjs/fallback/rmSync.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, /**
 * Fallback rmSync implementation for Node < 14.14.
 * Matches Node.js fs.rmSync API.
 */ "default", {
    enumerable: true,
    get: function() {
        return fallbackRmSync;
    }
});
var _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
var _fixWinEPERMts = require("./fixWinEPERM.js");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var RETRYABLE_CODES = [
    'EBUSY',
    'EMFILE',
    'ENFILE',
    'ENOTEMPTY',
    'EPERM'
];
/**
 * Busy-wait for sync retry delay.
 */ function busyWait(ms) {
    var end = Date.now() + ms;
    while(Date.now() < end){
    // Busy wait
    }
}
/**
 * Remove a file synchronously.
 */ function unlinkSync(path, options) {
    for(var attempt = 0; attempt <= options.maxRetries; attempt++){
        try {
            _fs.default.unlinkSync(path);
            return;
        } catch (err) {
            var error = err;
            // Handle ENOENT
            if (error.code === 'ENOENT') {
                if (options.force) return;
                throw error;
            }
            // Try EPERM fix on Windows
            if ((0, _fixWinEPERMts.shouldFixEPERM)(error)) {
                try {
                    (0, _fixWinEPERMts.fixWinEPERMSync)(path, error);
                    return;
                } catch (_fixErr) {
                // Fall through to retry logic
                }
            }
            // Check if retryable
            if (RETRYABLE_CODES.indexOf(error.code || '') === -1 || attempt >= options.maxRetries) {
                throw error;
            }
            // Wait before retry (linear backoff like Node.js)
            busyWait(options.retryDelay);
        }
    }
}
/**
 * Remove a directory synchronously (non-recursive).
 */ function rmdirSync(path, options) {
    for(var attempt = 0; attempt <= options.maxRetries; attempt++){
        try {
            _fs.default.rmdirSync(path);
            return;
        } catch (err) {
            var error = err;
            if (error.code === 'ENOENT') {
                if (options.force) return;
                throw error;
            }
            if (RETRYABLE_CODES.indexOf(error.code || '') === -1 || attempt >= options.maxRetries) {
                throw error;
            }
            busyWait(options.retryDelay);
        }
    }
}
/**
 * Remove directory contents recursively.
 */ function rmdirRecursiveSync(path, options) {
    var entries;
    try {
        entries = _fs.default.readdirSync(path);
    } catch (err) {
        var error = err;
        if (error.code === 'ENOENT' && options.force) return;
        throw error;
    }
    for(var i = 0; i < entries.length; i++){
        var entry = entries[i];
        var entryPath = "".concat(path, "/").concat(entry);
        var stats = void 0;
        try {
            stats = _fs.default.lstatSync(entryPath);
        } catch (err) {
            var error1 = err;
            if (error1.code === 'ENOENT' && options.force) continue;
            throw error1;
        }
        if (stats.isDirectory()) {
            rmdirRecursiveSync(entryPath, options);
        } else {
            unlinkSync(entryPath, options);
        }
    }
    // Now remove the directory itself
    rmdirSync(path, options);
}
function fallbackRmSync(path, options) {
    var _ref, _ref1, _ref2, _ref3;
    var opts = {
        recursive: (_ref = options === null || options === void 0 ? void 0 : options.recursive) !== null && _ref !== void 0 ? _ref : false,
        force: (_ref1 = options === null || options === void 0 ? void 0 : options.force) !== null && _ref1 !== void 0 ? _ref1 : false,
        maxRetries: (_ref2 = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref2 !== void 0 ? _ref2 : 0,
        retryDelay: (_ref3 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref3 !== void 0 ? _ref3 : 100
    };
    var stats;
    try {
        stats = _fs.default.lstatSync(path);
    } catch (err) {
        var error = err;
        if (error.code === 'ENOENT') {
            if (opts.force) return;
            throw error;
        }
        throw error;
    }
    if (stats.isDirectory()) {
        if (!opts.recursive) {
            var _$err = new Error("EISDIR: illegal operation on a directory, rm '".concat(path, "'"));
            _$err.code = 'EISDIR';
            _$err.syscall = 'rm';
            _$err.path = path;
            throw _$err;
        }
        rmdirRecursiveSync(path, opts);
    } else {
        unlinkSync(path, opts);
    }
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
