// @zen-browser/surfer/dist/utils/dispatch.js
'use strict'
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.dispatch = exports.configDispatch = exports.removeTimestamp = void 0
const execa_1 = __importDefault(require('execa'))
const constants_1 = require('../constants')
const log_1 = require('../log')
const removeTimestamp = (input) => input.replace(/\s\d{1,5}:\d\d\.\d\d /g, '')
exports.removeTimestamp = removeTimestamp
const configDispatch = (cmd, config) => {
  // Provide a default logger if none was specified by the user
  const logger = config?.logger || ((data) => log_1.log.info(data))
  // Decide what shell we should be using. False will use the system default
  let shell = false
  if (config?.shell) {
    switch (config.shell) {
      // Don't change anything if we are using the default shell
      case 'default': {
        break
      }
      case 'unix': {
        // Bash path provides a unix shell on windows
        shell = constants_1.BASH_PATH || false
        break
      }
      default: {
        log_1.log.error(`dispatch() does not understand the shell '${shell}'`)
        break
      }
    }
  }
  const output = []
  const handle = (data, killOnError) => {
    const dataAsString = data.toString()
    for (const line of dataAsString.split('\n')) {
      if (line.length > 0) {
        output.push(line)
        logger((0, exports.removeTimestamp)(line))
      }
    }
    if (killOnError) {
      log_1.log.error('Command failed. See error above.')
    }
  }
  return new Promise((resolve) => {
    const proc = (0, execa_1.default)(cmd, config?.args, {
      cwd: config?.cwd || process.cwd(),
      shell: shell,
      env: {
        ...config?.env,
        ...process.env,
      },
    })
    proc.stdout?.on('data', (d) => handle(d))
    proc.stderr?.on('data', (d) => handle(d))
    proc.stdout?.on('error', (d) => handle(d, config?.killOnError || false))
    proc.stderr?.on('error', (d) => handle(d, config?.killOnError || false))
    proc.on('exit', (code) => {
      resolve({ success: code === 0, output })
    })
  })
}
exports.configDispatch = configDispatch
/**
 * @deprecated Use configDispatch instead
 */
const dispatch = (
  cmd,
  arguments_,
  cwd,
  killOnError,
  logger = (data) => log_1.log.info(data)
) => {
  return (0, exports.configDispatch)(cmd, {
    args: arguments_,
    cwd: cwd,
    killOnError: killOnError,
    logger: logger,
  })
}
exports.dispatch = dispatch
