// @zen-browser/surfer/dist/utils/fs.js
'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.getSize =
  exports.ensureEmpty =
  exports.filesExist =
  exports.appendToFileSync =
  exports.mkdirpSync =
  exports.mkdirp =
  exports.ensureDirectory =
  exports.walkDirectoryTree =
  exports.walkDirectory =
  exports.windowsPathToUnix =
    void 0
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const node_fs_1 = require('node:fs')
const promises_1 = require('node:fs/promises')
const node_path_1 = require('node:path')
const log_1 = require('../log')
/**
 * On windows, converts a windows style path to a unix path. On unix, passes the
 * output through to the other side
 *
 * @param path The path that you want to be converted to a unix path
 * @returns A unix path
 */
const windowsPathToUnix = (path) =>
  process.platform == 'win32'
    ? (process.env.SURFER_SIGNING_MODE ? '/c' : '') +
      path.replace(/[\\]+/g, '/').replace(/^([a-zA-Z]+:|\.\/)/, '')
    : path
exports.windowsPathToUnix = windowsPathToUnix
async function walkDirectory(directory, ignore) {
  const output = []
  if (!(0, node_path_1.isAbsolute)(directory)) {
    log_1.log.askForReport()
    log_1.log.error('Please provide an absolute input to walkDirectory')
  }
  try {
    const directoryContents = await (0, promises_1.readdir)(directory)
    for (const file of directoryContents) {
      const fullPath = (0, node_path_1.join)(directory, file)
      const fStat = await (0, promises_1.stat)(fullPath)
      // If the file is in the ignore list, skip it
      if (ignore && fullPath.startsWith(ignore)) {
        continue
      }
      if (fStat.isDirectory()) {
        for (const newFile of await walkDirectory(fullPath)) {
          output.push(newFile)
        }
      } else {
        output.push(fullPath)
      }
    }
  } catch (error) {
    log_1.log.askForReport()
    log_1.log.error(error)
  }
  return output
}
exports.walkDirectory = walkDirectory
async function walkDirectoryTree(directory) {
  const output = {}
  if (!(0, node_path_1.isAbsolute)(directory)) {
    log_1.log.askForReport()
    log_1.log.error('Please provide an absolute input to walkDirectory')
  }
  try {
    const directoryContents = await (0, promises_1.readdir)(directory)
    const currentOut = []
    for (const file of directoryContents) {
      const fullPath = (0, node_path_1.join)(directory, file)
      const fStat = await (0, promises_1.stat)(fullPath)
      if (fStat.isDirectory()) {
        output[file] = await walkDirectoryTree(fullPath)
      } else {
        currentOut.push(fullPath)
      }
    }
    output['.'] = currentOut
  } catch (error) {
    log_1.log.askForReport()
    log_1.log.error(error)
  }
  return output
}
exports.walkDirectoryTree = walkDirectoryTree
async function ensureDirectory(directory) {
  if (!(0, node_fs_1.existsSync)(directory)) {
    await mkdirp(directory)
  }
}
exports.ensureDirectory = ensureDirectory
function mkdirp(directory) {
  return (0, promises_1.mkdir)(directory, { recursive: true })
}
exports.mkdirp = mkdirp
function mkdirpSync(directory) {
  return (0, node_fs_1.mkdirSync)(directory, { recursive: true })
}
exports.mkdirpSync = mkdirpSync
function appendToFileSync(fileName, content) {
  const file = (0, node_fs_1.openSync)(fileName, 'a')
  ;(0, node_fs_1.writeSync)(file, content)
  ;(0, node_fs_1.closeSync)(file)
}
exports.appendToFileSync = appendToFileSync
function filesExist(files) {
  return files.every((file) => (0, node_fs_1.existsSync)(file))
}
exports.filesExist = filesExist
function ensureEmpty(path) {
  if ((0, node_fs_1.existsSync)(path)) {
    ;(0, node_fs_1.rmSync)(path, { recursive: true })
  }
  ;(0, node_fs_1.mkdirSync)(path, { recursive: true })
}
exports.ensureEmpty = ensureEmpty
async function getSize(path) {
  const fileInfo = await (0, promises_1.stat)(path)
  return fileInfo.size
}
exports.getSize = getSize
