// @zen-browser/surfer/dist/utils/task-list.js
'use strict'
var __importDefault =
  (this && this.__importDefault) ||
  function (mod) {
    return mod && mod.__esModule ? mod : { default: mod }
  }
Object.defineProperty(exports, '__esModule', { value: true })
exports.TaskList = void 0
const kleur_1 = __importDefault(require('kleur'))
class TaskList {
  constructor(tasks) {
    this.loggingStyle = 'normal'
    this.loggingIndentation = ''
    this.loggingOnError = 'throw'
    this.tasks = tasks
  }
  style(style) {
    this.loggingStyle = style
    return this
  }
  indent(indentation) {
    this.loggingIndentation = indentation
    return this
  }
  onError(mode) {
    this.loggingOnError = mode
    return this
  }
  log(type, name) {
    if (this.loggingStyle == 'silent') return
    let prefix = this.loggingIndentation
    const prefixTemplate = `[${type.toUpperCase()}]`
    switch (type) {
      case 'start': {
        prefix += kleur_1.default.bold().gray(prefixTemplate)
        break
      }
      case 'finish': {
        prefix += kleur_1.default.bold().green(prefixTemplate)
        break
      }
      case 'fail': {
        prefix += kleur_1.default.bold().red(prefixTemplate)
        break
      }
      case 'skip': {
        prefix += kleur_1.default.bold().yellow(prefixTemplate)
        break
      }
      case 'info': {
        prefix += '  '
        prefix += kleur_1.default.bold().cyan(prefixTemplate)
        break
      }
    }
    console.log(`${prefix} ${name}`)
  }
  async run() {
    for (const task of this.tasks) {
      if (task.skip && (await task.skip())) {
        this.log('skip', task.name)
        continue
      }
      if (task.long) {
        this.log('start', task.name)
      }
      try {
        const result = await task.task((message) => this.log('info', message))
        if (result instanceof TaskList) {
          // We want to provide a start point if none exists already
          if (!task.long) {
            this.log('start', task.name)
          }
          await result.indent(this.loggingIndentation + '  ').run()
        }
      } catch (error) {
        if (this.loggingOnError == 'throw') {
          this.log('fail', task.name)
          throw error
        }
        if (this.loggingOnError == 'inline') {
          this.log('fail', `${task.name}: ${error}`)
        }
        this.error = error
      }
      this.log('finish', task.name)
    }
    if (this.error) {
      throw this.error
    }
  }
}
exports.TaskList = TaskList
