// cross-spawn-cb/dist/cjs/workers/async.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return worker;
    }
});
var _onone = /*#__PURE__*/ _interop_require_default(require("on-one"));
var _constantsts = require("../constants.js");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function worker(cp, options, callback) {
    callback = typeof options === 'function' ? options : callback;
    options = typeof options === 'function' ? {} : options || {};
    // collect output
    var res = {
        pid: 0,
        output: [],
        stdout: null,
        stderr: null,
        status: null,
        signal: null
    };
    var stdout = [];
    var stderr = [];
    if (options.encoding && cp.stdout) cp.stdout.on('data', stdout.push.bind(stdout));
    if (options.encoding && cp.stderr) cp.stderr.on('data', stderr.push.bind(stderr));
    // some versions of node emit both an error and close
    (0, _onone.default)(cp, [
        'error',
        'close'
    ], function(err, status, signal) {
        if (err) {
            if (err.code === 'OK') return; // ignore
            return callback(err);
        }
        // prepare result
        res.pid = cp.pid;
        res.status = status;
        res.signal = signal;
        if (options.encoding && cp.stdout) {
            res.stdout = Buffer.concat(stdout);
            if (options.encoding !== 'binary') res.stdout = res.stdout.toString(options.encoding);
        }
        if (options.encoding && cp.stderr) {
            res.stderr = Buffer.concat(stderr);
            if (options.encoding !== 'binary') res.stderr = res.stderr.toString(options.encoding);
        }
        res.output = [
            null,
            res.stdout,
            res.stderr
        ];
        if (res.status === null) res.status = 0; // patch: early node on windows could return null
        // process errors
        var exitErr = res.status !== 0 ? new Error("Non-zero exit code: ".concat(res.status)) : null;
        if (exitErr) {
            for(var key in res){
                if (_constantsts.spawnKeys.indexOf(key) < 0) continue;
                exitErr[key] = Buffer.isBuffer(res[key]) ? res[key].toString(options.encoding || 'utf8') : res[key];
            }
        }
        exitErr ? callback(exitErr) : callback(null, res);
    });
    // pipe input
    if (options.input && (typeof options.input === 'string' || Buffer.isBuffer(options.input))) {
        cp.stdin.end(options.input);
    }
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
