// function-exec-sync/dist/cjs/serialize-javascript.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return serialize;
    }
});
var _randombytes = /*#__PURE__*/ _interop_require_default(require("randombytes"));
function _instanceof(left, right) {
    if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
        return !!right[Symbol.hasInstance](left);
    } else {
        return left instanceof right;
    }
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
// Patched for backwards compatiblity from: https://github.com/yahoo/serialize-javascript/blob/main/LICENSE
var hasMap = typeof Map !== 'undefined';
var hasSet = typeof Set !== 'undefined';
var hasURL = typeof URL !== 'undefined';
var hasBigInt = typeof BigInt !== 'undefined';
var isArray = Array.isArray || function(x) {
    return Object.prototype.toString.call(x) === '[object Array]';
};
// Generate an internal UID to make the regexp pattern harder to guess.
var UID_LENGTH = 16;
var UID = generateUID();
var PLACE_HOLDER_REGEXP = new RegExp('(\\\\)?"@__(F|R|D|M|S|A|U|I|B|L)-'.concat(UID, '-(\\d+)__@"'), 'g');
var IS_NATIVE_CODE_REGEXP = /\{\s*\[native code\]\s*\}/g;
var IS_PURE_FUNCTION = /function.*?\(/;
var IS_ARROW_FUNCTION = /.*?=>.*?/;
var UNSAFE_CHARS_REGEXP = /[<>/\u2028\u2029]/g;
var RESERVED_SYMBOLS = [
    '*',
    'async'
];
// Mapping of unsafe HTML and invalid JavaScript line terminator chars to their
// Unicode char counterparts which are safe to use in JavaScript strings.
var ESCAPED_CHARS = {
    '<': '\\u003C',
    '>': '\\u003E',
    '/': '\\u002F',
    '\u2028': '\\u2028',
    '\u2029': '\\u2029'
};
function escapeUnsafeChars(unsafeChar) {
    return ESCAPED_CHARS[unsafeChar];
}
function generateUID() {
    var bytes = (0, _randombytes.default)(UID_LENGTH);
    var result = '';
    for(var i = 0; i < UID_LENGTH; ++i){
        result += bytes[i].toString(16);
    }
    return result;
}
function deleteFunctions(obj) {
    var functionKeys = [];
    for(var key in obj){
        if (typeof obj[key] === 'function') {
            functionKeys.push(key);
        }
    }
    for(var i = 0; i < functionKeys.length; i++){
        delete obj[functionKeys[i]];
    }
}
function serialize(obj, options) {
    var opts = {};
    if (options) {
        // Backwards-compatibility for `space` as the second argument.
        if (typeof options === 'number' || typeof options === 'string') {
            opts = {
                space: options
            };
        } else {
            opts = options;
        }
    }
    var functions = [];
    var regexps = [];
    var dates = [];
    var maps = [];
    var sets = [];
    var arrays = [];
    var undefs = [];
    var infinities = [];
    var bigInts = [];
    var urls = [];
    // Returns placeholders for functions and regexps (identified by index)
    // which are later replaced by their string representation.
    function replacer(key, value) {
        // For nested function
        if (opts.ignoreFunction) {
            deleteFunctions(value);
        }
        if (hasBigInt && !value && value !== undefined && value !== BigInt(0)) {
            return value;
        }
        // If the value is an object w/ a toJSON method, toJSON is called before
        // the replacer runs, so we use this[key] to get the non-toJSONed value.
        var origValue = this[key];
        var type = typeof origValue === "undefined" ? "undefined" : _type_of(origValue);
        if (type === 'object') {
            if (_instanceof(origValue, RegExp)) {
                return "@__R-".concat(UID, "-").concat(regexps.push(origValue) - 1, "__@");
            }
            if (_instanceof(origValue, Date)) {
                return "@__D-".concat(UID, "-").concat(dates.push(origValue) - 1, "__@");
            }
            if (hasMap && _instanceof(origValue, Map)) {
                return "@__M-".concat(UID, "-").concat(maps.push(origValue) - 1, "__@");
            }
            if (hasSet && _instanceof(origValue, Set)) {
                return "@__S-".concat(UID, "-").concat(sets.push(origValue) - 1, "__@");
            }
            if (isArray(origValue)) {
                var isSparse = origValue.filter(function() {
                    return true;
                }).length !== origValue.length;
                if (isSparse) {
                    return "@__A-".concat(UID, "-").concat(arrays.push(origValue) - 1, "__@");
                }
            }
            if (hasURL && _instanceof(origValue, URL)) {
                return "@__L-".concat(UID, "-").concat(urls.push(origValue) - 1, "__@");
            }
        }
        if (type === 'function') {
            return "@__F-".concat(UID, "-").concat(functions.push(origValue) - 1, "__@");
        }
        if (type === 'undefined') {
            return "@__U-".concat(UID, "-").concat(undefs.push(origValue) - 1, "__@");
        }
        if (type === 'number' && !Number.isNaN(origValue) && !Number.isFinite(origValue)) {
            return "@__I-".concat(UID, "-").concat(infinities.push(origValue) - 1, "__@");
        }
        if (type === 'bigint') {
            return "@__B-".concat(UID, "-").concat(bigInts.push(origValue) - 1, "__@");
        }
        return value;
    }
    function serializeFunc(fn) {
        var serializedFn = fn.toString();
        if (IS_NATIVE_CODE_REGEXP.test(serializedFn)) {
            throw new TypeError("Serializing native function: ".concat(fn.name));
        }
        // pure functions, example: {key: function() {}}
        if (IS_PURE_FUNCTION.test(serializedFn)) {
            return serializedFn;
        }
        // arrow functions, example: arg1 => arg1+5
        if (IS_ARROW_FUNCTION.test(serializedFn)) {
            return serializedFn;
        }
        var argsStartsAt = serializedFn.indexOf('(');
        var def = serializedFn.substr(0, argsStartsAt).trim().split(' ').filter(function(val) {
            return val.length > 0;
        });
        var nonReservedSymbols = def.filter(function(val) {
            return RESERVED_SYMBOLS.indexOf(val) === -1;
        });
        // enhanced literal objects, example: {key() {}}
        if (nonReservedSymbols.length > 0) {
            return "".concat(def.indexOf('async') > -1 ? 'async ' : '', "function").concat(def.join('').indexOf('*') > -1 ? '*' : '').concat(serializedFn.substr(argsStartsAt));
        }
        // arrow functions
        return serializedFn;
    }
    // Check if the parameter is function
    var objToSerialize = obj;
    if (opts.ignoreFunction && typeof objToSerialize === 'function') {
        objToSerialize = undefined;
    }
    // Protects against `JSON.stringify()` returning `undefined`, by serializing
    // to the literal string: "undefined".
    if (objToSerialize === undefined) {
        return String(objToSerialize);
    }
    var str;
    // Creates a JSON string representation of the value.
    // NOTE: Node 0.12 goes into slow mode with extra JSON.stringify() args.
    if (opts.isJSON && !opts.space) {
        str = JSON.stringify(objToSerialize);
    } else {
        str = JSON.stringify(objToSerialize, opts.isJSON ? null : replacer, opts.space);
    }
    // Protects against `JSON.stringify()` returning `undefined`, by serializing
    // to the literal string: "undefined".
    if (typeof str !== 'string') {
        return String(str);
    }
    // Replace unsafe HTML and invalid JavaScript line terminator chars with
    // their safe Unicode char counterpart. This _must_ happen before the
    // regexps and functions are serialized and added back to the string.
    if (opts.unsafe !== true) {
        str = str.replace(UNSAFE_CHARS_REGEXP, escapeUnsafeChars);
    }
    if (functions.length === 0 && regexps.length === 0 && dates.length === 0 && maps.length === 0 && sets.length === 0 && arrays.length === 0 && undefs.length === 0 && infinities.length === 0 && bigInts.length === 0 && urls.length === 0) {
        return str;
    }
    // Replaces all occurrences of function, regexp, date, map and set placeholders in the
    // JSON string with their string representations. If the original value can
    // not be found, then `undefined` is used.
    return str.replace(PLACE_HOLDER_REGEXP, function(match, backSlash, type, valueIndex) {
        // The placeholder may not be preceded by a backslash. This is to prevent
        // replacing things like `"a\"@__R-<UID>-0__@"` and thus outputting
        // invalid JS.
        if (backSlash) {
            return match;
        }
        if (type === 'D') {
            return 'new Date("'.concat(dates[valueIndex].toISOString(), '")');
        }
        if (type === 'R') {
            return "new RegExp(".concat(serialize(regexps[valueIndex].source), ', "').concat(regexps[valueIndex].flags, '")');
        }
        if (type === 'M') {
            return "new Map(".concat(serialize(Array.from(maps[valueIndex].entries()), opts), ")");
        }
        if (type === 'S') {
            return "new Set(".concat(serialize(Array.from(sets[valueIndex].values()), opts), ")");
        }
        if (type === 'A') {
            return "Array.prototype.slice.call(".concat(serialize(Object.assign({
                length: arrays[valueIndex].length
            }, arrays[valueIndex]), opts), ")");
        }
        if (type === 'U') {
            return 'undefined';
        }
        if (type === 'I') {
            return String(infinities[valueIndex]);
        }
        if (type === 'B') {
            return 'BigInt("'.concat(bigInts[valueIndex], '")');
        }
        if (type === 'L') {
            return "new URL(".concat(serialize(urls[valueIndex].toString(), opts), ")");
        }
        var fn = functions[valueIndex];
        return serializeFunc(fn);
    });
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
