// fs-remove-compat/dist/esm/safeRm.js
import fs from 'fs';
import { fixWinEPERM, shouldFixEPERM } from './fallback/fixWinEPERM.js';
import fallbackRm from './fallback/rm.js';
import { getBackoffDelay, isRetryableError, SAFE_DEFAULTS } from './retry.js';
/**
 * Check if native fs.rm is available.
 */ const HAS_NATIVE_RM = typeof fs.rm === 'function';
function safeRm(path, optionsOrCallback, maybeCallback) {
    // Parse arguments
    if (typeof optionsOrCallback === 'function') {
        safeRmImpl(path, undefined, 0, optionsOrCallback);
    } else {
        safeRmImpl(path, optionsOrCallback, 0, maybeCallback);
    }
}
/**
 * Callback-based implementation with retry.
 */ function safeRmImpl(path, options, attempt, callback) {
    var _ref, _ref1, _ref2, _ref3;
    const opts = {
        recursive: (_ref = options === null || options === void 0 ? void 0 : options.recursive) !== null && _ref !== void 0 ? _ref : SAFE_DEFAULTS.recursive,
        force: (_ref1 = options === null || options === void 0 ? void 0 : options.force) !== null && _ref1 !== void 0 ? _ref1 : SAFE_DEFAULTS.force,
        maxRetries: (_ref2 = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref2 !== void 0 ? _ref2 : SAFE_DEFAULTS.maxRetries,
        retryDelay: (_ref3 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref3 !== void 0 ? _ref3 : SAFE_DEFAULTS.retryDelay
    };
    rmOnce(path, {
        ...opts,
        maxRetries: 0
    }, (err)=>{
        if (!err) return callback();
        // Handle ENOENT with force
        if (err.code === 'ENOENT' && opts.force) {
            return callback();
        }
        // Try EPERM fix on Windows
        if (shouldFixEPERM(err)) {
            return fixWinEPERM(path, err, (fixErr)=>{
                if (!fixErr) return callback();
                // Fall through to retry
                retryIfNeeded(path, options, attempt, err, callback);
            });
        }
        retryIfNeeded(path, options, attempt, err, callback);
    });
}
/**
 * Retry helper for callback style.
 */ function retryIfNeeded(path, options, attempt, err, callback) {
    var _ref, _ref1;
    const maxRetries = (_ref = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref !== void 0 ? _ref : SAFE_DEFAULTS.maxRetries;
    const retryDelay = (_ref1 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref1 !== void 0 ? _ref1 : SAFE_DEFAULTS.retryDelay;
    if (!isRetryableError(err) || attempt >= maxRetries) {
        callback(err);
        return;
    }
    const delay = getBackoffDelay(retryDelay, attempt);
    setTimeout(()=>{
        safeRmImpl(path, options, attempt + 1, callback);
    }, delay);
}
/**
 * Single rm attempt (no retry).
 */ function rmOnce(path, options, callback) {
    if (HAS_NATIVE_RM) {
        fs.rm(path, options, callback);
    } else {
        fallbackRm(path, options, callback);
    }
}
export default safeRm;
