// extract-base-iterator/dist/cjs/FileEntry.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return FileEntry;
    }
});
var _fsremovecompat = require("fs-remove-compat");
var _gracefulfs = /*#__PURE__*/ _interop_require_default(require("graceful-fs"));
var _mkdirpclassic = /*#__PURE__*/ _interop_require_default(require("mkdirp-classic"));
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _queuecb = /*#__PURE__*/ _interop_require_default(require("queue-cb"));
var _chmodts = /*#__PURE__*/ _interop_require_default(require("./fs/chmod.js"));
var _chownts = /*#__PURE__*/ _interop_require_default(require("./fs/chown.js"));
var _utimests = /*#__PURE__*/ _interop_require_default(require("./fs/utimes.js"));
var _indexts = require("./shared/index.js");
var _stripPathts = /*#__PURE__*/ _interop_require_default(require("./shared/stripPath.js"));
var _validateAttributests = /*#__PURE__*/ _interop_require_default(require("./validateAttributes.js"));
var _waitForAccessts = /*#__PURE__*/ _interop_require_default(require("./waitForAccess.js"));
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var MANDATORY_ATTRIBUTES = [
    'mode',
    'mtime',
    'path'
];
var FileEntry = /*#__PURE__*/ function() {
    "use strict";
    function FileEntry(attributes) {
        _class_call_check(this, FileEntry);
        (0, _validateAttributests.default)(attributes, MANDATORY_ATTRIBUTES);
        (0, _indexts.objectAssign)(this, attributes);
        if (this.basename === undefined) this.basename = _path.default.basename(this.path);
        if (this.type === undefined) this.type = 'file';
        if (this._writeFile === undefined) throw new Error('File this missing _writeFile. Please implement this method in your subclass');
    }
    var _proto = FileEntry.prototype;
    _proto.create = function create(dest, options, callback) {
        var _this = this;
        callback = typeof options === 'function' ? options : callback;
        options = typeof options === 'function' ? {} : options || {};
        if (typeof callback === 'function') {
            try {
                var normalizedPath = _path.default.normalize(this.path);
                var fullPath = _path.default.join(dest, (0, _stripPathts.default)(normalizedPath, options));
                var queue = new _queuecb.default(1);
                if (options.force) {
                    queue.defer(function(callback) {
                        (0, _fsremovecompat.rm)(fullPath, function(err) {
                            err && err.code !== 'ENOENT' ? callback(err) : callback();
                        });
                    });
                } else {
                    // Check if file exists - throw EEXIST if it does
                    queue.defer(function(callback) {
                        _gracefulfs.default.stat(fullPath, function(err) {
                            if (!err) {
                                var existsErr = new Error("EEXIST: file already exists, open '".concat(fullPath, "'"));
                                existsErr.code = 'EEXIST';
                                existsErr.path = fullPath;
                                return callback(existsErr);
                            }
                            // ENOENT means file doesn't exist - that's what we want
                            if (err.code === 'ENOENT') return callback();
                            // Other errors should be reported
                            callback(err);
                        });
                    });
                }
                queue.defer(_mkdirpclassic.default.bind(null, _path.default.dirname(fullPath)));
                queue.defer(this._writeFile.bind(this, fullPath, options));
                queue.defer(_waitForAccessts.default.bind(null, fullPath));
                queue.defer(_chmodts.default.bind(null, fullPath, this, options));
                queue.defer(_chownts.default.bind(null, fullPath, this, options));
                queue.defer(_utimests.default.bind(null, fullPath, this, options));
                queue.await(callback);
            } catch (err) {
                callback(err);
            }
            return;
        }
        return new Promise(function(resolve, reject) {
            return _this.create(dest, options, function(err, done) {
                return err ? reject(err) : resolve(done);
            });
        });
    };
    _proto.destroy = function destroy() {};
    return FileEntry;
}();
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
