// extract-base-iterator/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get DirectoryEntry () {
        return _DirectoryEntryts.default;
    },
    get FileEntry () {
        return _FileEntryts.default;
    },
    get LinkEntry () {
        return _LinkEntryts.default;
    },
    get SymbolicLinkEntry () {
        return _SymbolicLinkEntryts.default;
    },
    get default () {
        return ExtractBaseIterator;
    },
    get waitForAccess () {
        return _waitForAccessts.default;
    }
});
var _stackbaseiterator = /*#__PURE__*/ _interop_require_default(require("stack-base-iterator"));
var _DirectoryEntryts = /*#__PURE__*/ _interop_require_default(require("./DirectoryEntry.js"));
var _FileEntryts = /*#__PURE__*/ _interop_require_default(require("./FileEntry.js"));
var _LinkEntryts = /*#__PURE__*/ _interop_require_default(require("./LinkEntry.js"));
var _SymbolicLinkEntryts = /*#__PURE__*/ _interop_require_default(require("./SymbolicLinkEntry.js"));
_export_star(require("./shared/index.js"), exports);
_export_star(require("./types.js"), exports);
var _waitForAccessts = /*#__PURE__*/ _interop_require_default(require("./waitForAccess.js"));
function _assert_this_initialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _call_super(_this, derived, args) {
    derived = _get_prototype_of(derived);
    return _possible_constructor_return(_this, _is_native_reflect_construct() ? Reflect.construct(derived, args || [], _get_prototype_of(_this).constructor) : derived.apply(_this, args));
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
function _get_prototype_of(o) {
    _get_prototype_of = Object.setPrototypeOf ? Object.getPrototypeOf : function getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _get_prototype_of(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _set_prototype_of(subClass, superClass);
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _possible_constructor_return(self, call) {
    if (call && (_type_of(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assert_this_initialized(self);
}
function _set_prototype_of(o, p) {
    _set_prototype_of = Object.setPrototypeOf || function setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _set_prototype_of(o, p);
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _is_native_reflect_construct() {
    try {
        var result = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
    } catch (_) {}
    return (_is_native_reflect_construct = function() {
        return !!result;
    })();
}
var ExtractBaseIterator = /*#__PURE__*/ function(StackBaseIterator) {
    "use strict";
    _inherits(ExtractBaseIterator, StackBaseIterator);
    function ExtractBaseIterator() {
        _class_call_check(this, ExtractBaseIterator);
        return _call_super(this, ExtractBaseIterator, arguments);
    }
    return ExtractBaseIterator;
}(_stackbaseiterator.default);
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
