// 7z-iterator/dist/esm/nextEntry.js
import once from 'call-once-fn';
import { DirectoryEntry, SymbolicLinkEntry } from 'extract-base-iterator';
import path from 'path';
import FileEntry from './FileEntry.js';
export default function nextEntry(iterator, callback) {
    const internalIter = iterator._iterator;
    if (!internalIter) {
        callback(new Error('iterator missing'));
        return;
    }
    let entry = null;
    entry = internalIter.next();
    const nextCallback = once((err, entry)=>{
        // keep processing
        if (entry) iterator.push(nextEntry);
        err ? callback(err) : callback(null, entry ? {
            done: false,
            value: entry
        } : {
            done: true,
            value: null
        });
    });
    // done: signal iteration is complete (guard against stale lock)
    if (!iterator.lock || iterator.isDone() || !entry) return callback(null, {
        done: true,
        value: null
    });
    // Skip anti-files (these mark files to delete in delta archives)
    if (entry.isAntiFile) {
        iterator.push(nextEntry);
        return callback(null, null);
    }
    // Determine type from entry
    const type = entry.type;
    // Default modes (decimal values for Node 0.8 compatibility)
    // 0o755 = 493, 0o644 = 420
    const defaultMode = type === 'directory' ? 493 : 420;
    // Build attributes from 7z entry
    // mtime must be timestamp (number) for FileAttributes compatibility
    const mtimeDate = entry.mtime || new Date();
    const attributes = {
        path: entry.path.split(path.sep).filter(Boolean).join(path.sep),
        basename: entry.name,
        mtime: mtimeDate.getTime(),
        mode: entry.mode !== undefined ? entry.mode : defaultMode
    };
    switch(type){
        case 'directory':
            attributes.type = 'directory';
            return nextCallback(null, new DirectoryEntry(attributes));
        case 'link':
            {
                // For symlinks, the file content IS the symlink target path
                // Read the content to get the linkpath for SymbolicLinkEntry
                const parser = internalIter.getParser();
                const stream = parser.getEntryStream(entry);
                const chunks = [];
                stream.on('data', (chunk)=>{
                    chunks.push(chunk);
                });
                stream.on('end', ()=>{
                    const linkpath = Buffer.concat(chunks).toString('utf8');
                    const linkAttributes = {
                        path: attributes.path,
                        mtime: attributes.mtime,
                        mode: attributes.mode,
                        linkpath: linkpath
                    };
                    nextCallback(null, new SymbolicLinkEntry(linkAttributes));
                });
                stream.on('error', (streamErr)=>{
                    nextCallback(streamErr);
                });
                return;
            }
        case 'file':
            {
                attributes.type = 'file';
                attributes.size = entry.size;
                const parser = internalIter.getParser();
                const stream = parser.getEntryStream(entry);
                return nextCallback(null, new FileEntry(attributes, stream, iterator.lock, entry._canStream));
            }
    }
    return callback(new Error(`Unrecognized entry type: ${type}`));
}
