// 7z-iterator/dist/esm/SevenZipIterator.js
import BaseIterator, { Lock } from 'extract-base-iterator';
import { rmSync } from 'fs-remove-compat';
import fs from 'graceful-fs';
import path from 'path';
import Queue from 'queue-cb';
import shortHash from 'short-hash';
import tempSuffix from 'temp-suffix';
import { tmpdir } from './compat.js';
import streamToSource from './lib/streamToSource.js';
import nextEntry from './nextEntry.js';
import { setPassword } from './sevenz/codecs/index.js';
import { FileSource, SevenZipParser } from './sevenz/SevenZipParser.js';
/**
 * Iterator wrapper around SevenZipParser entries
 */ let EntryIterator = class EntryIterator {
    next() {
        if (this.index >= this.entries.length) {
            return null;
        }
        return this.entries[this.index++];
    }
    getParser() {
        return this.parser;
    }
    constructor(parser){
        this.index = 0;
        this.parser = parser;
        this.entries = parser.getEntries();
    }
};
let SevenZipIterator = class SevenZipIterator extends BaseIterator {
    end(err) {
        if (this.lock) {
            const lock = this.lock;
            this.lock = null; // Clear before release to prevent re-entrancy
            lock.err = err;
            lock.release();
        }
        // Don't call base end here - Lock.__destroy() handles it
        this._iterator = null;
    }
    /**
   * Check if streaming extraction is available for any folder in this archive.
   * Streaming is possible when folders use codecs like BZip2, Deflate, or Copy
   * that can decompress incrementally without buffering the entire input.
   *
   * @returns true if at least one folder supports streaming
   */ canStream() {
        if (!this._iterator) return false;
        const parser = this._iterator.getParser();
        if (!parser) return false;
        const entries = parser.getEntries();
        const checkedFolders = {};
        for(let i = 0; i < entries.length; i++){
            const folderIndex = entries[i]._folderIndex;
            if (folderIndex >= 0 && checkedFolders[folderIndex] === undefined) {
                checkedFolders[folderIndex] = parser.canStreamFolder(folderIndex);
                if (checkedFolders[folderIndex]) {
                    return true;
                }
            }
        }
        return false;
    }
    /**
   * Get entries sorted for optimal streaming extraction.
   *
   * Entries are sorted by:
   * 1. Folder index (process one folder at a time)
   * 2. Stream index within folder (for solid block streaming)
   *
   * This ordering allows multi-file solid folders to stream with
   * O(largest file) memory instead of O(folder size).
   *
   * @returns Array of entries in streaming order
   */ getStreamingOrder() {
        if (!this._iterator) return [];
        const parser = this._iterator.getParser();
        if (!parser) return [];
        const entries = parser.getEntries();
        // Create a copy and sort for streaming order
        const sorted = [];
        for(let i = 0; i < entries.length; i++){
            sorted.push(entries[i]);
        }
        sorted.sort((a, b)=>{
            // First by folder index
            if (a._folderIndex !== b._folderIndex) {
                return a._folderIndex - b._folderIndex;
            }
            // Then by stream index within folder
            return a._streamIndexInFolder - b._streamIndexInFolder;
        });
        return sorted;
    }
    constructor(source, options = {}){
        super(options);
        this.lock = new Lock();
        this.lock.onDestroy = (err)=>BaseIterator.prototype.end.call(this, err);
        const queue = new Queue(1);
        let cancelled = false;
        let archiveSource = null;
        const setup = ()=>{
            cancelled = true;
        };
        this.processing.push(setup);
        // Set password (or clear if not provided)
        setPassword(options.password || null);
        if (typeof source === 'string') {
            // File path input - use FileSource directly
            queue.defer((cb)=>{
                fs.stat(source, (statErr, stats)=>{
                    if (this.done || cancelled) return;
                    if (statErr) return cb(statErr);
                    fs.open(source, 'r', (err, fd)=>{
                        if (this.done || cancelled) return;
                        if (err) return cb(err);
                        archiveSource = new FileSource(fd, stats.size);
                        // Register cleanup for file descriptor
                        this.lock.registerCleanup(()=>{
                            fs.closeSync(fd);
                        });
                        cb();
                    });
                });
            });
        } else {
            // Stream input - write to temp file for random access
            // Register cleanup for source stream
            const stream = source;
            this.lock.registerCleanup(()=>{
                const s = stream;
                if (typeof s.destroy === 'function') s.destroy();
            });
            const tempPath = path.join(tmpdir(), '7z-iterator', shortHash(process.cwd()), tempSuffix('tmp.7z'));
            queue.defer((cb)=>{
                streamToSource(source, {
                    tempPath
                }, (err, result)=>{
                    if (this.done || cancelled) return;
                    if (err) return cb(err);
                    if (!result) return cb(new Error('No result from streamToSource'));
                    archiveSource = result.source;
                    // Register cleanup for file descriptor
                    this.lock.registerCleanup(()=>{
                        fs.closeSync(result.fd);
                    });
                    // Register cleanup for temp file
                    this.lock.registerCleanup(()=>{
                        try {
                            rmSync(result.tempPath);
                        } catch (_e) {
                        /* ignore */ }
                    });
                    cb();
                });
            });
        }
        // Parse and build iterator
        queue.defer((cb)=>{
            if (this.done || cancelled) return;
            if (!archiveSource) return cb(new Error('No archive source'));
            const parser = new SevenZipParser(archiveSource);
            parser.parse((parseErr)=>{
                if (parseErr) {
                    cb(parseErr);
                    return;
                }
                try {
                    this._iterator = new EntryIterator(parser);
                    cb();
                } catch (err) {
                    cb(err);
                }
            });
        });
        // start processing
        queue.await((err)=>{
            this.processing.remove(setup);
            if (this.done || cancelled) return;
            err ? this.end(err) : this.push(nextEntry);
        });
    }
};
export { SevenZipIterator as default };
