// @zen-browser/surfer/dist/index.js
#!/usr/bin/env node
'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.compatMode = exports.bin_name = exports.config = void 0
function getPlatform() {
  if (process.env.SURFER_PLATFORM) {
    return process.env.SURFER_PLATFORM
  }
  return process.platform
}
Object.defineProperty(process, 'surferPlatform', { value: getPlatform() })
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
const commander_1 = require('commander')
const node_fs_1 = require('node:fs')
const node_path_1 = require('node:path')
const utils_1 = require('./utils')
const cmds_1 = require('./cmds')
const constants_1 = require('./constants')
const update_check_1 = require('./middleware/update-check')
const register_command_1 = require('./middleware/register-command')
const log_1 = require('./log')
// We have to use a dynamic require here, otherwise the typescript compiler
// mucks up the directory structure
// eslint-disable-next-line @typescript-eslint/no-var-requires, unicorn/prefer-module
const { version: surferVersion } = require('../package.json')
exports.config = utils_1.config
const program = new commander_1.Command()
let reportedFFVersion
if (
  (0, node_fs_1.existsSync)(
    (0, node_path_1.resolve)(
      constants_1.ENGINE_DIR,
      'browser',
      'config',
      'version.txt'
    )
  )
) {
  const version = (0, node_fs_1.readFileSync)(
    (0, node_path_1.resolve)(
      constants_1.ENGINE_DIR,
      'browser',
      'config',
      'version.txt'
    )
  )
    .toString()
    .replace(/\n/g, '')
  if (version !== (0, utils_1.getFFVersionOrCandidate)())
    reportedFFVersion = version
}
exports.bin_name = constants_1.BIN_NAME
const programVersions = []
exports.compatMode = process.env.SURFER_COMPAT
for (const brand in exports.config.brands) {
  const brandConfig = exports.config.brands[brand]
  programVersions.push({
    name: brandConfig.brandFullName,
    value: brandConfig.release.displayVersion,
  })
}
program
  .storeOptionsAsProperties(false)
  .passCommandToAction(false)
  .name(exports.bin_name)
  .option('-v, --verbose', 'Outputs extra debugging messages to the console')
  .version(
    (0, utils_1.versionFormatter)([
      ...programVersions,
      {
        name: 'Firefox',
        value: `${exports.config.version.version} ${
          reportedFFVersion ? `(being reported as ${reportedFFVersion})` : ''
        }`,
      },
      { name: 'Surfer', value: surferVersion },
      reportedFFVersion
        ? `Mismatch detected between expected Firefox version and the actual version.\nYou may have downloaded the source code using a different version and\nthen switched to another branch.`
        : '',
    ])
  )
async function middleware(command) {
  // If the program is verbose, store that fact within the logger
  log_1.log.isDebug = program.opts().verbose
  await (0, update_check_1.updateCheck)()
  ;(0, register_command_1.registerCommand)(command.name())
}
for (const command of cmds_1.commands) {
  if (
    command.flags &&
    command.flags.platforms &&
    !command.flags.platforms.includes(process.surferPlatform)
  ) {
    continue
  }
  let buildCommand = program
    .command(command.cmd)
    .description(command.description)
    .aliases(command?.aliases || [])
  // Register all of the required options
  for (const opt of command?.options || []) {
    buildCommand =
      opt.parse !== undefined
        ? buildCommand.option(opt.arg, opt.description, opt.parse)
        : buildCommand.option(opt.arg, opt.description)
  }
  buildCommand = buildCommand.action(async (...arguments_) => {
    // Start loading the controller in the background whilst middleware is
    // executing
    const controller = command.requestController()
    if (!command.disableMiddleware) {
      await middleware(buildCommand)
    }
    // Finish loading the controller and execute it
    // eslint-disable-next-line @typescript-eslint/no-extra-semi
    ;(await controller)(...arguments_)
  })
}
process
  .on('uncaughtException', utils_1.errorHandler)
  .on('unhandledException', (error) => (0, utils_1.errorHandler)(error, true))
program.parse(process.argv)
