// fs-remove-compat/dist/esm/fallback/rm.js
import fs from 'fs';
import { fixWinEPERM, shouldFixEPERM } from './fixWinEPERM.js';
const RETRYABLE_CODES = [
    'EBUSY',
    'EMFILE',
    'ENFILE',
    'ENOTEMPTY',
    'EPERM'
];
/**
 * Remove a file asynchronously with retry logic.
 */ function unlinkWithRetry(path, options, attempt, callback) {
    fs.unlink(path, (err)=>{
        if (!err) return callback();
        // Handle ENOENT
        if (err.code === 'ENOENT') {
            if (options.force) return callback();
            return callback(err);
        }
        // Try EPERM fix on Windows
        if (shouldFixEPERM(err)) {
            return fixWinEPERM(path, err, (fixErr)=>{
                if (!fixErr) return callback();
                // Fall through to retry logic
                retryOrFail(path, options, attempt, err, callback, unlinkWithRetry);
            });
        }
        retryOrFail(path, options, attempt, err, callback, unlinkWithRetry);
    });
}
/**
 * Remove a directory asynchronously with retry logic.
 */ function rmdirWithRetry(path, options, attempt, callback) {
    fs.rmdir(path, (err)=>{
        if (!err) return callback();
        if (err.code === 'ENOENT') {
            if (options.force) return callback();
            return callback(err);
        }
        retryOrFail(path, options, attempt, err, callback, rmdirWithRetry);
    });
}
/**
 * Retry logic helper.
 */ function retryOrFail(path, options, attempt, err, callback, retryFn) {
    if (RETRYABLE_CODES.indexOf(err.code || '') === -1 || attempt >= options.maxRetries) {
        callback(err);
        return;
    }
    setTimeout(()=>{
        retryFn(path, options, attempt + 1, callback);
    }, options.retryDelay);
}
/**
 * Remove directory contents recursively.
 */ function rmdirRecursive(path, options, callback) {
    fs.readdir(path, (readErr, entries)=>{
        if (readErr) {
            if (readErr.code === 'ENOENT' && options.force) return callback();
            return callback(readErr);
        }
        let pending = entries.length;
        if (pending === 0) {
            return rmdirWithRetry(path, options, 0, callback);
        }
        let hasError = false;
        for(let i = 0; i < entries.length; i++){
            const entry = entries[i];
            const entryPath = `${path}/${entry}`;
            fs.lstat(entryPath, (statErr, stats)=>{
                if (hasError) return;
                if (statErr) {
                    if (statErr.code === 'ENOENT' && options.force) {
                        pending--;
                        if (pending === 0) rmdirWithRetry(path, options, 0, callback);
                        return;
                    }
                    hasError = true;
                    return callback(statErr);
                }
                const onDone = (err)=>{
                    if (hasError) return;
                    if (err) {
                        hasError = true;
                        return callback(err);
                    }
                    pending--;
                    if (pending === 0) {
                        rmdirWithRetry(path, options, 0, callback);
                    }
                };
                if (stats.isDirectory()) {
                    rmdirRecursive(entryPath, options, onDone);
                } else {
                    unlinkWithRetry(entryPath, options, 0, onDone);
                }
            });
        }
    });
}
/**
 * Fallback rm implementation for Node < 14.14.
 * Matches Node.js fs.rm API.
 */ export default function fallbackRm(path, options, callback) {
    var _ref, _ref1, _ref2, _ref3;
    const opts = {
        recursive: (_ref = options === null || options === void 0 ? void 0 : options.recursive) !== null && _ref !== void 0 ? _ref : false,
        force: (_ref1 = options === null || options === void 0 ? void 0 : options.force) !== null && _ref1 !== void 0 ? _ref1 : false,
        maxRetries: (_ref2 = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref2 !== void 0 ? _ref2 : 0,
        retryDelay: (_ref3 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref3 !== void 0 ? _ref3 : 100
    };
    fs.lstat(path, (statErr, stats)=>{
        if (statErr) {
            if (statErr.code === 'ENOENT') {
                if (opts.force) return callback();
                return callback(statErr);
            }
            return callback(statErr);
        }
        if (stats.isDirectory()) {
            if (!opts.recursive) {
                const err = new Error(`EISDIR: illegal operation on a directory, rm '${path}'`);
                err.code = 'EISDIR';
                err.syscall = 'rm';
                err.path = path;
                return callback(err);
            }
            rmdirRecursive(path, opts, callback);
        } else {
            unlinkWithRetry(path, opts, 0, callback);
        }
    });
}
