// fs-remove-compat/dist/esm/fallback/rmSync.js
import fs from 'fs';
import { fixWinEPERMSync, shouldFixEPERM } from './fixWinEPERM.js';
const RETRYABLE_CODES = [
    'EBUSY',
    'EMFILE',
    'ENFILE',
    'ENOTEMPTY',
    'EPERM'
];
/**
 * Busy-wait for sync retry delay.
 */ function busyWait(ms) {
    const end = Date.now() + ms;
    while(Date.now() < end){
    // Busy wait
    }
}
/**
 * Remove a file synchronously.
 */ function unlinkSync(path, options) {
    for(let attempt = 0; attempt <= options.maxRetries; attempt++){
        try {
            fs.unlinkSync(path);
            return;
        } catch (err) {
            const error = err;
            // Handle ENOENT
            if (error.code === 'ENOENT') {
                if (options.force) return;
                throw error;
            }
            // Try EPERM fix on Windows
            if (shouldFixEPERM(error)) {
                try {
                    fixWinEPERMSync(path, error);
                    return;
                } catch (_fixErr) {
                // Fall through to retry logic
                }
            }
            // Check if retryable
            if (RETRYABLE_CODES.indexOf(error.code || '') === -1 || attempt >= options.maxRetries) {
                throw error;
            }
            // Wait before retry (linear backoff like Node.js)
            busyWait(options.retryDelay);
        }
    }
}
/**
 * Remove a directory synchronously (non-recursive).
 */ function rmdirSync(path, options) {
    for(let attempt = 0; attempt <= options.maxRetries; attempt++){
        try {
            fs.rmdirSync(path);
            return;
        } catch (err) {
            const error = err;
            if (error.code === 'ENOENT') {
                if (options.force) return;
                throw error;
            }
            if (RETRYABLE_CODES.indexOf(error.code || '') === -1 || attempt >= options.maxRetries) {
                throw error;
            }
            busyWait(options.retryDelay);
        }
    }
}
/**
 * Remove directory contents recursively.
 */ function rmdirRecursiveSync(path, options) {
    let entries;
    try {
        entries = fs.readdirSync(path);
    } catch (err) {
        const error = err;
        if (error.code === 'ENOENT' && options.force) return;
        throw error;
    }
    for(let i = 0; i < entries.length; i++){
        const entry = entries[i];
        const entryPath = `${path}/${entry}`;
        let stats;
        try {
            stats = fs.lstatSync(entryPath);
        } catch (err) {
            const error = err;
            if (error.code === 'ENOENT' && options.force) continue;
            throw error;
        }
        if (stats.isDirectory()) {
            rmdirRecursiveSync(entryPath, options);
        } else {
            unlinkSync(entryPath, options);
        }
    }
    // Now remove the directory itself
    rmdirSync(path, options);
}
/**
 * Fallback rmSync implementation for Node < 14.14.
 * Matches Node.js fs.rmSync API.
 */ export default function fallbackRmSync(path, options) {
    var _ref, _ref1, _ref2, _ref3;
    const opts = {
        recursive: (_ref = options === null || options === void 0 ? void 0 : options.recursive) !== null && _ref !== void 0 ? _ref : false,
        force: (_ref1 = options === null || options === void 0 ? void 0 : options.force) !== null && _ref1 !== void 0 ? _ref1 : false,
        maxRetries: (_ref2 = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref2 !== void 0 ? _ref2 : 0,
        retryDelay: (_ref3 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref3 !== void 0 ? _ref3 : 100
    };
    let stats;
    try {
        stats = fs.lstatSync(path);
    } catch (err) {
        const error = err;
        if (error.code === 'ENOENT') {
            if (opts.force) return;
            throw error;
        }
        throw error;
    }
    if (stats.isDirectory()) {
        if (!opts.recursive) {
            const err = new Error(`EISDIR: illegal operation on a directory, rm '${path}'`);
            err.code = 'EISDIR';
            err.syscall = 'rm';
            err.path = path;
            throw err;
        }
        rmdirRecursiveSync(path, opts);
    } else {
        unlinkSync(path, opts);
    }
}
