// xz-compat/dist/esm/index.js
/**
 * XZ-Compat: XZ/LZMA Decompression Library
 *
 * Pure JavaScript implementation with optional native acceleration
 * via lzma-native on Node.js 10+.
 *
 * Works on Node.js 0.8+ with automatic performance optimization
 * when native bindings are available.
 */ // ============================================================================
// High-Level APIs (Recommended)
// ============================================================================
// 7z-specific decoders - accept properties separately, try native automatically
export { decode7zLzma, decode7zLzma2 } from './sevenz.js';
// XZ container format - self-describing, works great with native acceleration
export { createXZDecoder, decodeXZ } from './xz/Decoder.js';
// ============================================================================
// Low-Level APIs (Backward Compatibility)
// ============================================================================
// Raw LZMA decoders (for specialized use cases)
export { createLzma2Decoder, createLzmaDecoder, decodeLzma, decodeLzma2 } from './lzma/index.js';
// ============================================================================
// Supporting APIs
// ============================================================================
// Preprocessing filters (BCJ/Delta - used by 7z-iterator)
export * from './filters/index.js';
// Native acceleration utilities
export { isNativeAvailable } from './native.js';
