// @zen-browser/surfer/dist/commands/updates/browser.js
'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.generateBrowserUpdateFiles = exports.getPlatformConfig = void 0
const node_fs_1 = require('node:fs')
const promises_1 = require('node:fs/promises')
const ini_1 = require('ini')
const node_path_1 = require('node:path')
const xmlbuilder2_1 = require('xmlbuilder2')
const __1 = require('../..')
const constants_1 = require('../../constants')
const log_1 = require('../../log')
const utils_1 = require('../../utils')
/**
 * These are all of the different platforms that aus should deploy to. Note that
 * the names have been simplified and they are all only the ones that are
 * supported by Pulse Browser. If you need something else, open an issue on gh.
 *
 * Based off the code from mozrelease:
 * https://searchfox.org/mozilla-central/source/python/mozrelease/mozrelease/platforms.py
 * https://searchfox.org/mozilla-central/source/taskcluster/gecko_taskgraph/util/partials.py
 */
const ausPlatformsMap = {
  linux64: ['Linux_x86_64-gcc3'],
  linuxArm: ['Linux_aarch64-gcc3'],
  macosIntel: [
    'Darwin_x86_64-gcc3-u-i386-x86_64',
    'Darwin_x86-gcc3-u-i386-x86_64',
    'Darwin_x86-gcc3',
    'Darwin_x86_64-gcc3',
  ],
  macosArm: ['Darwin_aarch64-gcc3'],
  win64: ['WINNT_x86_64-msvc', 'WINNT_x86_64-msvc-x64'],
  winArm: ['WINNT_aarch64-msvc-aarch64'],
}
async function getPlatformConfig() {
  let platformINI = (0, node_path_1.join)(
    constants_1.OBJ_DIR,
    'dist',
    __1.config.binaryName,
    'platform.ini'
  )
  if (!(0, node_fs_1.existsSync)(platformINI))
    platformINI = (0, node_path_1.join)(
      constants_1.OBJ_DIR,
      'dist',
      'bin',
      'platform.ini'
    )
  const iniContents = await (0, promises_1.readFile)(platformINI)
  return (0, ini_1.parse)(iniContents.toString())
}
exports.getPlatformConfig = getPlatformConfig
function getReleaseMarName(releaseInfo) {
  let releaseMarName
  if (process.surferPlatform == 'win32') {
    if (__1.compatMode == 'x86_64') {
      releaseMarName = 'windows.mar'
    } else if (__1.compatMode == 'aarch64') {
      releaseMarName = 'windows-arm64.mar'
    }
  }
  if (process.surferPlatform == 'darwin') {
    releaseMarName = 'macos.mar' // universal binary
  }
  if (process.surferPlatform == 'linux') {
    if (__1.compatMode == 'x86_64') {
      releaseMarName = 'linux.mar'
    } else if (__1.compatMode == 'aarch64') {
      releaseMarName = 'linux-aarch64.mar'
    }
  }
  return releaseMarName
}
function getReleaseMarURL(releaseInfo) {
  const releaseMarName = getReleaseMarName(releaseInfo)
  let completeMarURL = `https://${
    __1.config.updateHostname || 'localhost:8000'
  }/${releaseMarName || 'output.mar'}`
  // The user is using github to distribute release binaries for this version.
  if (releaseInfo.github) {
    let releaseVersion = releaseInfo.displayVersion
    const channel = utils_1.dynamicConfig.get('brand')
    if (channel === 'twilight') {
      releaseVersion = 'twilight'
    }
    completeMarURL = `https://github.com/${releaseInfo.github.repo}/releases/download/${releaseVersion}/${releaseMarName}`
    log_1.log.info(`Using '${completeMarURL}' as the distribution url`)
  } else {
    log_1.log.warning(
      `No release information found! Default release location will be "${completeMarURL}"`
    )
  }
  return completeMarURL
}
async function writeUpdateFileToDisk(target, channel, updateObject) {
  const xmlPath = (0, node_path_1.join)(
    constants_1.DIST_DIR,
    'update',
    'browser',
    target,
    channel,
    'update.xml'
  )
  const document = (0, xmlbuilder2_1.create)(updateObject)
  ;(0, utils_1.ensureEmpty)((0, node_path_1.dirname)(xmlPath))
  await (0, promises_1.writeFile)(xmlPath, document.end({ prettyPrint: true }))
}
function getTargets() {
  if (process.surferPlatform == 'win32') {
    return __1.compatMode == 'aarch64'
      ? ausPlatformsMap.winArm
      : ausPlatformsMap.win64
  }
  if (process.surferPlatform == 'linux') {
    return __1.compatMode == 'aarch64'
      ? ausPlatformsMap.linuxArm
      : ausPlatformsMap.linux64
  }
  if (process.surferPlatform == 'darwin') {
    return [...ausPlatformsMap.macosArm, ...ausPlatformsMap.macosIntel]
  }
  log_1.log.error('Unknown platform')
  return []
}
async function generateBrowserUpdateFiles() {
  log_1.log.info('Creating browser AUS update files')
  const brandingKey = utils_1.dynamicConfig.get('brand')
  const channel = brandingKey
  const brandingDetails = __1.config.brands[brandingKey]
  const releaseInfo = brandingDetails.release
  const { displayVersion: version } = releaseInfo
  const marPath = utils_1.dynamicConfig.get('marPath')
  if (!marPath || marPath == '') {
    log_1.log.error(
      `No mar file has been built! Make sure you ran |${__1.bin_name} package| before this command`
    )
    return
  }
  // We need the sha512 hash of the mar file for the update file. AUS will use
  // this to ensure that the mar file has not been modified on the distribution
  // server
  const marHash = (0, utils_1.generateHash)(marPath, 'sha512')
  // We need platform information, primarily for the BuildID, but other stuff
  // might be helpful later
  const platform = await getPlatformConfig()
  const completeMarURL = getReleaseMarURL(releaseInfo)
  const updateObject = {
    updates: {
      update: {
        // TODO: Correct update type from semvar, store the old version somewhere
        '@type': 'minor',
        '@displayVersion': version,
        '@appVersion': version,
        '@platformVersion': (0, utils_1.getFFVersionOrCandidate)(),
        '@buildID': platform.Build.BuildID,
        patch: {
          // TODO: Partial patches might be nice for download speed
          '@type': 'complete',
          '@URL': completeMarURL,
          '@hashFunction': 'sha512',
          '@hashValue': await marHash,
          '@size': await (0, utils_1.getSize)(marPath),
        },
      },
    },
  }
  for (const target of getTargets()) {
    await writeUpdateFileToDisk(target, channel, updateObject)
  }
}
exports.generateBrowserUpdateFiles = generateBrowserUpdateFiles
