// install-module-linked/dist/cjs/cli.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
var _exitcompat = /*#__PURE__*/ _interop_require_default(require("exit-compat"));
var _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
var _getoptscompat = /*#__PURE__*/ _interop_require_default(require("getopts-compat"));
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _url = /*#__PURE__*/ _interop_require_default(require("url"));
var _clearts = /*#__PURE__*/ _interop_require_default(require("./lib/clear.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var __dirname = _path.default.dirname(typeof __filename !== 'undefined' ? __filename : _url.default.fileURLToPath(require("url").pathToFileURL(__filename).toString()));
var ERROR_CODE = 7;
function getVersion() {
    var packageJsonPath = _path.default.join(__dirname, '..', '..', 'package.json');
    var packageJson = JSON.parse(_fs.default.readFileSync(packageJsonPath, 'utf8'));
    return packageJson.version;
}
function printHelp(name) {
    var version = getVersion();
    console.log("".concat(name, " v").concat(version));
    console.log('');
    console.log("Usage: ".concat(name, " [options] [command]"));
    console.log('');
    console.log('Commands:');
    console.log('  clear    Clear the module cache');
    console.log('');
    console.log('Options:');
    console.log('  -v, --version    Print version number');
    console.log('  -h, --help       Print this help message');
}
var _default = function(argv, name) {
    name = name || 'iml';
    var options = (0, _getoptscompat.default)(argv, {
        alias: {
            version: 'v',
            help: 'h'
        },
        boolean: [
            'version',
            'help'
        ],
        stopEarly: true
    });
    if (options.version) {
        console.log(getVersion());
        (0, _exitcompat.default)(0);
        return;
    }
    if (options.help) {
        printHelp(name);
        (0, _exitcompat.default)(0);
        return;
    }
    var args = options._;
    if (!args.length) {
        console.log("Missing command. Example usage: ".concat(name, " [command]"));
        console.log('Run "'.concat(name, ' --help" for more information.'));
        (0, _exitcompat.default)(ERROR_CODE);
        return;
    }
    if (args[0] === 'clear') {
        (0, _clearts.default)();
        (0, _exitcompat.default)(0);
        return;
    }
    console.log("Unrecognized command: ".concat(args[0], ". Example usage: ").concat(name, " [command]"));
    console.log('Run "'.concat(name, ' --help" for more information.'));
    (0, _exitcompat.default)(ERROR_CODE);
};
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
