// get-file-compat/dist/esm/getContent.js
import { bind } from 'node-version-call-local';
import oo from 'on-one';
import path from 'path';
import url from 'url';
import { bufferFrom } from './compat.js';
import makeRequest from './lib/makeRequest.js';
const major = +process.versions.node.split('.')[0];
const minor = +process.versions.node.split('.')[1];
const noHTTPS = major === 0 && (minor <= 8 || minor === 12);
const __dirname = path.dirname(typeof __filename === 'undefined' ? url.fileURLToPath(import.meta.url) : __filename);
const workerPath = path.join(__dirname, '..', 'cjs', 'getContent.js');
function run(endpoint, encoding, options, callback) {
    makeRequest(endpoint, {
        timeout: options.timeout
    }, (err, res)=>{
        if (err) return callback(err);
        const chunks = [];
        res.on('data', (chunk)=>{
            chunks.push(chunk);
        });
        oo(res, [
            'error',
            'end',
            'close',
            'finish'
        ], (err)=>{
            if (err) return callback(err);
            const buffer = Buffer.concat(chunks);
            const content = encoding ? buffer.toString(encoding) : buffer;
            callback(null, {
                content,
                headers: res.headers,
                statusCode: res.statusCode
            });
        });
    });
}
// spawnOptions: false - no node/npm spawn (network only)
const call = bind('>0', workerPath, {
    callbacks: true,
    spawnOptions: false
});
function worker(endpoint, encoding, options, callback) {
    if (!noHTTPS) {
        run(endpoint, encoding, options, callback);
        return;
    }
    const enc = encoding || 'base64';
    call(endpoint, enc, options, (err, result)=>{
        if (err) return callback(err);
        const content = encoding ? result.content : bufferFrom(result.content, 'base64');
        callback(null, {
            content,
            headers: result.headers,
            statusCode: result.statusCode
        });
    });
}
export default function getContent(endpoint, encodingOrOptionsOrCallback, optionsOrCallback, callback) {
    let encoding = null;
    let options = {};
    let cb;
    // Parse arguments
    if (typeof encodingOrOptionsOrCallback === 'function') {
        cb = encodingOrOptionsOrCallback;
    } else if (typeof encodingOrOptionsOrCallback === 'string') {
        encoding = encodingOrOptionsOrCallback;
        if (typeof optionsOrCallback === 'function') {
            cb = optionsOrCallback;
        } else if (optionsOrCallback) {
            options = optionsOrCallback;
            cb = callback;
        }
    } else if (encodingOrOptionsOrCallback) {
        options = encodingOrOptionsOrCallback;
        if (typeof optionsOrCallback === 'function') {
            cb = optionsOrCallback;
        }
    }
    if (typeof cb === 'function') return worker(endpoint, encoding, options, cb);
    return new Promise((resolve, reject)=>worker(endpoint, encoding, options, (err, result)=>err ? reject(err) : resolve(result)));
}
