// fs-remove-compat/dist/cjs/retry.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get SAFE_DEFAULTS () {
        return SAFE_DEFAULTS;
    },
    get busyWait () {
        return busyWait;
    },
    get getBackoffDelay () {
        return getBackoffDelay;
    },
    get isRetryableError () {
        return isRetryableError;
    }
});
function _instanceof(left, right) {
    if (right != null && typeof Symbol !== "undefined" && right[Symbol.hasInstance]) {
        return !!right[Symbol.hasInstance](left);
    } else {
        return left instanceof right;
    }
}
/**
 * Retry utilities for safeRm/safeRmSync with exponential backoff.
 */ var isWindows = process.platform === 'win32' || /^(msys|cygwin)$/.test(process.env.OSTYPE);
var RETRYABLE_CODES = [
    'EBUSY',
    'EMFILE',
    'ENFILE',
    'ENOTEMPTY',
    'EPERM'
];
var BACKOFF_FACTOR = 1.2;
var SAFE_DEFAULTS = {
    recursive: true,
    force: true,
    maxRetries: isWindows ? 10 : 0,
    retryDelay: 100
};
function isRetryableError(err) {
    if (!_instanceof(err, Error)) return false;
    var code = err.code;
    return code !== undefined && RETRYABLE_CODES.indexOf(code) !== -1;
}
function getBackoffDelay(baseDelay, attempt) {
    return Math.floor(baseDelay * Math.pow(BACKOFF_FACTOR, attempt));
}
function busyWait(ms) {
    var end = Date.now() + ms;
    while(Date.now() < end){
    // Busy wait
    }
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
