// fs-remove-compat/dist/cjs/safeRmSync.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
var _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
var _fixWinEPERMts = require("./fallback/fixWinEPERM.js");
var _rmSyncts = /*#__PURE__*/ _interop_require_default(require("./fallback/rmSync.js"));
var _retryts = require("./retry.js");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
/**
 * Check if native fs.rmSync is available.
 */ var HAS_NATIVE_RM_SYNC = typeof _fs.default.rmSync === 'function';
/**
 * Remove a file or directory synchronously with Windows-friendly defaults.
 *
 * This is NOT a strict ponyfill - it provides enhanced behavior:
 * - Default maxRetries: 10 on Windows, 0 on POSIX
 * - Exponential backoff (1.2 factor) instead of linear
 * - EPERM chmod fix before retry on Windows
 *
 * Use this for CI/test cleanup where Windows file locking is common.
 * For strict Node.js fs.rmSync compatibility, use rmSync instead.
 *
 * @param path - Path to remove
 * @param options - Options (with Windows-friendly defaults)
 */ function safeRmSync(path, options) {
    var _ref, _ref1, _ref2, _ref3;
    var opts = {
        recursive: (_ref = options === null || options === void 0 ? void 0 : options.recursive) !== null && _ref !== void 0 ? _ref : _retryts.SAFE_DEFAULTS.recursive,
        force: (_ref1 = options === null || options === void 0 ? void 0 : options.force) !== null && _ref1 !== void 0 ? _ref1 : _retryts.SAFE_DEFAULTS.force,
        maxRetries: (_ref2 = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref2 !== void 0 ? _ref2 : _retryts.SAFE_DEFAULTS.maxRetries,
        retryDelay: (_ref3 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref3 !== void 0 ? _ref3 : _retryts.SAFE_DEFAULTS.retryDelay
    };
    // If no retries requested, use standard implementation
    if (opts.maxRetries === 0) {
        if (HAS_NATIVE_RM_SYNC) {
            _fs.default.rmSync(path, opts);
        } else {
            (0, _rmSyncts.default)(path, opts);
        }
        return;
    }
    // Retry with exponential backoff
    for(var attempt = 0; attempt <= opts.maxRetries; attempt++){
        try {
            if (HAS_NATIVE_RM_SYNC) {
                // Use native but pass maxRetries=0 since we handle retries ourselves
                _fs.default.rmSync(path, _object_spread_props(_object_spread({}, opts), {
                    maxRetries: 0
                }));
            } else {
                (0, _rmSyncts.default)(path, _object_spread_props(_object_spread({}, opts), {
                    maxRetries: 0
                }));
            }
            return; // Success
        } catch (err) {
            // Handle ENOENT with force
            if (err.code === 'ENOENT' && opts.force) {
                return;
            }
            // Try EPERM fix on Windows
            if ((0, _fixWinEPERMts.shouldFixEPERM)(err)) {
                try {
                    (0, _fixWinEPERMts.fixWinEPERMSync)(path, err);
                    return; // Success after chmod fix
                } catch (_fixErr) {
                // Fall through to retry
                }
            }
            // Check if we should retry
            if (!(0, _retryts.isRetryableError)(err) || attempt >= opts.maxRetries) {
                throw err;
            }
            // Exponential backoff
            var delay = (0, _retryts.getBackoffDelay)(opts.retryDelay, attempt);
            (0, _retryts.busyWait)(delay);
        }
    }
}
var _default = safeRmSync;
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
