// fs-remove-compat/dist/cjs/safeRm.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
var _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
var _fixWinEPERMts = require("./fallback/fixWinEPERM.js");
var _rmts = /*#__PURE__*/ _interop_require_default(require("./fallback/rm.js"));
var _retryts = require("./retry.js");
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
/**
 * Check if native fs.rm is available.
 */ var HAS_NATIVE_RM = typeof _fs.default.rm === 'function';
function safeRm(path, optionsOrCallback, maybeCallback) {
    // Parse arguments
    if (typeof optionsOrCallback === 'function') {
        safeRmImpl(path, undefined, 0, optionsOrCallback);
    } else {
        safeRmImpl(path, optionsOrCallback, 0, maybeCallback);
    }
}
/**
 * Callback-based implementation with retry.
 */ function safeRmImpl(path, options, attempt, callback) {
    var _ref, _ref1, _ref2, _ref3;
    var opts = {
        recursive: (_ref = options === null || options === void 0 ? void 0 : options.recursive) !== null && _ref !== void 0 ? _ref : _retryts.SAFE_DEFAULTS.recursive,
        force: (_ref1 = options === null || options === void 0 ? void 0 : options.force) !== null && _ref1 !== void 0 ? _ref1 : _retryts.SAFE_DEFAULTS.force,
        maxRetries: (_ref2 = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref2 !== void 0 ? _ref2 : _retryts.SAFE_DEFAULTS.maxRetries,
        retryDelay: (_ref3 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref3 !== void 0 ? _ref3 : _retryts.SAFE_DEFAULTS.retryDelay
    };
    rmOnce(path, _object_spread_props(_object_spread({}, opts), {
        maxRetries: 0
    }), function(err) {
        if (!err) return callback();
        // Handle ENOENT with force
        if (err.code === 'ENOENT' && opts.force) {
            return callback();
        }
        // Try EPERM fix on Windows
        if ((0, _fixWinEPERMts.shouldFixEPERM)(err)) {
            return (0, _fixWinEPERMts.fixWinEPERM)(path, err, function(fixErr) {
                if (!fixErr) return callback();
                // Fall through to retry
                retryIfNeeded(path, options, attempt, err, callback);
            });
        }
        retryIfNeeded(path, options, attempt, err, callback);
    });
}
/**
 * Retry helper for callback style.
 */ function retryIfNeeded(path, options, attempt, err, callback) {
    var _ref, _ref1;
    var maxRetries = (_ref = options === null || options === void 0 ? void 0 : options.maxRetries) !== null && _ref !== void 0 ? _ref : _retryts.SAFE_DEFAULTS.maxRetries;
    var retryDelay = (_ref1 = options === null || options === void 0 ? void 0 : options.retryDelay) !== null && _ref1 !== void 0 ? _ref1 : _retryts.SAFE_DEFAULTS.retryDelay;
    if (!(0, _retryts.isRetryableError)(err) || attempt >= maxRetries) {
        callback(err);
        return;
    }
    var delay = (0, _retryts.getBackoffDelay)(retryDelay, attempt);
    setTimeout(function() {
        safeRmImpl(path, options, attempt + 1, callback);
    }, delay);
}
/**
 * Single rm attempt (no retry).
 */ function rmOnce(path, options, callback) {
    if (HAS_NATIVE_RM) {
        _fs.default.rm(path, options, callback);
    } else {
        (0, _rmts.default)(path, options, callback);
    }
}
var _default = safeRm;
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
