// module-compat/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get extToModuleType () {
        return _extToModuleTypets.default;
    },
    get loadModule () {
        return _loadModulets.default;
    },
    get loadModuleSync () {
        return _loadModuleSyncts.default;
    },
    get moduleType () {
        return _moduleTypets.default;
    },
    get supportsESM () {
        return _supportsESMts.default;
    },
    get supportsSyncRequireESM () {
        return _supportsSyncRequireESMts.default;
    }
});
var _extToModuleTypets = /*#__PURE__*/ _interop_require_default(require("./extToModuleType.js"));
var _loadModulets = /*#__PURE__*/ _interop_require_default(require("./loadModule.js"));
var _loadModuleSyncts = /*#__PURE__*/ _interop_require_default(require("./loadModuleSync.js"));
var _moduleTypets = /*#__PURE__*/ _interop_require_default(require("./moduleType.js"));
var _supportsESMts = /*#__PURE__*/ _interop_require_default(require("./supportsESM.js"));
var _supportsSyncRequireESMts = /*#__PURE__*/ _interop_require_default(require("./supportsSyncRequireESM.js"));
_export_star(require("./types.js"), exports);
function _export_star(from, to) {
    Object.keys(from).forEach(function(k) {
        if (k !== "default" && !Object.prototype.hasOwnProperty.call(to, k)) {
            Object.defineProperty(to, k, {
                enumerable: true,
                get: function() {
                    return from[k];
                }
            });
        }
    });
    return from;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
