// fast-extract/dist/cjs/compat/flush-write-stream.js
// Writable stream with flush callback support for Node 0.8+
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return flushWriteStream;
    }
});
var _extractbaseiterator = require("extract-base-iterator");
var _bufferts = require("./buffer.js");
function _assert_this_initialized(self) {
    if (self === void 0) {
        throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
    }
    return self;
}
function _call_super(_this, derived, args) {
    derived = _get_prototype_of(derived);
    return _possible_constructor_return(_this, _is_native_reflect_construct() ? Reflect.construct(derived, args || [], _get_prototype_of(_this).constructor) : derived.apply(_this, args));
}
function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _get(target, property, receiver) {
    if (typeof Reflect !== "undefined" && Reflect.get) {
        _get = Reflect.get;
    } else {
        _get = function get(target, property, receiver) {
            var base = _super_prop_base(target, property);
            if (!base) return;
            var desc = Object.getOwnPropertyDescriptor(base, property);
            if (desc.get) {
                return desc.get.call(receiver || target);
            }
            return desc.value;
        };
    }
    return _get(target, property, receiver || target);
}
function _get_prototype_of(o) {
    _get_prototype_of = Object.setPrototypeOf ? Object.getPrototypeOf : function getPrototypeOf(o) {
        return o.__proto__ || Object.getPrototypeOf(o);
    };
    return _get_prototype_of(o);
}
function _inherits(subClass, superClass) {
    if (typeof superClass !== "function" && superClass !== null) {
        throw new TypeError("Super expression must either be null or a function");
    }
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            writable: true,
            configurable: true
        }
    });
    if (superClass) _set_prototype_of(subClass, superClass);
}
function _possible_constructor_return(self, call) {
    if (call && (_type_of(call) === "object" || typeof call === "function")) {
        return call;
    }
    return _assert_this_initialized(self);
}
function _set_prototype_of(o, p) {
    _set_prototype_of = Object.setPrototypeOf || function setPrototypeOf(o, p) {
        o.__proto__ = p;
        return o;
    };
    return _set_prototype_of(o, p);
}
function _super_prop_base(object, property) {
    while(!Object.prototype.hasOwnProperty.call(object, property)){
        object = _get_prototype_of(object);
        if (object === null) break;
    }
    return object;
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _is_native_reflect_construct() {
    try {
        var result = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function() {}));
    } catch (_) {}
    return (_is_native_reflect_construct = function() {
        return !!result;
    })();
}
// Signal buffer for pre-_final Node versions
var SIGNAL_FLUSH = (0, _bufferts.bufferFrom)([
    0
]);
var FlushWriteStream = /*#__PURE__*/ function(Writable) {
    "use strict";
    _inherits(FlushWriteStream, Writable);
    function FlushWriteStream(opts, writeFn, flushFn) {
        _class_call_check(this, FlushWriteStream);
        var _this;
        _this = _call_super(this, FlushWriteStream, [
            opts || {}
        ]), _this.destroyed = false;
        _this._writeFn = writeFn;
        _this._flushFn = flushFn;
        return _this;
    }
    var _proto = FlushWriteStream.prototype;
    // biome-ignore lint/suspicious/noExplicitAny: Object mode allows any chunk type
    _proto._write = function _write(chunk, enc, cb) {
        if (chunk === SIGNAL_FLUSH) {
            if (this._flushFn) {
                this._flushFn(cb);
            } else {
                cb();
            }
        } else {
            this._writeFn.call(this, chunk, enc, cb);
        }
    };
    // biome-ignore lint/suspicious/noExplicitAny: Matches stream.Writable.end signature
    _proto.end = function end(chunk, enc, cb) {
        var _this = this;
        if (!this._flushFn) return _get(_get_prototype_of(FlushWriteStream.prototype), "end", this).call(this, chunk, enc, cb);
        var ref, ref1;
        // Normalize arguments
        if (typeof chunk === 'function') ref = [
            null,
            chunk
        ], chunk = ref[0], cb = ref[1], ref;
        else if (typeof enc === 'function') ref1 = [
            null,
            enc
        ], enc = ref1[0], cb = ref1[1], ref1;
        if (chunk) this.write(chunk);
        // Wait for SIGNAL_FLUSH write to complete before ending stream
        // biome-ignore lint/suspicious/noExplicitAny: Access internal _writableState
        if (!this._writableState.ending) {
            this.write(SIGNAL_FLUSH, function() {
                _get(_get_prototype_of(FlushWriteStream.prototype), "end", _this).call(_this, cb);
            });
            return this;
        }
        return _get(_get_prototype_of(FlushWriteStream.prototype), "end", this).call(this, cb);
    };
    _proto.destroy = function destroy(err) {
        if (this.destroyed) return this;
        this.destroyed = true;
        if (err) this.emit('error', err);
        this.emit('close');
        return this;
    };
    return FlushWriteStream;
}(_extractbaseiterator.Writable);
function flushWriteStream(opts, writeFn, flushFn) {
    if (typeof opts === 'function') return new FlushWriteStream(undefined, opts, writeFn);
    return new FlushWriteStream(opts, writeFn, flushFn);
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
