// 7z-iterator/dist/cjs/sevenz/codecs/Aes.js
// 7z AES-256-CBC codec with SHA-256 key derivation
// Implements password-based decryption for encrypted 7z archives
//
// Properties format:
//   Byte 0: bits 0-5 = NumCyclesPower (iterations = 2^NumCyclesPower)
//           bit 6 = IV present flag
//           bit 7 = Salt present flag
//   Byte 1: upper nibble = salt size extension
//           lower nibble = IV size extension
//   Following bytes: salt data, then IV data
//
// Key derivation:
//   For each round (2^NumCyclesPower times):
//     hash = SHA256(salt + password_utf16le + round_counter_8bytes)
//   Final key = first 32 bytes of accumulated hash
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get createAesDecoder () {
        return createAesDecoder;
    },
    get decodeAes () {
        return decodeAes;
    },
    get getPassword () {
        return getPassword;
    },
    get setPassword () {
        return setPassword;
    }
});
var _crypto = /*#__PURE__*/ _interop_require_default(require("crypto"));
var _extractbaseiterator = require("extract-base-iterator");
var _createBufferingDecoderts = /*#__PURE__*/ _interop_require_default(require("./createBufferingDecoder.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
// Global password storage - set before decryption
var _password = null;
function setPassword(password) {
    _password = password;
}
function getPassword() {
    return _password;
}
/**
 * Parse AES properties to extract key derivation parameters
 */ function parseProperties(properties) {
    if (!properties || properties.length < 1) {
        throw new Error('AES: missing properties');
    }
    var b0 = properties[0];
    var numCyclesPower = b0 & 0x3f;
    // Check for special case: no salt/IV flags
    if ((b0 & 0xc0) === 0) {
        // No salt, no IV - use zeros
        return {
            numCyclesPower: numCyclesPower,
            salt: (0, _extractbaseiterator.allocBuffer)(0),
            iv: (0, _extractbaseiterator.allocBuffer)(16)
        };
    }
    if (properties.length < 2) {
        throw new Error('AES: properties too short');
    }
    var b1 = properties[1];
    // Calculate sizes
    // saltSize = ((b0 >> 7) & 1) + (b1 >> 4)
    // ivSize = ((b0 >> 6) & 1) + (b1 & 0x0F)
    var saltSize = (b0 >>> 7 & 1) + (b1 >>> 4);
    var ivSize = (b0 >>> 6 & 1) + (b1 & 0x0f);
    var expectedSize = 2 + saltSize + ivSize;
    if (properties.length < expectedSize) {
        throw new Error('AES: properties too short for salt/IV');
    }
    var salt = properties.slice(2, 2 + saltSize);
    var iv = (0, _extractbaseiterator.allocBuffer)(16);
    // Copy IV data (may be less than 16 bytes, rest is zeros)
    var ivData = properties.slice(2 + saltSize, 2 + saltSize + ivSize);
    for(var i = 0; i < ivData.length && i < 16; i++){
        iv[i] = ivData[i];
    }
    return {
        numCyclesPower: numCyclesPower,
        salt: salt,
        iv: iv
    };
}
/**
 * Convert password string to UTF-16LE buffer
 */ function passwordToUtf16LE(password) {
    var buf = (0, _extractbaseiterator.allocBuffer)(password.length * 2);
    for(var i = 0; i < password.length; i++){
        var code = password.charCodeAt(i);
        buf[i * 2] = code & 0xff;
        buf[i * 2 + 1] = code >>> 8 & 0xff;
    }
    return buf;
}
/**
 * Derive AES-256 key from password using 7z's SHA-256 iteration scheme
 *
 * Algorithm:
 *   For round = 0 to 2^numCyclesPower - 1:
 *     hash.update(salt)
 *     hash.update(password_utf16le)
 *     hash.update(round as 8-byte little-endian)
 *   key = hash.digest()
 */ function deriveKey(password, salt, numCyclesPower) {
    var passwordBuf = passwordToUtf16LE(password);
    var numRounds = Math.pow(2, numCyclesPower);
    // For special case 0x3F, don't iterate
    if (numCyclesPower === 0x3f) {
        // Direct concatenation mode
        var key = (0, _extractbaseiterator.allocBuffer)(32);
        var offset = 0;
        for(var j = 0; j < salt.length && offset < 32; j++){
            key[offset++] = salt[j];
        }
        for(var k = 0; k < passwordBuf.length && offset < 32; k++){
            key[offset++] = passwordBuf[k];
        }
        return key;
    }
    // Counter buffer (8 bytes, little-endian)
    var counter = (0, _extractbaseiterator.allocBuffer)(8);
    // Create hash and iterate
    var hash = _crypto.default.createHash('sha256');
    for(var round = 0; round < numRounds; round++){
        // Write round counter as little-endian 64-bit
        counter[0] = round & 0xff;
        counter[1] = round >>> 8 & 0xff;
        counter[2] = round >>> 16 & 0xff;
        counter[3] = round >>> 24 & 0xff;
        // Upper 32 bits - for large round counts
        var high = Math.floor(round / 0x100000000);
        counter[4] = high & 0xff;
        counter[5] = high >>> 8 & 0xff;
        counter[6] = high >>> 16 & 0xff;
        counter[7] = high >>> 24 & 0xff;
        hash.update(salt);
        hash.update(passwordBuf);
        hash.update(counter);
    }
    return hash.digest();
}
function decodeAes(input, properties, _unpackSize) {
    if (!_password) {
        throw new Error('AES: password required but not set');
    }
    if (!properties) {
        throw new Error('AES: properties required');
    }
    var params = parseProperties(properties);
    var key = deriveKey(_password, params.salt, params.numCyclesPower);
    // Create AES-256-CBC decipher
    var decipher = _crypto.default.createDecipheriv('aes-256-cbc', key, params.iv);
    decipher.setAutoPadding(false); // 7z doesn't use PKCS7 padding
    // Node 0.8 returns binary strings, newer Node returns Buffers
    // Use 'binary' encoding for compatibility
    // @ts-expect-error - 'binary' encoding is deprecated but required for Node 0.8 compatibility
    var decStr = decipher.update(input, 'binary', 'binary') + decipher.final('binary');
    var decrypted = (0, _extractbaseiterator.bufferFrom)(decStr, 'binary');
    return decrypted;
}
function createAesDecoder(properties, unpackSize) {
    return (0, _createBufferingDecoderts.default)(decodeAes, properties, unpackSize);
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
