// 7z-iterator/dist/cjs/sevenz/codecs/index.js
// Codec registry for 7z decompression
// Each codec provides a decode function and optionally a streaming decoder
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get decodeBcj2Multi () {
        return _Bcj2ts.decodeBcj2Multi;
    },
    get getCodec () {
        return getCodec;
    },
    get getCodecName () {
        return getCodecName;
    },
    get getPassword () {
        return _Aests.getPassword;
    },
    get isBcj2Codec () {
        return isBcj2Codec;
    },
    get isCodecSupported () {
        return isCodecSupported;
    },
    get registerCodec () {
        return registerCodec;
    },
    get setPassword () {
        return _Aests.setPassword;
    }
});
var _xzcompat = require("xz-compat");
var _constantsts = require("../constants.js");
var _Aests = require("./Aes.js");
var _Bcj2ts = require("./Bcj2.js");
var _BZip2ts = require("./BZip2.js");
var _Copyts = require("./Copy.js");
var _Deflatets = require("./Deflate.js");
var schedule = typeof setImmediate === 'function' ? setImmediate : function(fn) {
    return process.nextTick(fn);
};
function wrapSyncDecode(fn) {
    return function(input, properties, unpackSize, callback) {
        schedule(function() {
            try {
                // Convert BufferList to Buffer if needed (these simpler decoders need contiguous memory)
                var buf = Buffer.isBuffer(input) ? input : input.toBuffer();
                callback(null, fn(buf, properties, unpackSize));
            } catch (err) {
                callback(err);
            }
        });
    };
}
// Simple wrappers with validation that use xz-compat's optimized decode7zLzma/decode7zLzma2
function decodeLzma(input, properties, unpackSize, callback) {
    if (properties.length < 5) {
        throw new Error('LZMA requires 5-byte properties');
    }
    // Pass BufferLike directly - xz-compat's LZMA decoder handles both Buffer and BufferList
    (0, _xzcompat.decode7zLzma)(input, properties, unpackSize, callback);
}
function createLzmaDecoder(properties, unpackSize) {
    if (!properties || properties.length < 5) {
        throw new Error('LZMA requires 5-byte properties');
    }
    if (typeof unpackSize !== 'number' || unpackSize < 0) {
        throw new Error('LZMA requires known unpack size');
    }
    return (0, _xzcompat.createLzmaDecoder)(properties, unpackSize);
}
function decodeLzma2(input, properties, unpackSize, callback) {
    if (properties.length < 1) {
        throw new Error('LZMA2 requires properties byte');
    }
    // Pass BufferLike directly - xz-compat's LZMA2 decoder handles both Buffer and BufferList
    (0, _xzcompat.decode7zLzma2)(input, properties, unpackSize, callback);
}
function createLzma2Decoder(properties, _unpackSize) {
    if (!properties || properties.length < 1) {
        throw new Error('LZMA2 requires properties byte');
    }
    return (0, _xzcompat.createLzma2Decoder)(properties);
}
// Registry of supported codecs
var codecs = {};
/**
 * Convert codec ID bytes to a string key
 */ function codecIdToKey(id) {
    var parts = [];
    for(var i = 0; i < id.length; i++){
        parts.push(id[i].toString(16).toUpperCase());
    }
    return parts.join('-');
}
/**
 * Check if two codec IDs match
 */ function codecIdEquals(a, b) {
    if (a.length !== b.length) return false;
    for(var i = 0; i < a.length; i++){
        if (a[i] !== b[i]) return false;
    }
    return true;
}
function registerCodec(id, codec) {
    codecs[codecIdToKey(id)] = codec;
}
function getCodec(id) {
    var key = codecIdToKey(id);
    var codec = codecs[key];
    if (!codec) {
        throw (0, _constantsts.createCodedError)("Unsupported codec: ".concat(key), _constantsts.ErrorCode.UNSUPPORTED_CODEC);
    }
    return codec;
}
function isCodecSupported(id) {
    return codecs[codecIdToKey(id)] !== undefined;
}
function getCodecName(id) {
    if (codecIdEquals(id, _constantsts.CodecId.COPY)) return 'Copy';
    if (codecIdEquals(id, _constantsts.CodecId.LZMA)) return 'LZMA';
    if (codecIdEquals(id, _constantsts.CodecId.LZMA2)) return 'LZMA2';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ_X86)) return 'BCJ (x86)';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ_ARM)) return 'BCJ (ARM)';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ_ARMT)) return 'BCJ (ARM Thumb)';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ_ARM64)) return 'BCJ (ARM64)';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ_PPC)) return 'BCJ (PowerPC)';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ_IA64)) return 'BCJ (IA64)';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ_SPARC)) return 'BCJ (SPARC)';
    if (codecIdEquals(id, _constantsts.CodecId.BCJ2)) return 'BCJ2';
    if (codecIdEquals(id, _constantsts.CodecId.PPMD)) return 'PPMd';
    if (codecIdEquals(id, _constantsts.CodecId.DELTA)) return 'Delta';
    if (codecIdEquals(id, _constantsts.CodecId.DEFLATE)) return 'Deflate';
    if (codecIdEquals(id, _constantsts.CodecId.BZIP2)) return 'BZip2';
    if (codecIdEquals(id, _constantsts.CodecId.AES)) return 'AES-256';
    return "Unknown (".concat(codecIdToKey(id), ")");
}
function isBcj2Codec(id) {
    return codecIdEquals(id, _constantsts.CodecId.BCJ2);
}
// Register built-in codecs
// Copy codec (no compression)
registerCodec(_constantsts.CodecId.COPY, {
    decode: wrapSyncDecode(_Copyts.decodeCopy),
    createDecoder: _Copyts.createCopyDecoder
});
// LZMA codec
registerCodec(_constantsts.CodecId.LZMA, {
    decode: decodeLzma,
    createDecoder: createLzmaDecoder
});
// LZMA2 codec
registerCodec(_constantsts.CodecId.LZMA2, {
    decode: decodeLzma2,
    createDecoder: createLzma2Decoder
});
// BCJ (x86) filter
registerCodec(_constantsts.CodecId.BCJ_X86, {
    decode: wrapSyncDecode(_xzcompat.decodeBcj),
    createDecoder: _xzcompat.createBcjDecoder
});
// BCJ (ARM) filter
registerCodec(_constantsts.CodecId.BCJ_ARM, {
    decode: wrapSyncDecode(_xzcompat.decodeBcjArm),
    createDecoder: _xzcompat.createBcjArmDecoder
});
// BCJ (ARM Thumb) filter
registerCodec(_constantsts.CodecId.BCJ_ARMT, {
    decode: wrapSyncDecode(_xzcompat.decodeBcjArmt),
    createDecoder: _xzcompat.createBcjArmtDecoder
});
// BCJ (ARM64) filter
registerCodec(_constantsts.CodecId.BCJ_ARM64, {
    decode: wrapSyncDecode(_xzcompat.decodeBcjArm64),
    createDecoder: _xzcompat.createBcjArm64Decoder
});
// BCJ (PowerPC) filter
registerCodec(_constantsts.CodecId.BCJ_PPC, {
    decode: wrapSyncDecode(_xzcompat.decodeBcjPpc),
    createDecoder: _xzcompat.createBcjPpcDecoder
});
// BCJ (IA64) filter
registerCodec(_constantsts.CodecId.BCJ_IA64, {
    decode: wrapSyncDecode(_xzcompat.decodeBcjIa64),
    createDecoder: _xzcompat.createBcjIa64Decoder
});
// BCJ (SPARC) filter
registerCodec(_constantsts.CodecId.BCJ_SPARC, {
    decode: wrapSyncDecode(_xzcompat.decodeBcjSparc),
    createDecoder: _xzcompat.createBcjSparcDecoder
});
// Delta filter
registerCodec(_constantsts.CodecId.DELTA, {
    decode: wrapSyncDecode(_xzcompat.decodeDelta),
    createDecoder: _xzcompat.createDeltaDecoder
});
// Deflate codec
registerCodec(_constantsts.CodecId.DEFLATE, {
    decode: wrapSyncDecode(_Deflatets.decodeDeflate),
    createDecoder: _Deflatets.createDeflateDecoder
});
// BZip2 codec
registerCodec(_constantsts.CodecId.BZIP2, {
    decode: wrapSyncDecode(_BZip2ts.decodeBzip2),
    createDecoder: _BZip2ts.createBzip2Decoder
});
// AES-256-CBC codec (encryption)
registerCodec(_constantsts.CodecId.AES, {
    decode: wrapSyncDecode(_Aests.decodeAes),
    createDecoder: _Aests.createAesDecoder
});
// BCJ2 (x86-64) filter - multi-stream
// Note: BCJ2 requires special handling in SevenZipParser due to 4-stream architecture
registerCodec(_constantsts.CodecId.BCJ2, {
    decode: _Bcj2ts.decodeBcj2,
    createDecoder: _Bcj2ts.createBcj2Decoder
}); // Note: PPMd codec is not implemented. See FUTURE_ENHANCEMENTS.md
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
