// 7z-iterator/dist/cjs/sevenz/codecs/streams.js
// Shared stream wrappers for lzma-purejs codec interface
// These adapters convert between Buffer/lzma-purejs stream interfaces
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get createInputStream () {
        return createInputStream;
    },
    get createOutputStream () {
        return createOutputStream;
    }
});
var _extractbaseiterator = require("extract-base-iterator");
function createInputStream(buffer, offset, length) {
    var pos = 0;
    var end = Math.min(offset + length, buffer.length);
    var start = offset;
    return {
        readByte: function() {
            if (start + pos >= end) return -1;
            return buffer[start + pos++];
        },
        read: function(buf, bufOffset, len) {
            var bytesRead = 0;
            while(bytesRead < len && start + pos < end){
                buf[bufOffset + bytesRead] = buffer[start + pos];
                pos++;
                bytesRead++;
            }
            return bytesRead === 0 ? -1 : bytesRead;
        }
    };
}
function createOutputStream(expectedSize) {
    // Pre-allocation mode: single buffer, no concat needed
    // Includes bounds checking for safety on older Node.js versions
    if (expectedSize && expectedSize > 0) {
        var buffer = (0, _extractbaseiterator.allocBufferUnsafe)(expectedSize);
        var bufPos = 0;
        var bufLen = buffer.length;
        return {
            writeByte: function(b) {
                if (bufPos < bufLen) {
                    buffer[bufPos++] = b;
                }
            // Silently ignore overflow (should not happen with correct size)
            },
            write: function(buf, bufOffset, len) {
                for(var i = 0; i < len && bufPos < bufLen; i++){
                    buffer[bufPos++] = buf[bufOffset + i];
                }
                return len;
            },
            flush: function() {
            // No-op for pre-allocated buffer
            },
            toBuffer: function() {
                // Return only the used portion
                return bufPos < buffer.length ? buffer.slice(0, bufPos) : buffer;
            }
        };
    }
    // Chunked mode: accumulate in 64KB chunks (fallback for unknown size)
    var chunks = [];
    var CHUNK_SIZE = 65536; // 64KB chunks for better memory efficiency
    var currentChunk = (0, _extractbaseiterator.allocBufferUnsafe)(CHUNK_SIZE);
    var pos = 0;
    return {
        writeByte: function(b) {
            currentChunk[pos++] = b;
            if (pos >= CHUNK_SIZE) {
                chunks.push(currentChunk);
                currentChunk = (0, _extractbaseiterator.allocBufferUnsafe)(CHUNK_SIZE);
                pos = 0;
            }
        },
        write: function write(buf, bufOffset, len) {
            for(var i = 0; i < len; i++){
                this.writeByte(buf[bufOffset + i]);
            }
            return len;
        },
        flush: function() {
            if (pos > 0) {
                // Only keep the used portion of the current chunk
                chunks.push(currentChunk.slice(0, pos));
                currentChunk = (0, _extractbaseiterator.allocBufferUnsafe)(CHUNK_SIZE);
                pos = 0;
            }
        },
        toBuffer: function toBuffer() {
            this.flush();
            // Optimization: if single chunk, return it directly
            if (chunks.length === 1) {
                return chunks[0];
            }
            return Buffer.concat(chunks);
        }
    };
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
