// get-file-compat/dist/cjs/getContent.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return getContent;
    }
});
var _nodeversioncalllocal = require("node-version-call-local");
var _onone = /*#__PURE__*/ _interop_require_default(require("on-one"));
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _url = /*#__PURE__*/ _interop_require_default(require("url"));
var _compatts = require("./compat.js");
var _makeRequestts = /*#__PURE__*/ _interop_require_default(require("./lib/makeRequest.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var major = +process.versions.node.split('.')[0];
var minor = +process.versions.node.split('.')[1];
var noHTTPS = major === 0 && (minor <= 8 || minor === 12);
var __dirname = _path.default.dirname(typeof __filename === 'undefined' ? _url.default.fileURLToPath(require("url").pathToFileURL(__filename).toString()) : __filename);
var workerPath = _path.default.join(__dirname, '..', 'cjs', 'getContent.js');
function run(endpoint, encoding, options, callback) {
    (0, _makeRequestts.default)(endpoint, {
        timeout: options.timeout
    }, function(err, res) {
        if (err) return callback(err);
        var chunks = [];
        res.on('data', function(chunk) {
            chunks.push(chunk);
        });
        (0, _onone.default)(res, [
            'error',
            'end',
            'close',
            'finish'
        ], function(err) {
            if (err) return callback(err);
            var buffer = Buffer.concat(chunks);
            var content = encoding ? buffer.toString(encoding) : buffer;
            callback(null, {
                content: content,
                headers: res.headers,
                statusCode: res.statusCode
            });
        });
    });
}
// spawnOptions: false - no node/npm spawn (network only)
var call = (0, _nodeversioncalllocal.bind)('>0', workerPath, {
    callbacks: true,
    spawnOptions: false
});
function worker(endpoint, encoding, options, callback) {
    if (!noHTTPS) {
        run(endpoint, encoding, options, callback);
        return;
    }
    var enc = encoding || 'base64';
    call(endpoint, enc, options, function(err, result) {
        if (err) return callback(err);
        var content = encoding ? result.content : (0, _compatts.bufferFrom)(result.content, 'base64');
        callback(null, {
            content: content,
            headers: result.headers,
            statusCode: result.statusCode
        });
    });
}
function getContent(endpoint, encodingOrOptionsOrCallback, optionsOrCallback, callback) {
    var encoding = null;
    var options = {};
    var cb;
    // Parse arguments
    if (typeof encodingOrOptionsOrCallback === 'function') {
        cb = encodingOrOptionsOrCallback;
    } else if (typeof encodingOrOptionsOrCallback === 'string') {
        encoding = encodingOrOptionsOrCallback;
        if (typeof optionsOrCallback === 'function') {
            cb = optionsOrCallback;
        } else if (optionsOrCallback) {
            options = optionsOrCallback;
            cb = callback;
        }
    } else if (encodingOrOptionsOrCallback) {
        options = encodingOrOptionsOrCallback;
        if (typeof optionsOrCallback === 'function') {
            cb = optionsOrCallback;
        }
    }
    if (typeof cb === 'function') return worker(endpoint, encoding, options, cb);
    return new Promise(function(resolve, reject) {
        return worker(endpoint, encoding, options, function(err, result) {
            return err ? reject(err) : resolve(result);
        });
    });
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
