// get-file-compat/dist/cjs/getFile.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return getFile;
    }
});
var _fs = /*#__PURE__*/ _interop_require_wildcard(require("fs"));
var _mkdirpclassic = /*#__PURE__*/ _interop_require_default(require("mkdirp-classic"));
var _nodeversioncalllocal = require("node-version-call-local");
var _onone = /*#__PURE__*/ _interop_require_default(require("on-one"));
var _path = /*#__PURE__*/ _interop_require_wildcard(require("path"));
var _pump = /*#__PURE__*/ _interop_require_default(require("pump"));
var _url = /*#__PURE__*/ _interop_require_default(require("url"));
var _makeRequestts = /*#__PURE__*/ _interop_require_default(require("./lib/makeRequest.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
// node <= 0.8 does not support https and node 0.12 certs cannot be trusted
var major = +process.versions.node.split('.')[0];
var minor = +process.versions.node.split('.')[1];
var noHTTPS = major === 0 && (minor <= 8 || minor === 12);
var __dirname = _path.dirname(typeof __filename === 'undefined' ? _url.default.fileURLToPath(require("url").pathToFileURL(__filename).toString()) : __filename);
var workerPath = _path.join(__dirname, '..', 'cjs', 'getFile.js');
function run(endpoint, dest, options, callback) {
    (0, _makeRequestts.default)(endpoint, {
        timeout: options.timeout
    }, function(err, res) {
        if (err) return callback(err);
        (0, _mkdirpclassic.default)(_path.dirname(dest), function(err) {
            if (err && err.code !== 'EEXIST') return callback(err);
            var stream = (0, _pump.default)(res, _fs.createWriteStream(dest));
            (0, _onone.default)(stream, [
                'error',
                'end',
                'close',
                'finish'
            ], function(err) {
                err ? callback(err) : callback(null, {
                    path: dest,
                    headers: res.headers,
                    statusCode: res.statusCode
                });
            });
        });
    });
}
// spawnOptions: false - no node/npm spawn (network + fs only)
var worker = noHTTPS ? (0, _nodeversioncalllocal.bind)('>0', workerPath, {
    callbacks: true,
    spawnOptions: false
}) : run;
function getFile(endpoint, dest, optionsOrCallback, callback) {
    var options = typeof optionsOrCallback === 'function' ? {} : optionsOrCallback || {};
    var cb = typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
    if (typeof cb === 'function') return worker(endpoint, dest, options, cb);
    return new Promise(function(resolve, reject) {
        return worker(endpoint, dest, options, function(err, result) {
            return err ? reject(err) : resolve(result);
        });
    });
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
