// get-file-compat/dist/cjs/head.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return head;
    }
});
var _nodeversioncalllocal = require("node-version-call-local");
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _url = /*#__PURE__*/ _interop_require_default(require("url"));
var _makeRequestts = /*#__PURE__*/ _interop_require_default(require("./lib/makeRequest.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
// node <= 0.8 does not support https and node 0.12 certs cannot be trusted
var major = +process.versions.node.split('.')[0];
var minor = +process.versions.node.split('.')[1];
var noHTTPS = major === 0 && (minor <= 8 || minor === 12);
var __dirname = _path.default.dirname(typeof __filename === 'undefined' ? _url.default.fileURLToPath(require("url").pathToFileURL(__filename).toString()) : __filename);
var workerPath = _path.default.join(__dirname, '..', 'cjs', 'head.js');
function run(endpoint, options, callback) {
    (0, _makeRequestts.default)(endpoint, {
        method: 'HEAD',
        timeout: options.timeout
    }, function(err, res) {
        if (err) return callback(err);
        res.resume(); // Discard any body
        callback(null, {
            statusCode: res.statusCode,
            headers: res.headers
        });
    });
}
// spawnOptions: false - no node/npm spawn (network only)
var worker = noHTTPS ? (0, _nodeversioncalllocal.bind)('>0', workerPath, {
    callbacks: true,
    spawnOptions: false
}) : run;
function head(endpoint, optionsOrCallback, callback) {
    var options = typeof optionsOrCallback === 'function' ? {} : optionsOrCallback || {};
    var cb = typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
    if (typeof cb === 'function') return worker(endpoint, options, cb);
    return new Promise(function(resolve, reject) {
        return worker(endpoint, options, function(err, response) {
            return err ? reject(err) : resolve(response);
        });
    });
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
