// function-exec-sync/dist/cjs/workers/runFunction.cjs
// CJS worker - uses CJS build of module-compat
// Only supports loading CJS modules (or ESM on Node 23+ via require)
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var compat = require('async-compat');
var fs = require('fs');
var loadModule = require('module-compat').loadModule;
var path = require('path');
var serialize = require(path.join(__dirname, '..', 'serialize-javascript.js'));
var input = process.argv[2];
var output = process.argv[3];
function writeResult(result) {
    fs.writeFile(output, serialize(result), 'utf8', function() {
        process.exit(0);
    });
}
function writeError(error) {
    var errorObj = {
        message: error.message,
        stack: error.stack
    };
    for(var key in error)errorObj[key] = error[key];
    writeResult({
        error: errorObj
    });
}
// get data
try {
    // biome-ignore lint/security/noGlobalEval: Deserialize worker data
    var workerData = eval("(".concat(fs.readFileSync(input, 'utf8'), ")"));
    // set up env
    if (process.cwd() !== workerData.cwd) process.chdir(workerData.cwd);
    for(var key in workerData.env)process.env[key] = workerData.env[key];
    // load and call function
    var loadOptions = {
        moduleType: workerData.moduleType || 'auto',
        interop: workerData.interop || 'default'
    };
    loadModule(workerData.filePath, loadOptions, function(err, fn) {
        if (err) {
            writeError(err);
            return;
        }
        if (typeof fn !== 'function') {
            writeResult({
                value: fn
            });
        } else {
            var args = [
                fn,
                workerData.callbacks
            ].concat(workerData.args);
            args.push(function(err, value) {
                err ? writeError(err) : writeResult({
                    value: value
                });
            });
            compat.asyncFunction.apply(null, args);
        }
    });
} catch (err) {
    writeError(err);
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
