// function-exec-sync/dist/cjs/workers/runFunction.mjs
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _asynccompat = /*#__PURE__*/ _interop_require_default(require("async-compat"));
var _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
var _module = /*#__PURE__*/ _interop_require_default(require("module"));
var _modulecompat = require("module-compat");
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _url = /*#__PURE__*/ _interop_require_default(require("url"));
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_without_holes(arr) {
    if (Array.isArray(arr)) return _array_like_to_array(arr);
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _iterable_to_array(iter) {
    if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter);
}
function _non_iterable_spread() {
    throw new TypeError("Invalid attempt to spread non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _to_consumable_array(arr) {
    return _array_without_holes(arr) || _iterable_to_array(arr) || _unsupported_iterable_to_array(arr) || _non_iterable_spread();
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
// ESM worker - uses ESM build of module-compat
// Supports loading both CJS and ESM modules via import()
var _require = _module.default.createRequire(require("url").pathToFileURL(__filename).toString());
var __dirname = _path.default.dirname(_url.default.fileURLToPath(require("url").pathToFileURL(__filename).toString()));
// Require serialize from CJS dist (ESM dist files can't be required on Node 12-22)
var serialize = _require(_path.default.join(__dirname, '..', '..', 'cjs', 'serialize-javascript.js'));
var input = process.argv[2];
var output = process.argv[3];
function writeResult(result) {
    _fs.default.writeFile(output, serialize(result), 'utf8', function() {
        process.exit(0);
    });
}
function writeError(error) {
    var errorObj = {
        message: error.message,
        stack: error.stack
    };
    for(var key in error)errorObj[key] = error[key];
    writeResult({
        error: errorObj
    });
}
// get data
try {
    // biome-ignore lint/security/noGlobalEval: Deserialize worker data
    var workerData = eval("(".concat(_fs.default.readFileSync(input, 'utf8'), ")"));
    // set up env
    if (process.cwd() !== workerData.cwd) process.chdir(workerData.cwd);
    for(var key in workerData.env)process.env[key] = workerData.env[key];
    // load and call function
    var loadOptions = {
        moduleType: workerData.moduleType || 'auto',
        interop: workerData.interop || 'default'
    };
    (0, _modulecompat.loadModule)(workerData.filePath, loadOptions, function(err, fn) {
        if (err) {
            writeError(err);
            return;
        }
        if (typeof fn !== 'function') {
            writeResult({
                value: fn
            });
        } else {
            var args = [
                fn,
                workerData.callbacks
            ].concat(_to_consumable_array(workerData.args));
            args.push(function(err, value) {
                err ? writeError(err) : writeResult({
                    value: value
                });
            });
            _asynccompat.default.asyncFunction.apply(null, args);
        }
    });
} catch (err) {
    writeError(err);
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
