// cross-spawn-cb/dist/cjs/shim/index.js
/**
 * Minimal cross-spawn shim for Windows compatibility on legacy Node.js
 *
 * This replaces the bundled cross-spawn v6.0.5 (~1,950 lines) with a minimal
 * implementation (~200 lines) that uses external dependencies:
 * - which@^1.0.0 (command resolution)
 * - env-path-key@^1.0.0 (cross-platform PATH key detection)
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get _enoent () {
        return _enoent;
    },
    get _parse () {
        return _parsets.parse;
    },
    get // Main export
    default () {
        return _default;
    },
    get spawn () {
        return spawn;
    },
    get sync () {
        return spawnSync;
    }
});
var _child_process = /*#__PURE__*/ _interop_require_default(require("child_process"));
var _enoentts = require("./enoent.js");
var _parsets = require("./parse.js");
var _spawnSyncts = /*#__PURE__*/ _interop_require_default(require("./spawnSync.js"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function spawn(command, args, options) {
    var parsed = (0, _parsets.parse)(command, args, options);
    var spawned = _child_process.default.spawn(parsed.command, parsed.args, parsed.options);
    (0, _enoentts.hookChildProcess)(spawned, parsed);
    return spawned;
}
function spawnSync(command, args, options) {
    var parsed = (0, _parsets.parse)(command, args, options);
    var result = (0, _spawnSyncts.default)(parsed.command, parsed.args, parsed.options);
    result.error = result.error || (0, _enoentts.verifyENOENTSync)(result.status, parsed);
    return result;
}
var _default = spawn;
var _enoent = {
    hookChildProcess: _enoentts.hookChildProcess,
    verifyENOENT: _enoentts.verifyENOENT,
    verifyENOENTSync: _enoentts.verifyENOENTSync,
    notFoundError: _enoentts.notFoundError
};
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
