// cross-spawn-cb/dist/cjs/shim/parse.js
/**
 * Command parsing for cross-platform spawn
 * Simplified from cross-spawn v6.0.5
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "parse", {
    enumerable: true,
    get: function() {
        return parse;
    }
});
var _path = /*#__PURE__*/ _interop_require_default(require("path"));
var _constantsts = require("../constants.js");
var _escapets = require("./escape.js");
var _resolvets = require("./resolve.js");
var _shebangts = require("./shebang.js");
function _array_like_to_array(arr, len) {
    if (len == null || len > arr.length) len = arr.length;
    for(var i = 0, arr2 = new Array(len); i < len; i++)arr2[i] = arr[i];
    return arr2;
}
function _array_with_holes(arr) {
    if (Array.isArray(arr)) return arr;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _iterable_to_array_limit(arr, i) {
    var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"];
    if (_i == null) return;
    var _arr = [];
    var _n = true;
    var _d = false;
    var _s, _e;
    try {
        for(_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true){
            _arr.push(_s.value);
            if (i && _arr.length === i) break;
        }
    } catch (err) {
        _d = true;
        _e = err;
    } finally{
        try {
            if (!_n && _i["return"] != null) _i["return"]();
        } finally{
            if (_d) throw _e;
        }
    }
    return _arr;
}
function _non_iterable_rest() {
    throw new TypeError("Invalid attempt to destructure non-iterable instance.\\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function _sliced_to_array(arr, i) {
    return _array_with_holes(arr) || _iterable_to_array_limit(arr, i) || _unsupported_iterable_to_array(arr, i) || _non_iterable_rest();
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
function _unsupported_iterable_to_array(o, minLen) {
    if (!o) return;
    if (typeof o === "string") return _array_like_to_array(o, minLen);
    var n = Object.prototype.toString.call(o).slice(8, -1);
    if (n === "Object" && o.constructor) n = o.constructor.name;
    if (n === "Map" || n === "Set") return Array.from(n);
    if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _array_like_to_array(o, minLen);
}
var isExecutableRegExp = /\.(?:com|exe)$/i;
var isCmdShimRegExp = /node_modules[\\/].bin[\\/][^\\/]+\.cmd$/i;
// Node ^4.8.0 || ^5.7.0 || >= 6.0.0 supports shell option natively
// See: https://github.com/nodejs/node/pull/4598
var _process_versions_node_split_map = _sliced_to_array(process.versions.node.split('.').map(Number), 2), major = _process_versions_node_split_map[0], minor = _process_versions_node_split_map[1];
var supportsShellOption = major >= 6 || major === 5 && minor >= 7 || major === 4 && minor >= 8;
function detectShebang(parsed) {
    parsed.file = (0, _resolvets.resolveCommand)(parsed);
    var shebang = parsed.file && (0, _shebangts.readShebang)(parsed.file);
    if (shebang) {
        parsed.args.unshift(parsed.file);
        parsed.command = shebang;
        return (0, _resolvets.resolveCommand)(parsed);
    }
    return parsed.file;
}
function parseNonShell(parsed) {
    if (!_constantsts.isWindows) return parsed;
    // Detect & add support for shebangs
    var commandFile = detectShebang(parsed);
    // We don't need a shell if the command filename is an executable
    var needsShell = !isExecutableRegExp.test(commandFile || '');
    if (parsed.options.forceShell || needsShell) {
        // Need to double escape meta chars if the command is a cmd-shim
        var needsDoubleEscapeMetaChars = isCmdShimRegExp.test(commandFile || '');
        // Normalize posix paths into OS compatible paths
        parsed.command = _path.default.normalize(parsed.command);
        // Escape command & arguments
        parsed.command = (0, _escapets.escapeCommand)(parsed.command);
        parsed.args = parsed.args.map(function(arg) {
            return (0, _escapets.escapeArgument)(arg, needsDoubleEscapeMetaChars);
        });
        var shellCommand = [
            parsed.command
        ].concat(parsed.args).join(' ');
        parsed.args = [
            '/d',
            '/s',
            '/c',
            '"'.concat(shellCommand, '"')
        ];
        parsed.command = process.env.comspec || 'cmd.exe';
        parsed.options.windowsVerbatimArguments = true;
    }
    return parsed;
}
function parseShell(parsed) {
    // If node supports the shell option, there's no need to mimic its behavior
    if (supportsShellOption) {
        return parsed;
    }
    // Mimic node shell option for older Node versions
    var shellCommand = [
        parsed.command
    ].concat(parsed.args).join(' ');
    if (_constantsts.isWindows) {
        parsed.command = typeof parsed.options.shell === 'string' ? parsed.options.shell : process.env.comspec || 'cmd.exe';
        parsed.args = [
            '/d',
            '/s',
            '/c',
            '"'.concat(shellCommand, '"')
        ];
        parsed.options.windowsVerbatimArguments = true;
    } else {
        if (typeof parsed.options.shell === 'string') {
            parsed.command = parsed.options.shell;
        } else if (process.platform === 'android') {
            parsed.command = '/system/bin/sh';
        } else {
            parsed.command = '/bin/sh';
        }
        parsed.args = [
            '-c',
            shellCommand
        ];
    }
    return parsed;
}
// Handle NODE/npm_node_execpath env vars for node commands
var NODES = [
    'node',
    'node.exe',
    'node.cmd'
];
function parse(command, args, options) {
    options = (typeof args === "undefined" ? "undefined" : _type_of(args)) && !Array.isArray(args) ? {} : _object_spread({}, options || {});
    args = (typeof args === "undefined" ? "undefined" : _type_of(args)) && !Array.isArray(args) ? [] : args.slice(0) || [];
    // Build parsed object with original command/args BEFORE any substitution
    var parsed = {
        command: command,
        args: args,
        options: options,
        file: undefined,
        original: {
            command: command,
            args: args
        }
    };
    // Substitute node command with NODE env var if set
    if (NODES.indexOf(_path.default.basename(parsed.command).toLowerCase()) >= 0) {
        var env = options.env || process.env;
        parsed.command = env.NODE || env.npm_node_execpath || parsed.command;
    }
    return options.shell ? parseShell(parsed) : parseNonShell(parsed);
}
/* CJS INTEROP */ if (exports.__esModule && exports.default) { try { Object.defineProperty(exports.default, '__esModule', { value: true }); for (var key in exports) { exports.default[key] = exports[key]; } } catch (_) {}; module.exports = exports.default; }
